// @prisma/client/internal/prismaNamespaceBrowser.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * WARNING: This is an internal file that is subject to change!
 *
 * 🛑 Under no circumstances should you import this file directly! 🛑
 *
 * All exports from this file are wrapped under a `Prisma` namespace object in the browser.ts file.
 * While this enables partial backward compatibility, it is not part of the stable public API.
 *
 * If you are looking for your Models, Enums, and Input Types, please import them from the respective
 * model files in the `model` directory!
 */

import * as runtime from "@prisma/client/runtime/index-browser"

export type * from '../models'
export type * from './prismaNamespace'

export const Decimal = runtime.Decimal


export const NullTypes = {
  DbNull: runtime.NullTypes.DbNull as (new (secret: never) => typeof runtime.DbNull),
  JsonNull: runtime.NullTypes.JsonNull as (new (secret: never) => typeof runtime.JsonNull),
  AnyNull: runtime.NullTypes.AnyNull as (new (secret: never) => typeof runtime.AnyNull),
}
/**
 * Helper for filtering JSON entries that have `null` on the database (empty on the db)
 *
 * @see https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-on-a-json-field
 */
export const DbNull = runtime.DbNull

/**
 * Helper for filtering JSON entries that have JSON `null` values (not empty on the db)
 *
 * @see https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-on-a-json-field
 */
export const JsonNull = runtime.JsonNull

/**
 * Helper for filtering JSON entries that are `Prisma.DbNull` or `Prisma.JsonNull`
 *
 * @see https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-on-a-json-field
 */
export const AnyNull = runtime.AnyNull


export const ModelName = {
  User: 'User',
  Account: 'Account',
  Session: 'Session',
  VerificationToken: 'VerificationToken',
  Authenticator: 'Authenticator',
  Chama: 'Chama',
  ChamaLoan: 'ChamaLoan',
  Projects: 'Projects',
  Member: 'Member',
  BalanceSheet: 'BalanceSheet',
  Saving: 'Saving',
  Deduction: 'Deduction',
  LongTermLoan: 'LongTermLoan',
  LongTermLoanRepayment: 'LongTermLoanRepayment',
  ShortTermLoan: 'ShortTermLoan',
  Penalty: 'Penalty',
  Dividend: 'Dividend'
} as const

export type ModelName = (typeof ModelName)[keyof typeof ModelName]

/*
 * Enums
 */

export const UserScalarFieldEnum = {
  id: 'id',
  name: 'name',
  email: 'email',
  emailVerified: 'emailVerified',
  image: 'image',
  role: 'role',
  chamaId: 'chamaId',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type UserScalarFieldEnum = (typeof UserScalarFieldEnum)[keyof typeof UserScalarFieldEnum]


export const AccountScalarFieldEnum = {
  id: 'id',
  userId: 'userId',
  type: 'type',
  provider: 'provider',
  providerAccountId: 'providerAccountId',
  refresh_token: 'refresh_token',
  access_token: 'access_token',
  expires_at: 'expires_at',
  token_type: 'token_type',
  scope: 'scope',
  id_token: 'id_token',
  session_state: 'session_state',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type AccountScalarFieldEnum = (typeof AccountScalarFieldEnum)[keyof typeof AccountScalarFieldEnum]


export const SessionScalarFieldEnum = {
  id: 'id',
  sessionToken: 'sessionToken',
  userId: 'userId',
  expires: 'expires',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type SessionScalarFieldEnum = (typeof SessionScalarFieldEnum)[keyof typeof SessionScalarFieldEnum]


export const VerificationTokenScalarFieldEnum = {
  id: 'id',
  identifier: 'identifier',
  token: 'token',
  expires: 'expires'
} as const

export type VerificationTokenScalarFieldEnum = (typeof VerificationTokenScalarFieldEnum)[keyof typeof VerificationTokenScalarFieldEnum]


export const AuthenticatorScalarFieldEnum = {
  credentialID: 'credentialID',
  userId: 'userId',
  providerAccountId: 'providerAccountId',
  credentialPublicKey: 'credentialPublicKey',
  counter: 'counter',
  credentialDeviceType: 'credentialDeviceType',
  credentialBackedUp: 'credentialBackedUp',
  transports: 'transports'
} as const

export type AuthenticatorScalarFieldEnum = (typeof AuthenticatorScalarFieldEnum)[keyof typeof AuthenticatorScalarFieldEnum]


export const ChamaScalarFieldEnum = {
  id: 'id',
  name: 'name',
  location: 'location',
  profilePic: 'profilePic',
  registrationDocument: 'registrationDocument',
  minimumSavings: 'minimumSavings',
  nextMeeting: 'nextMeeting',
  registrationId: 'registrationId',
  adminId: 'adminId',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type ChamaScalarFieldEnum = (typeof ChamaScalarFieldEnum)[keyof typeof ChamaScalarFieldEnum]


export const ChamaLoanScalarFieldEnum = {
  id: 'id',
  chamaId: 'chamaId',
  lender: 'lender',
  loanAmount: 'loanAmount',
  interest: 'interest',
  termMonths: 'termMonths',
  dueDate: 'dueDate',
  paymentDate: 'paymentDate',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type ChamaLoanScalarFieldEnum = (typeof ChamaLoanScalarFieldEnum)[keyof typeof ChamaLoanScalarFieldEnum]


export const ProjectsScalarFieldEnum = {
  id: 'id',
  chamaId: 'chamaId',
  projectName: 'projectName',
  estimatedCost: 'estimatedCost',
  actualCost: 'actualCost',
  sharesUsed: 'sharesUsed',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type ProjectsScalarFieldEnum = (typeof ProjectsScalarFieldEnum)[keyof typeof ProjectsScalarFieldEnum]


export const MemberScalarFieldEnum = {
  id: 'id',
  chamaId: 'chamaId',
  userId: 'userId',
  name: 'name',
  email: 'email',
  phoneNumber: 'phoneNumber',
  role: 'role',
  lastSeen: 'lastSeen',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type MemberScalarFieldEnum = (typeof MemberScalarFieldEnum)[keyof typeof MemberScalarFieldEnum]


export const BalanceSheetScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  totalSavings: 'totalSavings',
  totalLongTermLoan: 'totalLongTermLoan',
  totalShortTermLoan: 'totalShortTermLoan',
  totalDividend: 'totalDividend',
  totalPenalties: 'totalPenalties',
  totalUnpaid: 'totalUnpaid',
  fixedSavings: 'fixedSavings',
  welfare: 'welfare',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type BalanceSheetScalarFieldEnum = (typeof BalanceSheetScalarFieldEnum)[keyof typeof BalanceSheetScalarFieldEnum]


export const SavingScalarFieldEnum = {
  id: 'id',
  amount: 'amount',
  mPesaRef: 'mPesaRef',
  authorizedBy: 'authorizedBy',
  memberId: 'memberId',
  deductionId: 'deductionId',
  savings: 'savings',
  welfare: 'welfare',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type SavingScalarFieldEnum = (typeof SavingScalarFieldEnum)[keyof typeof SavingScalarFieldEnum]


export const DeductionScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  longTermLoanId: 'longTermLoanId',
  shortTermLoanId: 'shortTermLoanId',
  penaltyId: 'penaltyId',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type DeductionScalarFieldEnum = (typeof DeductionScalarFieldEnum)[keyof typeof DeductionScalarFieldEnum]


export const LongTermLoanScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  penaltyId: 'penaltyId',
  principle: 'principle',
  loanAmount: 'loanAmount',
  interest: 'interest',
  termMonths: 'termMonths',
  dueDate: 'dueDate',
  paymentDate: 'paymentDate',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type LongTermLoanScalarFieldEnum = (typeof LongTermLoanScalarFieldEnum)[keyof typeof LongTermLoanScalarFieldEnum]


export const LongTermLoanRepaymentScalarFieldEnum = {
  id: 'id',
  longTermLoanId: 'longTermLoanId',
  dueDate: 'dueDate',
  paymentDate: 'paymentDate',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type LongTermLoanRepaymentScalarFieldEnum = (typeof LongTermLoanRepaymentScalarFieldEnum)[keyof typeof LongTermLoanRepaymentScalarFieldEnum]


export const ShortTermLoanScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  penaltyId: 'penaltyId',
  principle: 'principle',
  loanAmount: 'loanAmount',
  interest: 'interest',
  termMonths: 'termMonths',
  dueDate: 'dueDate',
  paymentDate: 'paymentDate',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type ShortTermLoanScalarFieldEnum = (typeof ShortTermLoanScalarFieldEnum)[keyof typeof ShortTermLoanScalarFieldEnum]


export const PenaltyScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  penaltyAmount: 'penaltyAmount',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type PenaltyScalarFieldEnum = (typeof PenaltyScalarFieldEnum)[keyof typeof PenaltyScalarFieldEnum]


export const DividendScalarFieldEnum = {
  id: 'id',
  memberId: 'memberId',
  dividendAmount: 'dividendAmount',
  dueDate: 'dueDate',
  paidAt: 'paidAt',
  status: 'status',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt'
} as const

export type DividendScalarFieldEnum = (typeof DividendScalarFieldEnum)[keyof typeof DividendScalarFieldEnum]


export const SortOrder = {
  asc: 'asc',
  desc: 'desc'
} as const

export type SortOrder = (typeof SortOrder)[keyof typeof SortOrder]


export const QueryMode = {
  default: 'default',
  insensitive: 'insensitive'
} as const

export type QueryMode = (typeof QueryMode)[keyof typeof QueryMode]
