// @prisma/client/models/Chama.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `Chama` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model Chama
 * 
 */
export type ChamaModel = runtime.Types.Result.DefaultSelection<Prisma.$ChamaPayload>

export type AggregateChama = {
  _count: ChamaCountAggregateOutputType | null
  _avg: ChamaAvgAggregateOutputType | null
  _sum: ChamaSumAggregateOutputType | null
  _min: ChamaMinAggregateOutputType | null
  _max: ChamaMaxAggregateOutputType | null
}

export type ChamaAvgAggregateOutputType = {
  minimumSavings: number | null
}

export type ChamaSumAggregateOutputType = {
  minimumSavings: number | null
}

export type ChamaMinAggregateOutputType = {
  id: string | null
  name: string | null
  location: string | null
  profilePic: string | null
  registrationDocument: string | null
  minimumSavings: number | null
  nextMeeting: Date | null
  registrationId: string | null
  adminId: string | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type ChamaMaxAggregateOutputType = {
  id: string | null
  name: string | null
  location: string | null
  profilePic: string | null
  registrationDocument: string | null
  minimumSavings: number | null
  nextMeeting: Date | null
  registrationId: string | null
  adminId: string | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type ChamaCountAggregateOutputType = {
  id: number
  name: number
  location: number
  profilePic: number
  registrationDocument: number
  minimumSavings: number
  nextMeeting: number
  registrationId: number
  adminId: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type ChamaAvgAggregateInputType = {
  minimumSavings?: true
}

export type ChamaSumAggregateInputType = {
  minimumSavings?: true
}

export type ChamaMinAggregateInputType = {
  id?: true
  name?: true
  location?: true
  profilePic?: true
  registrationDocument?: true
  minimumSavings?: true
  nextMeeting?: true
  registrationId?: true
  adminId?: true
  createdAt?: true
  updatedAt?: true
}

export type ChamaMaxAggregateInputType = {
  id?: true
  name?: true
  location?: true
  profilePic?: true
  registrationDocument?: true
  minimumSavings?: true
  nextMeeting?: true
  registrationId?: true
  adminId?: true
  createdAt?: true
  updatedAt?: true
}

export type ChamaCountAggregateInputType = {
  id?: true
  name?: true
  location?: true
  profilePic?: true
  registrationDocument?: true
  minimumSavings?: true
  nextMeeting?: true
  registrationId?: true
  adminId?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type ChamaAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Chama to aggregate.
   */
  where?: Prisma.ChamaWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Chamas to fetch.
   */
  orderBy?: Prisma.ChamaOrderByWithRelationInput | Prisma.ChamaOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.ChamaWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Chamas from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Chamas.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned Chamas
  **/
  _count?: true | ChamaCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: ChamaAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: ChamaSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: ChamaMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: ChamaMaxAggregateInputType
}

export type GetChamaAggregateType<T extends ChamaAggregateArgs> = {
      [P in keyof T & keyof AggregateChama]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateChama[P]>
    : Prisma.GetScalarType<T[P], AggregateChama[P]>
}




export type ChamaGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ChamaWhereInput
  orderBy?: Prisma.ChamaOrderByWithAggregationInput | Prisma.ChamaOrderByWithAggregationInput[]
  by: Prisma.ChamaScalarFieldEnum[] | Prisma.ChamaScalarFieldEnum
  having?: Prisma.ChamaScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: ChamaCountAggregateInputType | true
  _avg?: ChamaAvgAggregateInputType
  _sum?: ChamaSumAggregateInputType
  _min?: ChamaMinAggregateInputType
  _max?: ChamaMaxAggregateInputType
}

export type ChamaGroupByOutputType = {
  id: string
  name: string
  location: string
  profilePic: string | null
  registrationDocument: string | null
  minimumSavings: number | null
  nextMeeting: Date | null
  registrationId: string | null
  adminId: string
  createdAt: Date
  updatedAt: Date
  _count: ChamaCountAggregateOutputType | null
  _avg: ChamaAvgAggregateOutputType | null
  _sum: ChamaSumAggregateOutputType | null
  _min: ChamaMinAggregateOutputType | null
  _max: ChamaMaxAggregateOutputType | null
}

type GetChamaGroupByPayload<T extends ChamaGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<ChamaGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof ChamaGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], ChamaGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], ChamaGroupByOutputType[P]>
      }
    >
  >



export type ChamaWhereInput = {
  AND?: Prisma.ChamaWhereInput | Prisma.ChamaWhereInput[]
  OR?: Prisma.ChamaWhereInput[]
  NOT?: Prisma.ChamaWhereInput | Prisma.ChamaWhereInput[]
  id?: Prisma.StringFilter<"Chama"> | string
  name?: Prisma.StringFilter<"Chama"> | string
  location?: Prisma.StringFilter<"Chama"> | string
  profilePic?: Prisma.StringNullableFilter<"Chama"> | string | null
  registrationDocument?: Prisma.StringNullableFilter<"Chama"> | string | null
  minimumSavings?: Prisma.FloatNullableFilter<"Chama"> | number | null
  nextMeeting?: Prisma.DateTimeNullableFilter<"Chama"> | Date | string | null
  registrationId?: Prisma.StringNullableFilter<"Chama"> | string | null
  adminId?: Prisma.StringFilter<"Chama"> | string
  createdAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
  admin?: Prisma.XOR<Prisma.UserScalarRelationFilter, Prisma.UserWhereInput>
  members?: Prisma.MemberListRelationFilter
  loans?: Prisma.ChamaLoanListRelationFilter
  projects?: Prisma.ProjectsListRelationFilter
}

export type ChamaOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  location?: Prisma.SortOrder
  profilePic?: Prisma.SortOrder
  registrationDocument?: Prisma.SortOrder
  minimumSavings?: Prisma.SortOrder
  nextMeeting?: Prisma.SortOrder
  registrationId?: Prisma.SortOrder
  adminId?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  admin?: Prisma.UserOrderByWithRelationInput
  members?: Prisma.MemberOrderByRelationAggregateInput
  loans?: Prisma.ChamaLoanOrderByRelationAggregateInput
  projects?: Prisma.ProjectsOrderByRelationAggregateInput
}

export type ChamaWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.ChamaWhereInput | Prisma.ChamaWhereInput[]
  OR?: Prisma.ChamaWhereInput[]
  NOT?: Prisma.ChamaWhereInput | Prisma.ChamaWhereInput[]
  name?: Prisma.StringFilter<"Chama"> | string
  location?: Prisma.StringFilter<"Chama"> | string
  profilePic?: Prisma.StringNullableFilter<"Chama"> | string | null
  registrationDocument?: Prisma.StringNullableFilter<"Chama"> | string | null
  minimumSavings?: Prisma.FloatNullableFilter<"Chama"> | number | null
  nextMeeting?: Prisma.DateTimeNullableFilter<"Chama"> | Date | string | null
  registrationId?: Prisma.StringNullableFilter<"Chama"> | string | null
  adminId?: Prisma.StringFilter<"Chama"> | string
  createdAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
  admin?: Prisma.XOR<Prisma.UserScalarRelationFilter, Prisma.UserWhereInput>
  members?: Prisma.MemberListRelationFilter
  loans?: Prisma.ChamaLoanListRelationFilter
  projects?: Prisma.ProjectsListRelationFilter
}, "id">

export type ChamaOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  location?: Prisma.SortOrder
  profilePic?: Prisma.SortOrder
  registrationDocument?: Prisma.SortOrder
  minimumSavings?: Prisma.SortOrder
  nextMeeting?: Prisma.SortOrder
  registrationId?: Prisma.SortOrder
  adminId?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.ChamaCountOrderByAggregateInput
  _avg?: Prisma.ChamaAvgOrderByAggregateInput
  _max?: Prisma.ChamaMaxOrderByAggregateInput
  _min?: Prisma.ChamaMinOrderByAggregateInput
  _sum?: Prisma.ChamaSumOrderByAggregateInput
}

export type ChamaScalarWhereWithAggregatesInput = {
  AND?: Prisma.ChamaScalarWhereWithAggregatesInput | Prisma.ChamaScalarWhereWithAggregatesInput[]
  OR?: Prisma.ChamaScalarWhereWithAggregatesInput[]
  NOT?: Prisma.ChamaScalarWhereWithAggregatesInput | Prisma.ChamaScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"Chama"> | string
  name?: Prisma.StringWithAggregatesFilter<"Chama"> | string
  location?: Prisma.StringWithAggregatesFilter<"Chama"> | string
  profilePic?: Prisma.StringNullableWithAggregatesFilter<"Chama"> | string | null
  registrationDocument?: Prisma.StringNullableWithAggregatesFilter<"Chama"> | string | null
  minimumSavings?: Prisma.FloatNullableWithAggregatesFilter<"Chama"> | number | null
  nextMeeting?: Prisma.DateTimeNullableWithAggregatesFilter<"Chama"> | Date | string | null
  registrationId?: Prisma.StringNullableWithAggregatesFilter<"Chama"> | string | null
  adminId?: Prisma.StringWithAggregatesFilter<"Chama"> | string
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"Chama"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"Chama"> | Date | string
}

export type ChamaCreateInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  admin: Prisma.UserCreateNestedOneWithoutChamaInput
  members?: Prisma.MemberCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsCreateNestedManyWithoutChamaInput
}

export type ChamaUncheckedCreateInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  adminId: string
  createdAt?: Date | string
  updatedAt?: Date | string
  members?: Prisma.MemberUncheckedCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanUncheckedCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsUncheckedCreateNestedManyWithoutChamaInput
}

export type ChamaUpdateInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  admin?: Prisma.UserUpdateOneRequiredWithoutChamaNestedInput
  members?: Prisma.MemberUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  adminId?: Prisma.StringFieldUpdateOperationsInput | string
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  members?: Prisma.MemberUncheckedUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUncheckedUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUncheckedUpdateManyWithoutChamaNestedInput
}

export type ChamaCreateManyInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  adminId: string
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ChamaUpdateManyMutationInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ChamaUncheckedUpdateManyInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  adminId?: Prisma.StringFieldUpdateOperationsInput | string
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ChamaListRelationFilter = {
  every?: Prisma.ChamaWhereInput
  some?: Prisma.ChamaWhereInput
  none?: Prisma.ChamaWhereInput
}

export type ChamaOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type ChamaCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  location?: Prisma.SortOrder
  profilePic?: Prisma.SortOrder
  registrationDocument?: Prisma.SortOrder
  minimumSavings?: Prisma.SortOrder
  nextMeeting?: Prisma.SortOrder
  registrationId?: Prisma.SortOrder
  adminId?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ChamaAvgOrderByAggregateInput = {
  minimumSavings?: Prisma.SortOrder
}

export type ChamaMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  location?: Prisma.SortOrder
  profilePic?: Prisma.SortOrder
  registrationDocument?: Prisma.SortOrder
  minimumSavings?: Prisma.SortOrder
  nextMeeting?: Prisma.SortOrder
  registrationId?: Prisma.SortOrder
  adminId?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ChamaMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  name?: Prisma.SortOrder
  location?: Prisma.SortOrder
  profilePic?: Prisma.SortOrder
  registrationDocument?: Prisma.SortOrder
  minimumSavings?: Prisma.SortOrder
  nextMeeting?: Prisma.SortOrder
  registrationId?: Prisma.SortOrder
  adminId?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ChamaSumOrderByAggregateInput = {
  minimumSavings?: Prisma.SortOrder
}

export type ChamaScalarRelationFilter = {
  is?: Prisma.ChamaWhereInput
  isNot?: Prisma.ChamaWhereInput
}

export type ChamaCreateNestedManyWithoutAdminInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput> | Prisma.ChamaCreateWithoutAdminInput[] | Prisma.ChamaUncheckedCreateWithoutAdminInput[]
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutAdminInput | Prisma.ChamaCreateOrConnectWithoutAdminInput[]
  createMany?: Prisma.ChamaCreateManyAdminInputEnvelope
  connect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
}

export type ChamaUncheckedCreateNestedManyWithoutAdminInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput> | Prisma.ChamaCreateWithoutAdminInput[] | Prisma.ChamaUncheckedCreateWithoutAdminInput[]
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutAdminInput | Prisma.ChamaCreateOrConnectWithoutAdminInput[]
  createMany?: Prisma.ChamaCreateManyAdminInputEnvelope
  connect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
}

export type ChamaUpdateManyWithoutAdminNestedInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput> | Prisma.ChamaCreateWithoutAdminInput[] | Prisma.ChamaUncheckedCreateWithoutAdminInput[]
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutAdminInput | Prisma.ChamaCreateOrConnectWithoutAdminInput[]
  upsert?: Prisma.ChamaUpsertWithWhereUniqueWithoutAdminInput | Prisma.ChamaUpsertWithWhereUniqueWithoutAdminInput[]
  createMany?: Prisma.ChamaCreateManyAdminInputEnvelope
  set?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  disconnect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  delete?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  connect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  update?: Prisma.ChamaUpdateWithWhereUniqueWithoutAdminInput | Prisma.ChamaUpdateWithWhereUniqueWithoutAdminInput[]
  updateMany?: Prisma.ChamaUpdateManyWithWhereWithoutAdminInput | Prisma.ChamaUpdateManyWithWhereWithoutAdminInput[]
  deleteMany?: Prisma.ChamaScalarWhereInput | Prisma.ChamaScalarWhereInput[]
}

export type ChamaUncheckedUpdateManyWithoutAdminNestedInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput> | Prisma.ChamaCreateWithoutAdminInput[] | Prisma.ChamaUncheckedCreateWithoutAdminInput[]
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutAdminInput | Prisma.ChamaCreateOrConnectWithoutAdminInput[]
  upsert?: Prisma.ChamaUpsertWithWhereUniqueWithoutAdminInput | Prisma.ChamaUpsertWithWhereUniqueWithoutAdminInput[]
  createMany?: Prisma.ChamaCreateManyAdminInputEnvelope
  set?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  disconnect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  delete?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  connect?: Prisma.ChamaWhereUniqueInput | Prisma.ChamaWhereUniqueInput[]
  update?: Prisma.ChamaUpdateWithWhereUniqueWithoutAdminInput | Prisma.ChamaUpdateWithWhereUniqueWithoutAdminInput[]
  updateMany?: Prisma.ChamaUpdateManyWithWhereWithoutAdminInput | Prisma.ChamaUpdateManyWithWhereWithoutAdminInput[]
  deleteMany?: Prisma.ChamaScalarWhereInput | Prisma.ChamaScalarWhereInput[]
}

export type NullableFloatFieldUpdateOperationsInput = {
  set?: number | null
  increment?: number
  decrement?: number
  multiply?: number
  divide?: number
  unset?: boolean
}

export type ChamaCreateNestedOneWithoutLoansInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutLoansInput, Prisma.ChamaUncheckedCreateWithoutLoansInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutLoansInput
  connect?: Prisma.ChamaWhereUniqueInput
}

export type ChamaUpdateOneRequiredWithoutLoansNestedInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutLoansInput, Prisma.ChamaUncheckedCreateWithoutLoansInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutLoansInput
  upsert?: Prisma.ChamaUpsertWithoutLoansInput
  connect?: Prisma.ChamaWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.ChamaUpdateToOneWithWhereWithoutLoansInput, Prisma.ChamaUpdateWithoutLoansInput>, Prisma.ChamaUncheckedUpdateWithoutLoansInput>
}

export type ChamaCreateNestedOneWithoutProjectsInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutProjectsInput, Prisma.ChamaUncheckedCreateWithoutProjectsInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutProjectsInput
  connect?: Prisma.ChamaWhereUniqueInput
}

export type ChamaUpdateOneRequiredWithoutProjectsNestedInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutProjectsInput, Prisma.ChamaUncheckedCreateWithoutProjectsInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutProjectsInput
  upsert?: Prisma.ChamaUpsertWithoutProjectsInput
  connect?: Prisma.ChamaWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.ChamaUpdateToOneWithWhereWithoutProjectsInput, Prisma.ChamaUpdateWithoutProjectsInput>, Prisma.ChamaUncheckedUpdateWithoutProjectsInput>
}

export type ChamaCreateNestedOneWithoutMembersInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutMembersInput, Prisma.ChamaUncheckedCreateWithoutMembersInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutMembersInput
  connect?: Prisma.ChamaWhereUniqueInput
}

export type ChamaUpdateOneRequiredWithoutMembersNestedInput = {
  create?: Prisma.XOR<Prisma.ChamaCreateWithoutMembersInput, Prisma.ChamaUncheckedCreateWithoutMembersInput>
  connectOrCreate?: Prisma.ChamaCreateOrConnectWithoutMembersInput
  upsert?: Prisma.ChamaUpsertWithoutMembersInput
  connect?: Prisma.ChamaWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.ChamaUpdateToOneWithWhereWithoutMembersInput, Prisma.ChamaUpdateWithoutMembersInput>, Prisma.ChamaUncheckedUpdateWithoutMembersInput>
}

export type ChamaCreateWithoutAdminInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  members?: Prisma.MemberCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsCreateNestedManyWithoutChamaInput
}

export type ChamaUncheckedCreateWithoutAdminInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  members?: Prisma.MemberUncheckedCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanUncheckedCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsUncheckedCreateNestedManyWithoutChamaInput
}

export type ChamaCreateOrConnectWithoutAdminInput = {
  where: Prisma.ChamaWhereUniqueInput
  create: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput>
}

export type ChamaCreateManyAdminInputEnvelope = {
  data: Prisma.ChamaCreateManyAdminInput | Prisma.ChamaCreateManyAdminInput[]
}

export type ChamaUpsertWithWhereUniqueWithoutAdminInput = {
  where: Prisma.ChamaWhereUniqueInput
  update: Prisma.XOR<Prisma.ChamaUpdateWithoutAdminInput, Prisma.ChamaUncheckedUpdateWithoutAdminInput>
  create: Prisma.XOR<Prisma.ChamaCreateWithoutAdminInput, Prisma.ChamaUncheckedCreateWithoutAdminInput>
}

export type ChamaUpdateWithWhereUniqueWithoutAdminInput = {
  where: Prisma.ChamaWhereUniqueInput
  data: Prisma.XOR<Prisma.ChamaUpdateWithoutAdminInput, Prisma.ChamaUncheckedUpdateWithoutAdminInput>
}

export type ChamaUpdateManyWithWhereWithoutAdminInput = {
  where: Prisma.ChamaScalarWhereInput
  data: Prisma.XOR<Prisma.ChamaUpdateManyMutationInput, Prisma.ChamaUncheckedUpdateManyWithoutAdminInput>
}

export type ChamaScalarWhereInput = {
  AND?: Prisma.ChamaScalarWhereInput | Prisma.ChamaScalarWhereInput[]
  OR?: Prisma.ChamaScalarWhereInput[]
  NOT?: Prisma.ChamaScalarWhereInput | Prisma.ChamaScalarWhereInput[]
  id?: Prisma.StringFilter<"Chama"> | string
  name?: Prisma.StringFilter<"Chama"> | string
  location?: Prisma.StringFilter<"Chama"> | string
  profilePic?: Prisma.StringNullableFilter<"Chama"> | string | null
  registrationDocument?: Prisma.StringNullableFilter<"Chama"> | string | null
  minimumSavings?: Prisma.FloatNullableFilter<"Chama"> | number | null
  nextMeeting?: Prisma.DateTimeNullableFilter<"Chama"> | Date | string | null
  registrationId?: Prisma.StringNullableFilter<"Chama"> | string | null
  adminId?: Prisma.StringFilter<"Chama"> | string
  createdAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Chama"> | Date | string
}

export type ChamaCreateWithoutLoansInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  admin: Prisma.UserCreateNestedOneWithoutChamaInput
  members?: Prisma.MemberCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsCreateNestedManyWithoutChamaInput
}

export type ChamaUncheckedCreateWithoutLoansInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  adminId: string
  createdAt?: Date | string
  updatedAt?: Date | string
  members?: Prisma.MemberUncheckedCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsUncheckedCreateNestedManyWithoutChamaInput
}

export type ChamaCreateOrConnectWithoutLoansInput = {
  where: Prisma.ChamaWhereUniqueInput
  create: Prisma.XOR<Prisma.ChamaCreateWithoutLoansInput, Prisma.ChamaUncheckedCreateWithoutLoansInput>
}

export type ChamaUpsertWithoutLoansInput = {
  update: Prisma.XOR<Prisma.ChamaUpdateWithoutLoansInput, Prisma.ChamaUncheckedUpdateWithoutLoansInput>
  create: Prisma.XOR<Prisma.ChamaCreateWithoutLoansInput, Prisma.ChamaUncheckedCreateWithoutLoansInput>
  where?: Prisma.ChamaWhereInput
}

export type ChamaUpdateToOneWithWhereWithoutLoansInput = {
  where?: Prisma.ChamaWhereInput
  data: Prisma.XOR<Prisma.ChamaUpdateWithoutLoansInput, Prisma.ChamaUncheckedUpdateWithoutLoansInput>
}

export type ChamaUpdateWithoutLoansInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  admin?: Prisma.UserUpdateOneRequiredWithoutChamaNestedInput
  members?: Prisma.MemberUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateWithoutLoansInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  adminId?: Prisma.StringFieldUpdateOperationsInput | string
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  members?: Prisma.MemberUncheckedUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUncheckedUpdateManyWithoutChamaNestedInput
}

export type ChamaCreateWithoutProjectsInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  admin: Prisma.UserCreateNestedOneWithoutChamaInput
  members?: Prisma.MemberCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanCreateNestedManyWithoutChamaInput
}

export type ChamaUncheckedCreateWithoutProjectsInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  adminId: string
  createdAt?: Date | string
  updatedAt?: Date | string
  members?: Prisma.MemberUncheckedCreateNestedManyWithoutChamaInput
  loans?: Prisma.ChamaLoanUncheckedCreateNestedManyWithoutChamaInput
}

export type ChamaCreateOrConnectWithoutProjectsInput = {
  where: Prisma.ChamaWhereUniqueInput
  create: Prisma.XOR<Prisma.ChamaCreateWithoutProjectsInput, Prisma.ChamaUncheckedCreateWithoutProjectsInput>
}

export type ChamaUpsertWithoutProjectsInput = {
  update: Prisma.XOR<Prisma.ChamaUpdateWithoutProjectsInput, Prisma.ChamaUncheckedUpdateWithoutProjectsInput>
  create: Prisma.XOR<Prisma.ChamaCreateWithoutProjectsInput, Prisma.ChamaUncheckedCreateWithoutProjectsInput>
  where?: Prisma.ChamaWhereInput
}

export type ChamaUpdateToOneWithWhereWithoutProjectsInput = {
  where?: Prisma.ChamaWhereInput
  data: Prisma.XOR<Prisma.ChamaUpdateWithoutProjectsInput, Prisma.ChamaUncheckedUpdateWithoutProjectsInput>
}

export type ChamaUpdateWithoutProjectsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  admin?: Prisma.UserUpdateOneRequiredWithoutChamaNestedInput
  members?: Prisma.MemberUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateWithoutProjectsInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  adminId?: Prisma.StringFieldUpdateOperationsInput | string
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  members?: Prisma.MemberUncheckedUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUncheckedUpdateManyWithoutChamaNestedInput
}

export type ChamaCreateWithoutMembersInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
  admin: Prisma.UserCreateNestedOneWithoutChamaInput
  loans?: Prisma.ChamaLoanCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsCreateNestedManyWithoutChamaInput
}

export type ChamaUncheckedCreateWithoutMembersInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  adminId: string
  createdAt?: Date | string
  updatedAt?: Date | string
  loans?: Prisma.ChamaLoanUncheckedCreateNestedManyWithoutChamaInput
  projects?: Prisma.ProjectsUncheckedCreateNestedManyWithoutChamaInput
}

export type ChamaCreateOrConnectWithoutMembersInput = {
  where: Prisma.ChamaWhereUniqueInput
  create: Prisma.XOR<Prisma.ChamaCreateWithoutMembersInput, Prisma.ChamaUncheckedCreateWithoutMembersInput>
}

export type ChamaUpsertWithoutMembersInput = {
  update: Prisma.XOR<Prisma.ChamaUpdateWithoutMembersInput, Prisma.ChamaUncheckedUpdateWithoutMembersInput>
  create: Prisma.XOR<Prisma.ChamaCreateWithoutMembersInput, Prisma.ChamaUncheckedCreateWithoutMembersInput>
  where?: Prisma.ChamaWhereInput
}

export type ChamaUpdateToOneWithWhereWithoutMembersInput = {
  where?: Prisma.ChamaWhereInput
  data: Prisma.XOR<Prisma.ChamaUpdateWithoutMembersInput, Prisma.ChamaUncheckedUpdateWithoutMembersInput>
}

export type ChamaUpdateWithoutMembersInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  admin?: Prisma.UserUpdateOneRequiredWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateWithoutMembersInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  adminId?: Prisma.StringFieldUpdateOperationsInput | string
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  loans?: Prisma.ChamaLoanUncheckedUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUncheckedUpdateManyWithoutChamaNestedInput
}

export type ChamaCreateManyAdminInput = {
  id?: string
  name: string
  location: string
  profilePic?: string | null
  registrationDocument?: string | null
  minimumSavings?: number | null
  nextMeeting?: Date | string | null
  registrationId?: string | null
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ChamaUpdateWithoutAdminInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  members?: Prisma.MemberUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateWithoutAdminInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  members?: Prisma.MemberUncheckedUpdateManyWithoutChamaNestedInput
  loans?: Prisma.ChamaLoanUncheckedUpdateManyWithoutChamaNestedInput
  projects?: Prisma.ProjectsUncheckedUpdateManyWithoutChamaNestedInput
}

export type ChamaUncheckedUpdateManyWithoutAdminInput = {
  name?: Prisma.StringFieldUpdateOperationsInput | string
  location?: Prisma.StringFieldUpdateOperationsInput | string
  profilePic?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  registrationDocument?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  minimumSavings?: Prisma.NullableFloatFieldUpdateOperationsInput | number | null
  nextMeeting?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  registrationId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}


/**
 * Count Type ChamaCountOutputType
 */

export type ChamaCountOutputType = {
  members: number
  loans: number
  projects: number
}

export type ChamaCountOutputTypeSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  members?: boolean | ChamaCountOutputTypeCountMembersArgs
  loans?: boolean | ChamaCountOutputTypeCountLoansArgs
  projects?: boolean | ChamaCountOutputTypeCountProjectsArgs
}

/**
 * ChamaCountOutputType without action
 */
export type ChamaCountOutputTypeDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ChamaCountOutputType
   */
  select?: Prisma.ChamaCountOutputTypeSelect<ExtArgs> | null
}

/**
 * ChamaCountOutputType without action
 */
export type ChamaCountOutputTypeCountMembersArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.MemberWhereInput
}

/**
 * ChamaCountOutputType without action
 */
export type ChamaCountOutputTypeCountLoansArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ChamaLoanWhereInput
}

/**
 * ChamaCountOutputType without action
 */
export type ChamaCountOutputTypeCountProjectsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ProjectsWhereInput
}


export type ChamaSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  location?: boolean
  profilePic?: boolean
  registrationDocument?: boolean
  minimumSavings?: boolean
  nextMeeting?: boolean
  registrationId?: boolean
  adminId?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  admin?: boolean | Prisma.UserDefaultArgs<ExtArgs>
  members?: boolean | Prisma.Chama$membersArgs<ExtArgs>
  loans?: boolean | Prisma.Chama$loansArgs<ExtArgs>
  projects?: boolean | Prisma.Chama$projectsArgs<ExtArgs>
  _count?: boolean | Prisma.ChamaCountOutputTypeDefaultArgs<ExtArgs>
}, ExtArgs["result"]["chama"]>



export type ChamaSelectScalar = {
  id?: boolean
  name?: boolean
  location?: boolean
  profilePic?: boolean
  registrationDocument?: boolean
  minimumSavings?: boolean
  nextMeeting?: boolean
  registrationId?: boolean
  adminId?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type ChamaOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "name" | "location" | "profilePic" | "registrationDocument" | "minimumSavings" | "nextMeeting" | "registrationId" | "adminId" | "createdAt" | "updatedAt", ExtArgs["result"]["chama"]>
export type ChamaInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  admin?: boolean | Prisma.UserDefaultArgs<ExtArgs>
  members?: boolean | Prisma.Chama$membersArgs<ExtArgs>
  loans?: boolean | Prisma.Chama$loansArgs<ExtArgs>
  projects?: boolean | Prisma.Chama$projectsArgs<ExtArgs>
  _count?: boolean | Prisma.ChamaCountOutputTypeDefaultArgs<ExtArgs>
}

export type $ChamaPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "Chama"
  objects: {
    admin: Prisma.$UserPayload<ExtArgs>
    members: Prisma.$MemberPayload<ExtArgs>[]
    loans: Prisma.$ChamaLoanPayload<ExtArgs>[]
    projects: Prisma.$ProjectsPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    name: string
    location: string
    profilePic: string | null
    registrationDocument: string | null
    minimumSavings: number | null
    nextMeeting: Date | null
    registrationId: string | null
    adminId: string
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["chama"]>
  composites: {}
}

export type ChamaGetPayload<S extends boolean | null | undefined | ChamaDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$ChamaPayload, S>

export type ChamaCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<ChamaFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: ChamaCountAggregateInputType | true
  }

export interface ChamaDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['Chama'], meta: { name: 'Chama' } }
  /**
   * Find zero or one Chama that matches the filter.
   * @param {ChamaFindUniqueArgs} args - Arguments to find a Chama
   * @example
   * // Get one Chama
   * const chama = await prisma.chama.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends ChamaFindUniqueArgs>(args: Prisma.SelectSubset<T, ChamaFindUniqueArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Chama that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {ChamaFindUniqueOrThrowArgs} args - Arguments to find a Chama
   * @example
   * // Get one Chama
   * const chama = await prisma.chama.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends ChamaFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, ChamaFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Chama that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaFindFirstArgs} args - Arguments to find a Chama
   * @example
   * // Get one Chama
   * const chama = await prisma.chama.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends ChamaFindFirstArgs>(args?: Prisma.SelectSubset<T, ChamaFindFirstArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Chama that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaFindFirstOrThrowArgs} args - Arguments to find a Chama
   * @example
   * // Get one Chama
   * const chama = await prisma.chama.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends ChamaFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, ChamaFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Chamas that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Chamas
   * const chamas = await prisma.chama.findMany()
   * 
   * // Get first 10 Chamas
   * const chamas = await prisma.chama.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const chamaWithIdOnly = await prisma.chama.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends ChamaFindManyArgs>(args?: Prisma.SelectSubset<T, ChamaFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Chama.
   * @param {ChamaCreateArgs} args - Arguments to create a Chama.
   * @example
   * // Create one Chama
   * const Chama = await prisma.chama.create({
   *   data: {
   *     // ... data to create a Chama
   *   }
   * })
   * 
   */
  create<T extends ChamaCreateArgs>(args: Prisma.SelectSubset<T, ChamaCreateArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Chamas.
   * @param {ChamaCreateManyArgs} args - Arguments to create many Chamas.
   * @example
   * // Create many Chamas
   * const chama = await prisma.chama.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends ChamaCreateManyArgs>(args?: Prisma.SelectSubset<T, ChamaCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a Chama.
   * @param {ChamaDeleteArgs} args - Arguments to delete one Chama.
   * @example
   * // Delete one Chama
   * const Chama = await prisma.chama.delete({
   *   where: {
   *     // ... filter to delete one Chama
   *   }
   * })
   * 
   */
  delete<T extends ChamaDeleteArgs>(args: Prisma.SelectSubset<T, ChamaDeleteArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Chama.
   * @param {ChamaUpdateArgs} args - Arguments to update one Chama.
   * @example
   * // Update one Chama
   * const chama = await prisma.chama.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends ChamaUpdateArgs>(args: Prisma.SelectSubset<T, ChamaUpdateArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Chamas.
   * @param {ChamaDeleteManyArgs} args - Arguments to filter Chamas to delete.
   * @example
   * // Delete a few Chamas
   * const { count } = await prisma.chama.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends ChamaDeleteManyArgs>(args?: Prisma.SelectSubset<T, ChamaDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Chamas.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Chamas
   * const chama = await prisma.chama.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends ChamaUpdateManyArgs>(args: Prisma.SelectSubset<T, ChamaUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one Chama.
   * @param {ChamaUpsertArgs} args - Arguments to update or create a Chama.
   * @example
   * // Update or create a Chama
   * const chama = await prisma.chama.upsert({
   *   create: {
   *     // ... data to create a Chama
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Chama we want to update
   *   }
   * })
   */
  upsert<T extends ChamaUpsertArgs>(args: Prisma.SelectSubset<T, ChamaUpsertArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Chamas that matches the filter.
   * @param {ChamaFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const chama = await prisma.chama.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.ChamaFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a Chama.
   * @param {ChamaAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const chama = await prisma.chama.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.ChamaAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of Chamas.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaCountArgs} args - Arguments to filter Chamas to count.
   * @example
   * // Count the number of Chamas
   * const count = await prisma.chama.count({
   *   where: {
   *     // ... the filter for the Chamas we want to count
   *   }
   * })
  **/
  count<T extends ChamaCountArgs>(
    args?: Prisma.Subset<T, ChamaCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], ChamaCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Chama.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends ChamaAggregateArgs>(args: Prisma.Subset<T, ChamaAggregateArgs>): Prisma.PrismaPromise<GetChamaAggregateType<T>>

  /**
   * Group by Chama.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ChamaGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends ChamaGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: ChamaGroupByArgs['orderBy'] }
      : { orderBy?: ChamaGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, ChamaGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetChamaGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the Chama model
 */
readonly fields: ChamaFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for Chama.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__ChamaClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  admin<T extends Prisma.UserDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.UserDefaultArgs<ExtArgs>>): Prisma.Prisma__UserClient<runtime.Types.Result.GetResult<Prisma.$UserPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  members<T extends Prisma.Chama$membersArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Chama$membersArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$MemberPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  loans<T extends Prisma.Chama$loansArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Chama$loansArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ChamaLoanPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  projects<T extends Prisma.Chama$projectsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Chama$projectsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the Chama model
 */
export interface ChamaFieldRefs {
  readonly id: Prisma.FieldRef<"Chama", 'String'>
  readonly name: Prisma.FieldRef<"Chama", 'String'>
  readonly location: Prisma.FieldRef<"Chama", 'String'>
  readonly profilePic: Prisma.FieldRef<"Chama", 'String'>
  readonly registrationDocument: Prisma.FieldRef<"Chama", 'String'>
  readonly minimumSavings: Prisma.FieldRef<"Chama", 'Float'>
  readonly nextMeeting: Prisma.FieldRef<"Chama", 'DateTime'>
  readonly registrationId: Prisma.FieldRef<"Chama", 'String'>
  readonly adminId: Prisma.FieldRef<"Chama", 'String'>
  readonly createdAt: Prisma.FieldRef<"Chama", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"Chama", 'DateTime'>
}
    

// Custom InputTypes
/**
 * Chama findUnique
 */
export type ChamaFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter, which Chama to fetch.
   */
  where: Prisma.ChamaWhereUniqueInput
}

/**
 * Chama findUniqueOrThrow
 */
export type ChamaFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter, which Chama to fetch.
   */
  where: Prisma.ChamaWhereUniqueInput
}

/**
 * Chama findFirst
 */
export type ChamaFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter, which Chama to fetch.
   */
  where?: Prisma.ChamaWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Chamas to fetch.
   */
  orderBy?: Prisma.ChamaOrderByWithRelationInput | Prisma.ChamaOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Chamas.
   */
  cursor?: Prisma.ChamaWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Chamas from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Chamas.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Chamas.
   */
  distinct?: Prisma.ChamaScalarFieldEnum | Prisma.ChamaScalarFieldEnum[]
}

/**
 * Chama findFirstOrThrow
 */
export type ChamaFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter, which Chama to fetch.
   */
  where?: Prisma.ChamaWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Chamas to fetch.
   */
  orderBy?: Prisma.ChamaOrderByWithRelationInput | Prisma.ChamaOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Chamas.
   */
  cursor?: Prisma.ChamaWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Chamas from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Chamas.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Chamas.
   */
  distinct?: Prisma.ChamaScalarFieldEnum | Prisma.ChamaScalarFieldEnum[]
}

/**
 * Chama findMany
 */
export type ChamaFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter, which Chamas to fetch.
   */
  where?: Prisma.ChamaWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Chamas to fetch.
   */
  orderBy?: Prisma.ChamaOrderByWithRelationInput | Prisma.ChamaOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing Chamas.
   */
  cursor?: Prisma.ChamaWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Chamas from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Chamas.
   */
  skip?: number
  distinct?: Prisma.ChamaScalarFieldEnum | Prisma.ChamaScalarFieldEnum[]
}

/**
 * Chama create
 */
export type ChamaCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * The data needed to create a Chama.
   */
  data: Prisma.XOR<Prisma.ChamaCreateInput, Prisma.ChamaUncheckedCreateInput>
}

/**
 * Chama createMany
 */
export type ChamaCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many Chamas.
   */
  data: Prisma.ChamaCreateManyInput | Prisma.ChamaCreateManyInput[]
}

/**
 * Chama update
 */
export type ChamaUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * The data needed to update a Chama.
   */
  data: Prisma.XOR<Prisma.ChamaUpdateInput, Prisma.ChamaUncheckedUpdateInput>
  /**
   * Choose, which Chama to update.
   */
  where: Prisma.ChamaWhereUniqueInput
}

/**
 * Chama updateMany
 */
export type ChamaUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update Chamas.
   */
  data: Prisma.XOR<Prisma.ChamaUpdateManyMutationInput, Prisma.ChamaUncheckedUpdateManyInput>
  /**
   * Filter which Chamas to update
   */
  where?: Prisma.ChamaWhereInput
  /**
   * Limit how many Chamas to update.
   */
  limit?: number
}

/**
 * Chama upsert
 */
export type ChamaUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * The filter to search for the Chama to update in case it exists.
   */
  where: Prisma.ChamaWhereUniqueInput
  /**
   * In case the Chama found by the `where` argument doesn't exist, create a new Chama with this data.
   */
  create: Prisma.XOR<Prisma.ChamaCreateInput, Prisma.ChamaUncheckedCreateInput>
  /**
   * In case the Chama was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.ChamaUpdateInput, Prisma.ChamaUncheckedUpdateInput>
}

/**
 * Chama delete
 */
export type ChamaDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
  /**
   * Filter which Chama to delete.
   */
  where: Prisma.ChamaWhereUniqueInput
}

/**
 * Chama deleteMany
 */
export type ChamaDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Chamas to delete
   */
  where?: Prisma.ChamaWhereInput
  /**
   * Limit how many Chamas to delete.
   */
  limit?: number
}

/**
 * Chama findRaw
 */
export type ChamaFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Chama aggregateRaw
 */
export type ChamaAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Chama.members
 */
export type Chama$membersArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Member
   */
  select?: Prisma.MemberSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Member
   */
  omit?: Prisma.MemberOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.MemberInclude<ExtArgs> | null
  where?: Prisma.MemberWhereInput
  orderBy?: Prisma.MemberOrderByWithRelationInput | Prisma.MemberOrderByWithRelationInput[]
  cursor?: Prisma.MemberWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.MemberScalarFieldEnum | Prisma.MemberScalarFieldEnum[]
}

/**
 * Chama.loans
 */
export type Chama$loansArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ChamaLoan
   */
  select?: Prisma.ChamaLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ChamaLoan
   */
  omit?: Prisma.ChamaLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaLoanInclude<ExtArgs> | null
  where?: Prisma.ChamaLoanWhereInput
  orderBy?: Prisma.ChamaLoanOrderByWithRelationInput | Prisma.ChamaLoanOrderByWithRelationInput[]
  cursor?: Prisma.ChamaLoanWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.ChamaLoanScalarFieldEnum | Prisma.ChamaLoanScalarFieldEnum[]
}

/**
 * Chama.projects
 */
export type Chama$projectsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  where?: Prisma.ProjectsWhereInput
  orderBy?: Prisma.ProjectsOrderByWithRelationInput | Prisma.ProjectsOrderByWithRelationInput[]
  cursor?: Prisma.ProjectsWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.ProjectsScalarFieldEnum | Prisma.ProjectsScalarFieldEnum[]
}

/**
 * Chama without action
 */
export type ChamaDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Chama
   */
  select?: Prisma.ChamaSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Chama
   */
  omit?: Prisma.ChamaOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ChamaInclude<ExtArgs> | null
}
