// @prisma/client/models/Deduction.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `Deduction` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model Deduction
 * 
 */
export type DeductionModel = runtime.Types.Result.DefaultSelection<Prisma.$DeductionPayload>

export type AggregateDeduction = {
  _count: DeductionCountAggregateOutputType | null
  _min: DeductionMinAggregateOutputType | null
  _max: DeductionMaxAggregateOutputType | null
}

export type DeductionMinAggregateOutputType = {
  id: string | null
  memberId: string | null
  longTermLoanId: string | null
  shortTermLoanId: string | null
  penaltyId: string | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type DeductionMaxAggregateOutputType = {
  id: string | null
  memberId: string | null
  longTermLoanId: string | null
  shortTermLoanId: string | null
  penaltyId: string | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type DeductionCountAggregateOutputType = {
  id: number
  memberId: number
  longTermLoanId: number
  shortTermLoanId: number
  penaltyId: number
  status: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type DeductionMinAggregateInputType = {
  id?: true
  memberId?: true
  longTermLoanId?: true
  shortTermLoanId?: true
  penaltyId?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type DeductionMaxAggregateInputType = {
  id?: true
  memberId?: true
  longTermLoanId?: true
  shortTermLoanId?: true
  penaltyId?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type DeductionCountAggregateInputType = {
  id?: true
  memberId?: true
  longTermLoanId?: true
  shortTermLoanId?: true
  penaltyId?: true
  status?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type DeductionAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Deduction to aggregate.
   */
  where?: Prisma.DeductionWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Deductions to fetch.
   */
  orderBy?: Prisma.DeductionOrderByWithRelationInput | Prisma.DeductionOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.DeductionWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Deductions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Deductions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned Deductions
  **/
  _count?: true | DeductionCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: DeductionMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: DeductionMaxAggregateInputType
}

export type GetDeductionAggregateType<T extends DeductionAggregateArgs> = {
      [P in keyof T & keyof AggregateDeduction]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateDeduction[P]>
    : Prisma.GetScalarType<T[P], AggregateDeduction[P]>
}




export type DeductionGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.DeductionWhereInput
  orderBy?: Prisma.DeductionOrderByWithAggregationInput | Prisma.DeductionOrderByWithAggregationInput[]
  by: Prisma.DeductionScalarFieldEnum[] | Prisma.DeductionScalarFieldEnum
  having?: Prisma.DeductionScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: DeductionCountAggregateInputType | true
  _min?: DeductionMinAggregateInputType
  _max?: DeductionMaxAggregateInputType
}

export type DeductionGroupByOutputType = {
  id: string
  memberId: string
  longTermLoanId: string | null
  shortTermLoanId: string | null
  penaltyId: string | null
  status: $Enums.TransactionStatus
  createdAt: Date
  updatedAt: Date
  _count: DeductionCountAggregateOutputType | null
  _min: DeductionMinAggregateOutputType | null
  _max: DeductionMaxAggregateOutputType | null
}

type GetDeductionGroupByPayload<T extends DeductionGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<DeductionGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof DeductionGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], DeductionGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], DeductionGroupByOutputType[P]>
      }
    >
  >



export type DeductionWhereInput = {
  AND?: Prisma.DeductionWhereInput | Prisma.DeductionWhereInput[]
  OR?: Prisma.DeductionWhereInput[]
  NOT?: Prisma.DeductionWhereInput | Prisma.DeductionWhereInput[]
  id?: Prisma.StringFilter<"Deduction"> | string
  memberId?: Prisma.StringFilter<"Deduction"> | string
  longTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  shortTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  penaltyId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Deduction"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
  longTermLoan?: Prisma.XOR<Prisma.LongTermLoanNullableScalarRelationFilter, Prisma.LongTermLoanWhereInput> | null
  shortTermLoan?: Prisma.XOR<Prisma.ShortTermLoanNullableScalarRelationFilter, Prisma.ShortTermLoanWhereInput> | null
  penalty?: Prisma.XOR<Prisma.PenaltyNullableScalarRelationFilter, Prisma.PenaltyWhereInput> | null
  savings?: Prisma.SavingListRelationFilter
}

export type DeductionOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  shortTermLoanId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  member?: Prisma.MemberOrderByWithRelationInput
  longTermLoan?: Prisma.LongTermLoanOrderByWithRelationInput
  shortTermLoan?: Prisma.ShortTermLoanOrderByWithRelationInput
  penalty?: Prisma.PenaltyOrderByWithRelationInput
  savings?: Prisma.SavingOrderByRelationAggregateInput
}

export type DeductionWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.DeductionWhereInput | Prisma.DeductionWhereInput[]
  OR?: Prisma.DeductionWhereInput[]
  NOT?: Prisma.DeductionWhereInput | Prisma.DeductionWhereInput[]
  memberId?: Prisma.StringFilter<"Deduction"> | string
  longTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  shortTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  penaltyId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Deduction"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
  longTermLoan?: Prisma.XOR<Prisma.LongTermLoanNullableScalarRelationFilter, Prisma.LongTermLoanWhereInput> | null
  shortTermLoan?: Prisma.XOR<Prisma.ShortTermLoanNullableScalarRelationFilter, Prisma.ShortTermLoanWhereInput> | null
  penalty?: Prisma.XOR<Prisma.PenaltyNullableScalarRelationFilter, Prisma.PenaltyWhereInput> | null
  savings?: Prisma.SavingListRelationFilter
}, "id">

export type DeductionOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  shortTermLoanId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.DeductionCountOrderByAggregateInput
  _max?: Prisma.DeductionMaxOrderByAggregateInput
  _min?: Prisma.DeductionMinOrderByAggregateInput
}

export type DeductionScalarWhereWithAggregatesInput = {
  AND?: Prisma.DeductionScalarWhereWithAggregatesInput | Prisma.DeductionScalarWhereWithAggregatesInput[]
  OR?: Prisma.DeductionScalarWhereWithAggregatesInput[]
  NOT?: Prisma.DeductionScalarWhereWithAggregatesInput | Prisma.DeductionScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"Deduction"> | string
  memberId?: Prisma.StringWithAggregatesFilter<"Deduction"> | string
  longTermLoanId?: Prisma.StringNullableWithAggregatesFilter<"Deduction"> | string | null
  shortTermLoanId?: Prisma.StringNullableWithAggregatesFilter<"Deduction"> | string | null
  penaltyId?: Prisma.StringNullableWithAggregatesFilter<"Deduction"> | string | null
  status?: Prisma.EnumTransactionStatusWithAggregatesFilter<"Deduction"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"Deduction"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"Deduction"> | Date | string
}

export type DeductionCreateInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDeductionsInput
  longTermLoan?: Prisma.LongTermLoanCreateNestedOneWithoutDeductionsInput
  shortTermLoan?: Prisma.ShortTermLoanCreateNestedOneWithoutDeductionsInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutDeductionsInput
  savings?: Prisma.SavingCreateNestedManyWithoutDeductionInput
}

export type DeductionUncheckedCreateInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  savings?: Prisma.SavingUncheckedCreateNestedManyWithoutDeductionInput
}

export type DeductionUpdateInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDeductionsNestedInput
  longTermLoan?: Prisma.LongTermLoanUpdateOneWithoutDeductionsNestedInput
  shortTermLoan?: Prisma.ShortTermLoanUpdateOneWithoutDeductionsNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutDeductionsNestedInput
  savings?: Prisma.SavingUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  savings?: Prisma.SavingUncheckedUpdateManyWithoutDeductionNestedInput
}

export type DeductionCreateManyInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionUpdateManyMutationInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionUncheckedUpdateManyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionListRelationFilter = {
  every?: Prisma.DeductionWhereInput
  some?: Prisma.DeductionWhereInput
  none?: Prisma.DeductionWhereInput
}

export type DeductionOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type DeductionNullableScalarRelationFilter = {
  is?: Prisma.DeductionWhereInput | null
  isNot?: Prisma.DeductionWhereInput | null
}

export type DeductionCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  shortTermLoanId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DeductionMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  shortTermLoanId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DeductionMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  shortTermLoanId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DeductionCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput> | Prisma.DeductionCreateWithoutMemberInput[] | Prisma.DeductionUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutMemberInput | Prisma.DeductionCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.DeductionCreateManyMemberInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUncheckedCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput> | Prisma.DeductionCreateWithoutMemberInput[] | Prisma.DeductionUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutMemberInput | Prisma.DeductionCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.DeductionCreateManyMemberInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput> | Prisma.DeductionCreateWithoutMemberInput[] | Prisma.DeductionUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutMemberInput | Prisma.DeductionCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutMemberInput | Prisma.DeductionUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.DeductionCreateManyMemberInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutMemberInput | Prisma.DeductionUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutMemberInput | Prisma.DeductionUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionUncheckedUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput> | Prisma.DeductionCreateWithoutMemberInput[] | Prisma.DeductionUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutMemberInput | Prisma.DeductionCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutMemberInput | Prisma.DeductionUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.DeductionCreateManyMemberInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutMemberInput | Prisma.DeductionUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutMemberInput | Prisma.DeductionUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionCreateNestedOneWithoutSavingsInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutSavingsInput, Prisma.DeductionUncheckedCreateWithoutSavingsInput>
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutSavingsInput
  connect?: Prisma.DeductionWhereUniqueInput
}

export type DeductionUpdateOneWithoutSavingsNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutSavingsInput, Prisma.DeductionUncheckedCreateWithoutSavingsInput>
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutSavingsInput
  upsert?: Prisma.DeductionUpsertWithoutSavingsInput
  disconnect?: boolean
  delete?: Prisma.DeductionWhereInput | boolean
  connect?: Prisma.DeductionWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.DeductionUpdateToOneWithWhereWithoutSavingsInput, Prisma.DeductionUpdateWithoutSavingsInput>, Prisma.DeductionUncheckedUpdateWithoutSavingsInput>
}

export type DeductionCreateNestedManyWithoutLongTermLoanInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput> | Prisma.DeductionCreateWithoutLongTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput | Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyLongTermLoanInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUncheckedCreateNestedManyWithoutLongTermLoanInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput> | Prisma.DeductionCreateWithoutLongTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput | Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyLongTermLoanInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUpdateManyWithoutLongTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput> | Prisma.DeductionCreateWithoutLongTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput | Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutLongTermLoanInput | Prisma.DeductionUpsertWithWhereUniqueWithoutLongTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyLongTermLoanInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutLongTermLoanInput | Prisma.DeductionUpdateWithWhereUniqueWithoutLongTermLoanInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutLongTermLoanInput | Prisma.DeductionUpdateManyWithWhereWithoutLongTermLoanInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionUncheckedUpdateManyWithoutLongTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput> | Prisma.DeductionCreateWithoutLongTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput | Prisma.DeductionCreateOrConnectWithoutLongTermLoanInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutLongTermLoanInput | Prisma.DeductionUpsertWithWhereUniqueWithoutLongTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyLongTermLoanInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutLongTermLoanInput | Prisma.DeductionUpdateWithWhereUniqueWithoutLongTermLoanInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutLongTermLoanInput | Prisma.DeductionUpdateManyWithWhereWithoutLongTermLoanInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionCreateNestedManyWithoutShortTermLoanInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput> | Prisma.DeductionCreateWithoutShortTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput | Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyShortTermLoanInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUncheckedCreateNestedManyWithoutShortTermLoanInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput> | Prisma.DeductionCreateWithoutShortTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput | Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyShortTermLoanInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUpdateManyWithoutShortTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput> | Prisma.DeductionCreateWithoutShortTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput | Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutShortTermLoanInput | Prisma.DeductionUpsertWithWhereUniqueWithoutShortTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyShortTermLoanInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutShortTermLoanInput | Prisma.DeductionUpdateWithWhereUniqueWithoutShortTermLoanInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutShortTermLoanInput | Prisma.DeductionUpdateManyWithWhereWithoutShortTermLoanInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionUncheckedUpdateManyWithoutShortTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput> | Prisma.DeductionCreateWithoutShortTermLoanInput[] | Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput | Prisma.DeductionCreateOrConnectWithoutShortTermLoanInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutShortTermLoanInput | Prisma.DeductionUpsertWithWhereUniqueWithoutShortTermLoanInput[]
  createMany?: Prisma.DeductionCreateManyShortTermLoanInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutShortTermLoanInput | Prisma.DeductionUpdateWithWhereUniqueWithoutShortTermLoanInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutShortTermLoanInput | Prisma.DeductionUpdateManyWithWhereWithoutShortTermLoanInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionCreateNestedManyWithoutPenaltyInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput> | Prisma.DeductionCreateWithoutPenaltyInput[] | Prisma.DeductionUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutPenaltyInput | Prisma.DeductionCreateOrConnectWithoutPenaltyInput[]
  createMany?: Prisma.DeductionCreateManyPenaltyInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUncheckedCreateNestedManyWithoutPenaltyInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput> | Prisma.DeductionCreateWithoutPenaltyInput[] | Prisma.DeductionUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutPenaltyInput | Prisma.DeductionCreateOrConnectWithoutPenaltyInput[]
  createMany?: Prisma.DeductionCreateManyPenaltyInputEnvelope
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
}

export type DeductionUpdateManyWithoutPenaltyNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput> | Prisma.DeductionCreateWithoutPenaltyInput[] | Prisma.DeductionUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutPenaltyInput | Prisma.DeductionCreateOrConnectWithoutPenaltyInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutPenaltyInput | Prisma.DeductionUpsertWithWhereUniqueWithoutPenaltyInput[]
  createMany?: Prisma.DeductionCreateManyPenaltyInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutPenaltyInput | Prisma.DeductionUpdateWithWhereUniqueWithoutPenaltyInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutPenaltyInput | Prisma.DeductionUpdateManyWithWhereWithoutPenaltyInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionUncheckedUpdateManyWithoutPenaltyNestedInput = {
  create?: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput> | Prisma.DeductionCreateWithoutPenaltyInput[] | Prisma.DeductionUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.DeductionCreateOrConnectWithoutPenaltyInput | Prisma.DeductionCreateOrConnectWithoutPenaltyInput[]
  upsert?: Prisma.DeductionUpsertWithWhereUniqueWithoutPenaltyInput | Prisma.DeductionUpsertWithWhereUniqueWithoutPenaltyInput[]
  createMany?: Prisma.DeductionCreateManyPenaltyInputEnvelope
  set?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  disconnect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  delete?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  connect?: Prisma.DeductionWhereUniqueInput | Prisma.DeductionWhereUniqueInput[]
  update?: Prisma.DeductionUpdateWithWhereUniqueWithoutPenaltyInput | Prisma.DeductionUpdateWithWhereUniqueWithoutPenaltyInput[]
  updateMany?: Prisma.DeductionUpdateManyWithWhereWithoutPenaltyInput | Prisma.DeductionUpdateManyWithWhereWithoutPenaltyInput[]
  deleteMany?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
}

export type DeductionCreateWithoutMemberInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  longTermLoan?: Prisma.LongTermLoanCreateNestedOneWithoutDeductionsInput
  shortTermLoan?: Prisma.ShortTermLoanCreateNestedOneWithoutDeductionsInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutDeductionsInput
  savings?: Prisma.SavingCreateNestedManyWithoutDeductionInput
}

export type DeductionUncheckedCreateWithoutMemberInput = {
  id?: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  savings?: Prisma.SavingUncheckedCreateNestedManyWithoutDeductionInput
}

export type DeductionCreateOrConnectWithoutMemberInput = {
  where: Prisma.DeductionWhereUniqueInput
  create: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput>
}

export type DeductionCreateManyMemberInputEnvelope = {
  data: Prisma.DeductionCreateManyMemberInput | Prisma.DeductionCreateManyMemberInput[]
}

export type DeductionUpsertWithWhereUniqueWithoutMemberInput = {
  where: Prisma.DeductionWhereUniqueInput
  update: Prisma.XOR<Prisma.DeductionUpdateWithoutMemberInput, Prisma.DeductionUncheckedUpdateWithoutMemberInput>
  create: Prisma.XOR<Prisma.DeductionCreateWithoutMemberInput, Prisma.DeductionUncheckedCreateWithoutMemberInput>
}

export type DeductionUpdateWithWhereUniqueWithoutMemberInput = {
  where: Prisma.DeductionWhereUniqueInput
  data: Prisma.XOR<Prisma.DeductionUpdateWithoutMemberInput, Prisma.DeductionUncheckedUpdateWithoutMemberInput>
}

export type DeductionUpdateManyWithWhereWithoutMemberInput = {
  where: Prisma.DeductionScalarWhereInput
  data: Prisma.XOR<Prisma.DeductionUpdateManyMutationInput, Prisma.DeductionUncheckedUpdateManyWithoutMemberInput>
}

export type DeductionScalarWhereInput = {
  AND?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
  OR?: Prisma.DeductionScalarWhereInput[]
  NOT?: Prisma.DeductionScalarWhereInput | Prisma.DeductionScalarWhereInput[]
  id?: Prisma.StringFilter<"Deduction"> | string
  memberId?: Prisma.StringFilter<"Deduction"> | string
  longTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  shortTermLoanId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  penaltyId?: Prisma.StringNullableFilter<"Deduction"> | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Deduction"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Deduction"> | Date | string
}

export type DeductionCreateWithoutSavingsInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDeductionsInput
  longTermLoan?: Prisma.LongTermLoanCreateNestedOneWithoutDeductionsInput
  shortTermLoan?: Prisma.ShortTermLoanCreateNestedOneWithoutDeductionsInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutDeductionsInput
}

export type DeductionUncheckedCreateWithoutSavingsInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionCreateOrConnectWithoutSavingsInput = {
  where: Prisma.DeductionWhereUniqueInput
  create: Prisma.XOR<Prisma.DeductionCreateWithoutSavingsInput, Prisma.DeductionUncheckedCreateWithoutSavingsInput>
}

export type DeductionUpsertWithoutSavingsInput = {
  update: Prisma.XOR<Prisma.DeductionUpdateWithoutSavingsInput, Prisma.DeductionUncheckedUpdateWithoutSavingsInput>
  create: Prisma.XOR<Prisma.DeductionCreateWithoutSavingsInput, Prisma.DeductionUncheckedCreateWithoutSavingsInput>
  where?: Prisma.DeductionWhereInput
}

export type DeductionUpdateToOneWithWhereWithoutSavingsInput = {
  where?: Prisma.DeductionWhereInput
  data: Prisma.XOR<Prisma.DeductionUpdateWithoutSavingsInput, Prisma.DeductionUncheckedUpdateWithoutSavingsInput>
}

export type DeductionUpdateWithoutSavingsInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDeductionsNestedInput
  longTermLoan?: Prisma.LongTermLoanUpdateOneWithoutDeductionsNestedInput
  shortTermLoan?: Prisma.ShortTermLoanUpdateOneWithoutDeductionsNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutDeductionsNestedInput
}

export type DeductionUncheckedUpdateWithoutSavingsInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionCreateWithoutLongTermLoanInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDeductionsInput
  shortTermLoan?: Prisma.ShortTermLoanCreateNestedOneWithoutDeductionsInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutDeductionsInput
  savings?: Prisma.SavingCreateNestedManyWithoutDeductionInput
}

export type DeductionUncheckedCreateWithoutLongTermLoanInput = {
  id?: string
  memberId: string
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  savings?: Prisma.SavingUncheckedCreateNestedManyWithoutDeductionInput
}

export type DeductionCreateOrConnectWithoutLongTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  create: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput>
}

export type DeductionCreateManyLongTermLoanInputEnvelope = {
  data: Prisma.DeductionCreateManyLongTermLoanInput | Prisma.DeductionCreateManyLongTermLoanInput[]
}

export type DeductionUpsertWithWhereUniqueWithoutLongTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  update: Prisma.XOR<Prisma.DeductionUpdateWithoutLongTermLoanInput, Prisma.DeductionUncheckedUpdateWithoutLongTermLoanInput>
  create: Prisma.XOR<Prisma.DeductionCreateWithoutLongTermLoanInput, Prisma.DeductionUncheckedCreateWithoutLongTermLoanInput>
}

export type DeductionUpdateWithWhereUniqueWithoutLongTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  data: Prisma.XOR<Prisma.DeductionUpdateWithoutLongTermLoanInput, Prisma.DeductionUncheckedUpdateWithoutLongTermLoanInput>
}

export type DeductionUpdateManyWithWhereWithoutLongTermLoanInput = {
  where: Prisma.DeductionScalarWhereInput
  data: Prisma.XOR<Prisma.DeductionUpdateManyMutationInput, Prisma.DeductionUncheckedUpdateManyWithoutLongTermLoanInput>
}

export type DeductionCreateWithoutShortTermLoanInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDeductionsInput
  longTermLoan?: Prisma.LongTermLoanCreateNestedOneWithoutDeductionsInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutDeductionsInput
  savings?: Prisma.SavingCreateNestedManyWithoutDeductionInput
}

export type DeductionUncheckedCreateWithoutShortTermLoanInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  savings?: Prisma.SavingUncheckedCreateNestedManyWithoutDeductionInput
}

export type DeductionCreateOrConnectWithoutShortTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  create: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput>
}

export type DeductionCreateManyShortTermLoanInputEnvelope = {
  data: Prisma.DeductionCreateManyShortTermLoanInput | Prisma.DeductionCreateManyShortTermLoanInput[]
}

export type DeductionUpsertWithWhereUniqueWithoutShortTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  update: Prisma.XOR<Prisma.DeductionUpdateWithoutShortTermLoanInput, Prisma.DeductionUncheckedUpdateWithoutShortTermLoanInput>
  create: Prisma.XOR<Prisma.DeductionCreateWithoutShortTermLoanInput, Prisma.DeductionUncheckedCreateWithoutShortTermLoanInput>
}

export type DeductionUpdateWithWhereUniqueWithoutShortTermLoanInput = {
  where: Prisma.DeductionWhereUniqueInput
  data: Prisma.XOR<Prisma.DeductionUpdateWithoutShortTermLoanInput, Prisma.DeductionUncheckedUpdateWithoutShortTermLoanInput>
}

export type DeductionUpdateManyWithWhereWithoutShortTermLoanInput = {
  where: Prisma.DeductionScalarWhereInput
  data: Prisma.XOR<Prisma.DeductionUpdateManyMutationInput, Prisma.DeductionUncheckedUpdateManyWithoutShortTermLoanInput>
}

export type DeductionCreateWithoutPenaltyInput = {
  id?: string
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDeductionsInput
  longTermLoan?: Prisma.LongTermLoanCreateNestedOneWithoutDeductionsInput
  shortTermLoan?: Prisma.ShortTermLoanCreateNestedOneWithoutDeductionsInput
  savings?: Prisma.SavingCreateNestedManyWithoutDeductionInput
}

export type DeductionUncheckedCreateWithoutPenaltyInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  savings?: Prisma.SavingUncheckedCreateNestedManyWithoutDeductionInput
}

export type DeductionCreateOrConnectWithoutPenaltyInput = {
  where: Prisma.DeductionWhereUniqueInput
  create: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput>
}

export type DeductionCreateManyPenaltyInputEnvelope = {
  data: Prisma.DeductionCreateManyPenaltyInput | Prisma.DeductionCreateManyPenaltyInput[]
}

export type DeductionUpsertWithWhereUniqueWithoutPenaltyInput = {
  where: Prisma.DeductionWhereUniqueInput
  update: Prisma.XOR<Prisma.DeductionUpdateWithoutPenaltyInput, Prisma.DeductionUncheckedUpdateWithoutPenaltyInput>
  create: Prisma.XOR<Prisma.DeductionCreateWithoutPenaltyInput, Prisma.DeductionUncheckedCreateWithoutPenaltyInput>
}

export type DeductionUpdateWithWhereUniqueWithoutPenaltyInput = {
  where: Prisma.DeductionWhereUniqueInput
  data: Prisma.XOR<Prisma.DeductionUpdateWithoutPenaltyInput, Prisma.DeductionUncheckedUpdateWithoutPenaltyInput>
}

export type DeductionUpdateManyWithWhereWithoutPenaltyInput = {
  where: Prisma.DeductionScalarWhereInput
  data: Prisma.XOR<Prisma.DeductionUpdateManyMutationInput, Prisma.DeductionUncheckedUpdateManyWithoutPenaltyInput>
}

export type DeductionCreateManyMemberInput = {
  id?: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionUpdateWithoutMemberInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  longTermLoan?: Prisma.LongTermLoanUpdateOneWithoutDeductionsNestedInput
  shortTermLoan?: Prisma.ShortTermLoanUpdateOneWithoutDeductionsNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutDeductionsNestedInput
  savings?: Prisma.SavingUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateWithoutMemberInput = {
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  savings?: Prisma.SavingUncheckedUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateManyWithoutMemberInput = {
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionCreateManyLongTermLoanInput = {
  id?: string
  memberId: string
  shortTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionUpdateWithoutLongTermLoanInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDeductionsNestedInput
  shortTermLoan?: Prisma.ShortTermLoanUpdateOneWithoutDeductionsNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutDeductionsNestedInput
  savings?: Prisma.SavingUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateWithoutLongTermLoanInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  savings?: Prisma.SavingUncheckedUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateManyWithoutLongTermLoanInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionCreateManyShortTermLoanInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  penaltyId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionUpdateWithoutShortTermLoanInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDeductionsNestedInput
  longTermLoan?: Prisma.LongTermLoanUpdateOneWithoutDeductionsNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutDeductionsNestedInput
  savings?: Prisma.SavingUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateWithoutShortTermLoanInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  savings?: Prisma.SavingUncheckedUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateManyWithoutShortTermLoanInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DeductionCreateManyPenaltyInput = {
  id?: string
  memberId: string
  longTermLoanId?: string | null
  shortTermLoanId?: string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DeductionUpdateWithoutPenaltyInput = {
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDeductionsNestedInput
  longTermLoan?: Prisma.LongTermLoanUpdateOneWithoutDeductionsNestedInput
  shortTermLoan?: Prisma.ShortTermLoanUpdateOneWithoutDeductionsNestedInput
  savings?: Prisma.SavingUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateWithoutPenaltyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  savings?: Prisma.SavingUncheckedUpdateManyWithoutDeductionNestedInput
}

export type DeductionUncheckedUpdateManyWithoutPenaltyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  longTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  shortTermLoanId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}


/**
 * Count Type DeductionCountOutputType
 */

export type DeductionCountOutputType = {
  savings: number
}

export type DeductionCountOutputTypeSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  savings?: boolean | DeductionCountOutputTypeCountSavingsArgs
}

/**
 * DeductionCountOutputType without action
 */
export type DeductionCountOutputTypeDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the DeductionCountOutputType
   */
  select?: Prisma.DeductionCountOutputTypeSelect<ExtArgs> | null
}

/**
 * DeductionCountOutputType without action
 */
export type DeductionCountOutputTypeCountSavingsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.SavingWhereInput
}


export type DeductionSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  memberId?: boolean
  longTermLoanId?: boolean
  shortTermLoanId?: boolean
  penaltyId?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
  longTermLoan?: boolean | Prisma.Deduction$longTermLoanArgs<ExtArgs>
  shortTermLoan?: boolean | Prisma.Deduction$shortTermLoanArgs<ExtArgs>
  penalty?: boolean | Prisma.Deduction$penaltyArgs<ExtArgs>
  savings?: boolean | Prisma.Deduction$savingsArgs<ExtArgs>
  _count?: boolean | Prisma.DeductionCountOutputTypeDefaultArgs<ExtArgs>
}, ExtArgs["result"]["deduction"]>



export type DeductionSelectScalar = {
  id?: boolean
  memberId?: boolean
  longTermLoanId?: boolean
  shortTermLoanId?: boolean
  penaltyId?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type DeductionOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "memberId" | "longTermLoanId" | "shortTermLoanId" | "penaltyId" | "status" | "createdAt" | "updatedAt", ExtArgs["result"]["deduction"]>
export type DeductionInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
  longTermLoan?: boolean | Prisma.Deduction$longTermLoanArgs<ExtArgs>
  shortTermLoan?: boolean | Prisma.Deduction$shortTermLoanArgs<ExtArgs>
  penalty?: boolean | Prisma.Deduction$penaltyArgs<ExtArgs>
  savings?: boolean | Prisma.Deduction$savingsArgs<ExtArgs>
  _count?: boolean | Prisma.DeductionCountOutputTypeDefaultArgs<ExtArgs>
}

export type $DeductionPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "Deduction"
  objects: {
    member: Prisma.$MemberPayload<ExtArgs>
    longTermLoan: Prisma.$LongTermLoanPayload<ExtArgs> | null
    shortTermLoan: Prisma.$ShortTermLoanPayload<ExtArgs> | null
    penalty: Prisma.$PenaltyPayload<ExtArgs> | null
    savings: Prisma.$SavingPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    memberId: string
    longTermLoanId: string | null
    shortTermLoanId: string | null
    penaltyId: string | null
    status: $Enums.TransactionStatus
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["deduction"]>
  composites: {}
}

export type DeductionGetPayload<S extends boolean | null | undefined | DeductionDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$DeductionPayload, S>

export type DeductionCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<DeductionFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: DeductionCountAggregateInputType | true
  }

export interface DeductionDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['Deduction'], meta: { name: 'Deduction' } }
  /**
   * Find zero or one Deduction that matches the filter.
   * @param {DeductionFindUniqueArgs} args - Arguments to find a Deduction
   * @example
   * // Get one Deduction
   * const deduction = await prisma.deduction.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends DeductionFindUniqueArgs>(args: Prisma.SelectSubset<T, DeductionFindUniqueArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Deduction that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {DeductionFindUniqueOrThrowArgs} args - Arguments to find a Deduction
   * @example
   * // Get one Deduction
   * const deduction = await prisma.deduction.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends DeductionFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, DeductionFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Deduction that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionFindFirstArgs} args - Arguments to find a Deduction
   * @example
   * // Get one Deduction
   * const deduction = await prisma.deduction.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends DeductionFindFirstArgs>(args?: Prisma.SelectSubset<T, DeductionFindFirstArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Deduction that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionFindFirstOrThrowArgs} args - Arguments to find a Deduction
   * @example
   * // Get one Deduction
   * const deduction = await prisma.deduction.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends DeductionFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, DeductionFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Deductions that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Deductions
   * const deductions = await prisma.deduction.findMany()
   * 
   * // Get first 10 Deductions
   * const deductions = await prisma.deduction.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const deductionWithIdOnly = await prisma.deduction.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends DeductionFindManyArgs>(args?: Prisma.SelectSubset<T, DeductionFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Deduction.
   * @param {DeductionCreateArgs} args - Arguments to create a Deduction.
   * @example
   * // Create one Deduction
   * const Deduction = await prisma.deduction.create({
   *   data: {
   *     // ... data to create a Deduction
   *   }
   * })
   * 
   */
  create<T extends DeductionCreateArgs>(args: Prisma.SelectSubset<T, DeductionCreateArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Deductions.
   * @param {DeductionCreateManyArgs} args - Arguments to create many Deductions.
   * @example
   * // Create many Deductions
   * const deduction = await prisma.deduction.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends DeductionCreateManyArgs>(args?: Prisma.SelectSubset<T, DeductionCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a Deduction.
   * @param {DeductionDeleteArgs} args - Arguments to delete one Deduction.
   * @example
   * // Delete one Deduction
   * const Deduction = await prisma.deduction.delete({
   *   where: {
   *     // ... filter to delete one Deduction
   *   }
   * })
   * 
   */
  delete<T extends DeductionDeleteArgs>(args: Prisma.SelectSubset<T, DeductionDeleteArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Deduction.
   * @param {DeductionUpdateArgs} args - Arguments to update one Deduction.
   * @example
   * // Update one Deduction
   * const deduction = await prisma.deduction.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends DeductionUpdateArgs>(args: Prisma.SelectSubset<T, DeductionUpdateArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Deductions.
   * @param {DeductionDeleteManyArgs} args - Arguments to filter Deductions to delete.
   * @example
   * // Delete a few Deductions
   * const { count } = await prisma.deduction.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends DeductionDeleteManyArgs>(args?: Prisma.SelectSubset<T, DeductionDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Deductions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Deductions
   * const deduction = await prisma.deduction.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends DeductionUpdateManyArgs>(args: Prisma.SelectSubset<T, DeductionUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one Deduction.
   * @param {DeductionUpsertArgs} args - Arguments to update or create a Deduction.
   * @example
   * // Update or create a Deduction
   * const deduction = await prisma.deduction.upsert({
   *   create: {
   *     // ... data to create a Deduction
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Deduction we want to update
   *   }
   * })
   */
  upsert<T extends DeductionUpsertArgs>(args: Prisma.SelectSubset<T, DeductionUpsertArgs<ExtArgs>>): Prisma.Prisma__DeductionClient<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Deductions that matches the filter.
   * @param {DeductionFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const deduction = await prisma.deduction.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.DeductionFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a Deduction.
   * @param {DeductionAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const deduction = await prisma.deduction.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.DeductionAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of Deductions.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionCountArgs} args - Arguments to filter Deductions to count.
   * @example
   * // Count the number of Deductions
   * const count = await prisma.deduction.count({
   *   where: {
   *     // ... the filter for the Deductions we want to count
   *   }
   * })
  **/
  count<T extends DeductionCountArgs>(
    args?: Prisma.Subset<T, DeductionCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], DeductionCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Deduction.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends DeductionAggregateArgs>(args: Prisma.Subset<T, DeductionAggregateArgs>): Prisma.PrismaPromise<GetDeductionAggregateType<T>>

  /**
   * Group by Deduction.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DeductionGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends DeductionGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: DeductionGroupByArgs['orderBy'] }
      : { orderBy?: DeductionGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, DeductionGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetDeductionGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the Deduction model
 */
readonly fields: DeductionFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for Deduction.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__DeductionClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  member<T extends Prisma.MemberDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.MemberDefaultArgs<ExtArgs>>): Prisma.Prisma__MemberClient<runtime.Types.Result.GetResult<Prisma.$MemberPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  longTermLoan<T extends Prisma.Deduction$longTermLoanArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Deduction$longTermLoanArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>
  shortTermLoan<T extends Prisma.Deduction$shortTermLoanArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Deduction$shortTermLoanArgs<ExtArgs>>): Prisma.Prisma__ShortTermLoanClient<runtime.Types.Result.GetResult<Prisma.$ShortTermLoanPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>
  penalty<T extends Prisma.Deduction$penaltyArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Deduction$penaltyArgs<ExtArgs>>): Prisma.Prisma__PenaltyClient<runtime.Types.Result.GetResult<Prisma.$PenaltyPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>
  savings<T extends Prisma.Deduction$savingsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.Deduction$savingsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$SavingPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the Deduction model
 */
export interface DeductionFieldRefs {
  readonly id: Prisma.FieldRef<"Deduction", 'String'>
  readonly memberId: Prisma.FieldRef<"Deduction", 'String'>
  readonly longTermLoanId: Prisma.FieldRef<"Deduction", 'String'>
  readonly shortTermLoanId: Prisma.FieldRef<"Deduction", 'String'>
  readonly penaltyId: Prisma.FieldRef<"Deduction", 'String'>
  readonly status: Prisma.FieldRef<"Deduction", 'TransactionStatus'>
  readonly createdAt: Prisma.FieldRef<"Deduction", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"Deduction", 'DateTime'>
}
    

// Custom InputTypes
/**
 * Deduction findUnique
 */
export type DeductionFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter, which Deduction to fetch.
   */
  where: Prisma.DeductionWhereUniqueInput
}

/**
 * Deduction findUniqueOrThrow
 */
export type DeductionFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter, which Deduction to fetch.
   */
  where: Prisma.DeductionWhereUniqueInput
}

/**
 * Deduction findFirst
 */
export type DeductionFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter, which Deduction to fetch.
   */
  where?: Prisma.DeductionWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Deductions to fetch.
   */
  orderBy?: Prisma.DeductionOrderByWithRelationInput | Prisma.DeductionOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Deductions.
   */
  cursor?: Prisma.DeductionWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Deductions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Deductions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Deductions.
   */
  distinct?: Prisma.DeductionScalarFieldEnum | Prisma.DeductionScalarFieldEnum[]
}

/**
 * Deduction findFirstOrThrow
 */
export type DeductionFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter, which Deduction to fetch.
   */
  where?: Prisma.DeductionWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Deductions to fetch.
   */
  orderBy?: Prisma.DeductionOrderByWithRelationInput | Prisma.DeductionOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Deductions.
   */
  cursor?: Prisma.DeductionWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Deductions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Deductions.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Deductions.
   */
  distinct?: Prisma.DeductionScalarFieldEnum | Prisma.DeductionScalarFieldEnum[]
}

/**
 * Deduction findMany
 */
export type DeductionFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter, which Deductions to fetch.
   */
  where?: Prisma.DeductionWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Deductions to fetch.
   */
  orderBy?: Prisma.DeductionOrderByWithRelationInput | Prisma.DeductionOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing Deductions.
   */
  cursor?: Prisma.DeductionWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Deductions from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Deductions.
   */
  skip?: number
  distinct?: Prisma.DeductionScalarFieldEnum | Prisma.DeductionScalarFieldEnum[]
}

/**
 * Deduction create
 */
export type DeductionCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * The data needed to create a Deduction.
   */
  data: Prisma.XOR<Prisma.DeductionCreateInput, Prisma.DeductionUncheckedCreateInput>
}

/**
 * Deduction createMany
 */
export type DeductionCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many Deductions.
   */
  data: Prisma.DeductionCreateManyInput | Prisma.DeductionCreateManyInput[]
}

/**
 * Deduction update
 */
export type DeductionUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * The data needed to update a Deduction.
   */
  data: Prisma.XOR<Prisma.DeductionUpdateInput, Prisma.DeductionUncheckedUpdateInput>
  /**
   * Choose, which Deduction to update.
   */
  where: Prisma.DeductionWhereUniqueInput
}

/**
 * Deduction updateMany
 */
export type DeductionUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update Deductions.
   */
  data: Prisma.XOR<Prisma.DeductionUpdateManyMutationInput, Prisma.DeductionUncheckedUpdateManyInput>
  /**
   * Filter which Deductions to update
   */
  where?: Prisma.DeductionWhereInput
  /**
   * Limit how many Deductions to update.
   */
  limit?: number
}

/**
 * Deduction upsert
 */
export type DeductionUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * The filter to search for the Deduction to update in case it exists.
   */
  where: Prisma.DeductionWhereUniqueInput
  /**
   * In case the Deduction found by the `where` argument doesn't exist, create a new Deduction with this data.
   */
  create: Prisma.XOR<Prisma.DeductionCreateInput, Prisma.DeductionUncheckedCreateInput>
  /**
   * In case the Deduction was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.DeductionUpdateInput, Prisma.DeductionUncheckedUpdateInput>
}

/**
 * Deduction delete
 */
export type DeductionDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  /**
   * Filter which Deduction to delete.
   */
  where: Prisma.DeductionWhereUniqueInput
}

/**
 * Deduction deleteMany
 */
export type DeductionDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Deductions to delete
   */
  where?: Prisma.DeductionWhereInput
  /**
   * Limit how many Deductions to delete.
   */
  limit?: number
}

/**
 * Deduction findRaw
 */
export type DeductionFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Deduction aggregateRaw
 */
export type DeductionAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Deduction.longTermLoan
 */
export type Deduction$longTermLoanArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  where?: Prisma.LongTermLoanWhereInput
}

/**
 * Deduction.shortTermLoan
 */
export type Deduction$shortTermLoanArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the ShortTermLoan
   */
  select?: Prisma.ShortTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the ShortTermLoan
   */
  omit?: Prisma.ShortTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ShortTermLoanInclude<ExtArgs> | null
  where?: Prisma.ShortTermLoanWhereInput
}

/**
 * Deduction.penalty
 */
export type Deduction$penaltyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Penalty
   */
  select?: Prisma.PenaltySelect<ExtArgs> | null
  /**
   * Omit specific fields from the Penalty
   */
  omit?: Prisma.PenaltyOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.PenaltyInclude<ExtArgs> | null
  where?: Prisma.PenaltyWhereInput
}

/**
 * Deduction.savings
 */
export type Deduction$savingsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Saving
   */
  select?: Prisma.SavingSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Saving
   */
  omit?: Prisma.SavingOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.SavingInclude<ExtArgs> | null
  where?: Prisma.SavingWhereInput
  orderBy?: Prisma.SavingOrderByWithRelationInput | Prisma.SavingOrderByWithRelationInput[]
  cursor?: Prisma.SavingWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.SavingScalarFieldEnum | Prisma.SavingScalarFieldEnum[]
}

/**
 * Deduction without action
 */
export type DeductionDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
}
