// @prisma/client/models/Dividend.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `Dividend` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model Dividend
 * 
 */
export type DividendModel = runtime.Types.Result.DefaultSelection<Prisma.$DividendPayload>

export type AggregateDividend = {
  _count: DividendCountAggregateOutputType | null
  _avg: DividendAvgAggregateOutputType | null
  _sum: DividendSumAggregateOutputType | null
  _min: DividendMinAggregateOutputType | null
  _max: DividendMaxAggregateOutputType | null
}

export type DividendAvgAggregateOutputType = {
  dividendAmount: number | null
}

export type DividendSumAggregateOutputType = {
  dividendAmount: number | null
}

export type DividendMinAggregateOutputType = {
  id: string | null
  memberId: string | null
  dividendAmount: number | null
  dueDate: Date | null
  paidAt: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type DividendMaxAggregateOutputType = {
  id: string | null
  memberId: string | null
  dividendAmount: number | null
  dueDate: Date | null
  paidAt: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type DividendCountAggregateOutputType = {
  id: number
  memberId: number
  dividendAmount: number
  dueDate: number
  paidAt: number
  status: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type DividendAvgAggregateInputType = {
  dividendAmount?: true
}

export type DividendSumAggregateInputType = {
  dividendAmount?: true
}

export type DividendMinAggregateInputType = {
  id?: true
  memberId?: true
  dividendAmount?: true
  dueDate?: true
  paidAt?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type DividendMaxAggregateInputType = {
  id?: true
  memberId?: true
  dividendAmount?: true
  dueDate?: true
  paidAt?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type DividendCountAggregateInputType = {
  id?: true
  memberId?: true
  dividendAmount?: true
  dueDate?: true
  paidAt?: true
  status?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type DividendAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Dividend to aggregate.
   */
  where?: Prisma.DividendWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Dividends to fetch.
   */
  orderBy?: Prisma.DividendOrderByWithRelationInput | Prisma.DividendOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.DividendWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Dividends from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Dividends.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned Dividends
  **/
  _count?: true | DividendCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: DividendAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: DividendSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: DividendMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: DividendMaxAggregateInputType
}

export type GetDividendAggregateType<T extends DividendAggregateArgs> = {
      [P in keyof T & keyof AggregateDividend]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateDividend[P]>
    : Prisma.GetScalarType<T[P], AggregateDividend[P]>
}




export type DividendGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.DividendWhereInput
  orderBy?: Prisma.DividendOrderByWithAggregationInput | Prisma.DividendOrderByWithAggregationInput[]
  by: Prisma.DividendScalarFieldEnum[] | Prisma.DividendScalarFieldEnum
  having?: Prisma.DividendScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: DividendCountAggregateInputType | true
  _avg?: DividendAvgAggregateInputType
  _sum?: DividendSumAggregateInputType
  _min?: DividendMinAggregateInputType
  _max?: DividendMaxAggregateInputType
}

export type DividendGroupByOutputType = {
  id: string
  memberId: string
  dividendAmount: number
  dueDate: Date | null
  paidAt: Date | null
  status: $Enums.TransactionStatus
  createdAt: Date
  updatedAt: Date
  _count: DividendCountAggregateOutputType | null
  _avg: DividendAvgAggregateOutputType | null
  _sum: DividendSumAggregateOutputType | null
  _min: DividendMinAggregateOutputType | null
  _max: DividendMaxAggregateOutputType | null
}

type GetDividendGroupByPayload<T extends DividendGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<DividendGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof DividendGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], DividendGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], DividendGroupByOutputType[P]>
      }
    >
  >



export type DividendWhereInput = {
  AND?: Prisma.DividendWhereInput | Prisma.DividendWhereInput[]
  OR?: Prisma.DividendWhereInput[]
  NOT?: Prisma.DividendWhereInput | Prisma.DividendWhereInput[]
  id?: Prisma.StringFilter<"Dividend"> | string
  memberId?: Prisma.StringFilter<"Dividend"> | string
  dividendAmount?: Prisma.FloatFilter<"Dividend"> | number
  dueDate?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  paidAt?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Dividend"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
}

export type DividendOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  dividendAmount?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paidAt?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  member?: Prisma.MemberOrderByWithRelationInput
}

export type DividendWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.DividendWhereInput | Prisma.DividendWhereInput[]
  OR?: Prisma.DividendWhereInput[]
  NOT?: Prisma.DividendWhereInput | Prisma.DividendWhereInput[]
  memberId?: Prisma.StringFilter<"Dividend"> | string
  dividendAmount?: Prisma.FloatFilter<"Dividend"> | number
  dueDate?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  paidAt?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Dividend"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
}, "id">

export type DividendOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  dividendAmount?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paidAt?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.DividendCountOrderByAggregateInput
  _avg?: Prisma.DividendAvgOrderByAggregateInput
  _max?: Prisma.DividendMaxOrderByAggregateInput
  _min?: Prisma.DividendMinOrderByAggregateInput
  _sum?: Prisma.DividendSumOrderByAggregateInput
}

export type DividendScalarWhereWithAggregatesInput = {
  AND?: Prisma.DividendScalarWhereWithAggregatesInput | Prisma.DividendScalarWhereWithAggregatesInput[]
  OR?: Prisma.DividendScalarWhereWithAggregatesInput[]
  NOT?: Prisma.DividendScalarWhereWithAggregatesInput | Prisma.DividendScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"Dividend"> | string
  memberId?: Prisma.StringWithAggregatesFilter<"Dividend"> | string
  dividendAmount?: Prisma.FloatWithAggregatesFilter<"Dividend"> | number
  dueDate?: Prisma.DateTimeNullableWithAggregatesFilter<"Dividend"> | Date | string | null
  paidAt?: Prisma.DateTimeNullableWithAggregatesFilter<"Dividend"> | Date | string | null
  status?: Prisma.EnumTransactionStatusWithAggregatesFilter<"Dividend"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"Dividend"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"Dividend"> | Date | string
}

export type DividendCreateInput = {
  id?: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutDividendsInput
}

export type DividendUncheckedCreateInput = {
  id?: string
  memberId: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DividendUpdateInput = {
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutDividendsNestedInput
}

export type DividendUncheckedUpdateInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DividendCreateManyInput = {
  id?: string
  memberId: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DividendUpdateManyMutationInput = {
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DividendUncheckedUpdateManyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DividendListRelationFilter = {
  every?: Prisma.DividendWhereInput
  some?: Prisma.DividendWhereInput
  none?: Prisma.DividendWhereInput
}

export type DividendOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type DividendCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  dividendAmount?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paidAt?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DividendAvgOrderByAggregateInput = {
  dividendAmount?: Prisma.SortOrder
}

export type DividendMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  dividendAmount?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paidAt?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DividendMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  dividendAmount?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paidAt?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type DividendSumOrderByAggregateInput = {
  dividendAmount?: Prisma.SortOrder
}

export type DividendCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput> | Prisma.DividendCreateWithoutMemberInput[] | Prisma.DividendUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DividendCreateOrConnectWithoutMemberInput | Prisma.DividendCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.DividendCreateManyMemberInputEnvelope
  connect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
}

export type DividendUncheckedCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput> | Prisma.DividendCreateWithoutMemberInput[] | Prisma.DividendUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DividendCreateOrConnectWithoutMemberInput | Prisma.DividendCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.DividendCreateManyMemberInputEnvelope
  connect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
}

export type DividendUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput> | Prisma.DividendCreateWithoutMemberInput[] | Prisma.DividendUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DividendCreateOrConnectWithoutMemberInput | Prisma.DividendCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.DividendUpsertWithWhereUniqueWithoutMemberInput | Prisma.DividendUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.DividendCreateManyMemberInputEnvelope
  set?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  disconnect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  delete?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  connect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  update?: Prisma.DividendUpdateWithWhereUniqueWithoutMemberInput | Prisma.DividendUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.DividendUpdateManyWithWhereWithoutMemberInput | Prisma.DividendUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.DividendScalarWhereInput | Prisma.DividendScalarWhereInput[]
}

export type DividendUncheckedUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput> | Prisma.DividendCreateWithoutMemberInput[] | Prisma.DividendUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.DividendCreateOrConnectWithoutMemberInput | Prisma.DividendCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.DividendUpsertWithWhereUniqueWithoutMemberInput | Prisma.DividendUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.DividendCreateManyMemberInputEnvelope
  set?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  disconnect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  delete?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  connect?: Prisma.DividendWhereUniqueInput | Prisma.DividendWhereUniqueInput[]
  update?: Prisma.DividendUpdateWithWhereUniqueWithoutMemberInput | Prisma.DividendUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.DividendUpdateManyWithWhereWithoutMemberInput | Prisma.DividendUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.DividendScalarWhereInput | Prisma.DividendScalarWhereInput[]
}

export type DividendCreateWithoutMemberInput = {
  id?: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DividendUncheckedCreateWithoutMemberInput = {
  id?: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DividendCreateOrConnectWithoutMemberInput = {
  where: Prisma.DividendWhereUniqueInput
  create: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput>
}

export type DividendCreateManyMemberInputEnvelope = {
  data: Prisma.DividendCreateManyMemberInput | Prisma.DividendCreateManyMemberInput[]
}

export type DividendUpsertWithWhereUniqueWithoutMemberInput = {
  where: Prisma.DividendWhereUniqueInput
  update: Prisma.XOR<Prisma.DividendUpdateWithoutMemberInput, Prisma.DividendUncheckedUpdateWithoutMemberInput>
  create: Prisma.XOR<Prisma.DividendCreateWithoutMemberInput, Prisma.DividendUncheckedCreateWithoutMemberInput>
}

export type DividendUpdateWithWhereUniqueWithoutMemberInput = {
  where: Prisma.DividendWhereUniqueInput
  data: Prisma.XOR<Prisma.DividendUpdateWithoutMemberInput, Prisma.DividendUncheckedUpdateWithoutMemberInput>
}

export type DividendUpdateManyWithWhereWithoutMemberInput = {
  where: Prisma.DividendScalarWhereInput
  data: Prisma.XOR<Prisma.DividendUpdateManyMutationInput, Prisma.DividendUncheckedUpdateManyWithoutMemberInput>
}

export type DividendScalarWhereInput = {
  AND?: Prisma.DividendScalarWhereInput | Prisma.DividendScalarWhereInput[]
  OR?: Prisma.DividendScalarWhereInput[]
  NOT?: Prisma.DividendScalarWhereInput | Prisma.DividendScalarWhereInput[]
  id?: Prisma.StringFilter<"Dividend"> | string
  memberId?: Prisma.StringFilter<"Dividend"> | string
  dividendAmount?: Prisma.FloatFilter<"Dividend"> | number
  dueDate?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  paidAt?: Prisma.DateTimeNullableFilter<"Dividend"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"Dividend"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Dividend"> | Date | string
}

export type DividendCreateManyMemberInput = {
  id?: string
  dividendAmount?: number
  dueDate?: Date | string | null
  paidAt?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type DividendUpdateWithoutMemberInput = {
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DividendUncheckedUpdateWithoutMemberInput = {
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type DividendUncheckedUpdateManyWithoutMemberInput = {
  dividendAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  dueDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  paidAt?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}



export type DividendSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  memberId?: boolean
  dividendAmount?: boolean
  dueDate?: boolean
  paidAt?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
}, ExtArgs["result"]["dividend"]>



export type DividendSelectScalar = {
  id?: boolean
  memberId?: boolean
  dividendAmount?: boolean
  dueDate?: boolean
  paidAt?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type DividendOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "memberId" | "dividendAmount" | "dueDate" | "paidAt" | "status" | "createdAt" | "updatedAt", ExtArgs["result"]["dividend"]>
export type DividendInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
}

export type $DividendPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "Dividend"
  objects: {
    member: Prisma.$MemberPayload<ExtArgs>
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    memberId: string
    dividendAmount: number
    dueDate: Date | null
    paidAt: Date | null
    status: $Enums.TransactionStatus
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["dividend"]>
  composites: {}
}

export type DividendGetPayload<S extends boolean | null | undefined | DividendDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$DividendPayload, S>

export type DividendCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<DividendFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: DividendCountAggregateInputType | true
  }

export interface DividendDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['Dividend'], meta: { name: 'Dividend' } }
  /**
   * Find zero or one Dividend that matches the filter.
   * @param {DividendFindUniqueArgs} args - Arguments to find a Dividend
   * @example
   * // Get one Dividend
   * const dividend = await prisma.dividend.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends DividendFindUniqueArgs>(args: Prisma.SelectSubset<T, DividendFindUniqueArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Dividend that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {DividendFindUniqueOrThrowArgs} args - Arguments to find a Dividend
   * @example
   * // Get one Dividend
   * const dividend = await prisma.dividend.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends DividendFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, DividendFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Dividend that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendFindFirstArgs} args - Arguments to find a Dividend
   * @example
   * // Get one Dividend
   * const dividend = await prisma.dividend.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends DividendFindFirstArgs>(args?: Prisma.SelectSubset<T, DividendFindFirstArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Dividend that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendFindFirstOrThrowArgs} args - Arguments to find a Dividend
   * @example
   * // Get one Dividend
   * const dividend = await prisma.dividend.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends DividendFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, DividendFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Dividends that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Dividends
   * const dividends = await prisma.dividend.findMany()
   * 
   * // Get first 10 Dividends
   * const dividends = await prisma.dividend.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const dividendWithIdOnly = await prisma.dividend.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends DividendFindManyArgs>(args?: Prisma.SelectSubset<T, DividendFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Dividend.
   * @param {DividendCreateArgs} args - Arguments to create a Dividend.
   * @example
   * // Create one Dividend
   * const Dividend = await prisma.dividend.create({
   *   data: {
   *     // ... data to create a Dividend
   *   }
   * })
   * 
   */
  create<T extends DividendCreateArgs>(args: Prisma.SelectSubset<T, DividendCreateArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Dividends.
   * @param {DividendCreateManyArgs} args - Arguments to create many Dividends.
   * @example
   * // Create many Dividends
   * const dividend = await prisma.dividend.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends DividendCreateManyArgs>(args?: Prisma.SelectSubset<T, DividendCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a Dividend.
   * @param {DividendDeleteArgs} args - Arguments to delete one Dividend.
   * @example
   * // Delete one Dividend
   * const Dividend = await prisma.dividend.delete({
   *   where: {
   *     // ... filter to delete one Dividend
   *   }
   * })
   * 
   */
  delete<T extends DividendDeleteArgs>(args: Prisma.SelectSubset<T, DividendDeleteArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Dividend.
   * @param {DividendUpdateArgs} args - Arguments to update one Dividend.
   * @example
   * // Update one Dividend
   * const dividend = await prisma.dividend.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends DividendUpdateArgs>(args: Prisma.SelectSubset<T, DividendUpdateArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Dividends.
   * @param {DividendDeleteManyArgs} args - Arguments to filter Dividends to delete.
   * @example
   * // Delete a few Dividends
   * const { count } = await prisma.dividend.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends DividendDeleteManyArgs>(args?: Prisma.SelectSubset<T, DividendDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Dividends.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Dividends
   * const dividend = await prisma.dividend.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends DividendUpdateManyArgs>(args: Prisma.SelectSubset<T, DividendUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one Dividend.
   * @param {DividendUpsertArgs} args - Arguments to update or create a Dividend.
   * @example
   * // Update or create a Dividend
   * const dividend = await prisma.dividend.upsert({
   *   create: {
   *     // ... data to create a Dividend
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Dividend we want to update
   *   }
   * })
   */
  upsert<T extends DividendUpsertArgs>(args: Prisma.SelectSubset<T, DividendUpsertArgs<ExtArgs>>): Prisma.Prisma__DividendClient<runtime.Types.Result.GetResult<Prisma.$DividendPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Dividends that matches the filter.
   * @param {DividendFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const dividend = await prisma.dividend.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.DividendFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a Dividend.
   * @param {DividendAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const dividend = await prisma.dividend.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.DividendAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of Dividends.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendCountArgs} args - Arguments to filter Dividends to count.
   * @example
   * // Count the number of Dividends
   * const count = await prisma.dividend.count({
   *   where: {
   *     // ... the filter for the Dividends we want to count
   *   }
   * })
  **/
  count<T extends DividendCountArgs>(
    args?: Prisma.Subset<T, DividendCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], DividendCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Dividend.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends DividendAggregateArgs>(args: Prisma.Subset<T, DividendAggregateArgs>): Prisma.PrismaPromise<GetDividendAggregateType<T>>

  /**
   * Group by Dividend.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {DividendGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends DividendGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: DividendGroupByArgs['orderBy'] }
      : { orderBy?: DividendGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, DividendGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetDividendGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the Dividend model
 */
readonly fields: DividendFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for Dividend.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__DividendClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  member<T extends Prisma.MemberDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.MemberDefaultArgs<ExtArgs>>): Prisma.Prisma__MemberClient<runtime.Types.Result.GetResult<Prisma.$MemberPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the Dividend model
 */
export interface DividendFieldRefs {
  readonly id: Prisma.FieldRef<"Dividend", 'String'>
  readonly memberId: Prisma.FieldRef<"Dividend", 'String'>
  readonly dividendAmount: Prisma.FieldRef<"Dividend", 'Float'>
  readonly dueDate: Prisma.FieldRef<"Dividend", 'DateTime'>
  readonly paidAt: Prisma.FieldRef<"Dividend", 'DateTime'>
  readonly status: Prisma.FieldRef<"Dividend", 'TransactionStatus'>
  readonly createdAt: Prisma.FieldRef<"Dividend", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"Dividend", 'DateTime'>
}
    

// Custom InputTypes
/**
 * Dividend findUnique
 */
export type DividendFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter, which Dividend to fetch.
   */
  where: Prisma.DividendWhereUniqueInput
}

/**
 * Dividend findUniqueOrThrow
 */
export type DividendFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter, which Dividend to fetch.
   */
  where: Prisma.DividendWhereUniqueInput
}

/**
 * Dividend findFirst
 */
export type DividendFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter, which Dividend to fetch.
   */
  where?: Prisma.DividendWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Dividends to fetch.
   */
  orderBy?: Prisma.DividendOrderByWithRelationInput | Prisma.DividendOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Dividends.
   */
  cursor?: Prisma.DividendWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Dividends from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Dividends.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Dividends.
   */
  distinct?: Prisma.DividendScalarFieldEnum | Prisma.DividendScalarFieldEnum[]
}

/**
 * Dividend findFirstOrThrow
 */
export type DividendFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter, which Dividend to fetch.
   */
  where?: Prisma.DividendWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Dividends to fetch.
   */
  orderBy?: Prisma.DividendOrderByWithRelationInput | Prisma.DividendOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Dividends.
   */
  cursor?: Prisma.DividendWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Dividends from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Dividends.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Dividends.
   */
  distinct?: Prisma.DividendScalarFieldEnum | Prisma.DividendScalarFieldEnum[]
}

/**
 * Dividend findMany
 */
export type DividendFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter, which Dividends to fetch.
   */
  where?: Prisma.DividendWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Dividends to fetch.
   */
  orderBy?: Prisma.DividendOrderByWithRelationInput | Prisma.DividendOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing Dividends.
   */
  cursor?: Prisma.DividendWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Dividends from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Dividends.
   */
  skip?: number
  distinct?: Prisma.DividendScalarFieldEnum | Prisma.DividendScalarFieldEnum[]
}

/**
 * Dividend create
 */
export type DividendCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * The data needed to create a Dividend.
   */
  data: Prisma.XOR<Prisma.DividendCreateInput, Prisma.DividendUncheckedCreateInput>
}

/**
 * Dividend createMany
 */
export type DividendCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many Dividends.
   */
  data: Prisma.DividendCreateManyInput | Prisma.DividendCreateManyInput[]
}

/**
 * Dividend update
 */
export type DividendUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * The data needed to update a Dividend.
   */
  data: Prisma.XOR<Prisma.DividendUpdateInput, Prisma.DividendUncheckedUpdateInput>
  /**
   * Choose, which Dividend to update.
   */
  where: Prisma.DividendWhereUniqueInput
}

/**
 * Dividend updateMany
 */
export type DividendUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update Dividends.
   */
  data: Prisma.XOR<Prisma.DividendUpdateManyMutationInput, Prisma.DividendUncheckedUpdateManyInput>
  /**
   * Filter which Dividends to update
   */
  where?: Prisma.DividendWhereInput
  /**
   * Limit how many Dividends to update.
   */
  limit?: number
}

/**
 * Dividend upsert
 */
export type DividendUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * The filter to search for the Dividend to update in case it exists.
   */
  where: Prisma.DividendWhereUniqueInput
  /**
   * In case the Dividend found by the `where` argument doesn't exist, create a new Dividend with this data.
   */
  create: Prisma.XOR<Prisma.DividendCreateInput, Prisma.DividendUncheckedCreateInput>
  /**
   * In case the Dividend was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.DividendUpdateInput, Prisma.DividendUncheckedUpdateInput>
}

/**
 * Dividend delete
 */
export type DividendDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
  /**
   * Filter which Dividend to delete.
   */
  where: Prisma.DividendWhereUniqueInput
}

/**
 * Dividend deleteMany
 */
export type DividendDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Dividends to delete
   */
  where?: Prisma.DividendWhereInput
  /**
   * Limit how many Dividends to delete.
   */
  limit?: number
}

/**
 * Dividend findRaw
 */
export type DividendFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Dividend aggregateRaw
 */
export type DividendAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Dividend without action
 */
export type DividendDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Dividend
   */
  select?: Prisma.DividendSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Dividend
   */
  omit?: Prisma.DividendOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DividendInclude<ExtArgs> | null
}
