// @prisma/client/models/LongTermLoan.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `LongTermLoan` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model LongTermLoan
 * 
 */
export type LongTermLoanModel = runtime.Types.Result.DefaultSelection<Prisma.$LongTermLoanPayload>

export type AggregateLongTermLoan = {
  _count: LongTermLoanCountAggregateOutputType | null
  _avg: LongTermLoanAvgAggregateOutputType | null
  _sum: LongTermLoanSumAggregateOutputType | null
  _min: LongTermLoanMinAggregateOutputType | null
  _max: LongTermLoanMaxAggregateOutputType | null
}

export type LongTermLoanAvgAggregateOutputType = {
  principle: number | null
  loanAmount: number | null
  interest: number | null
  termMonths: number | null
}

export type LongTermLoanSumAggregateOutputType = {
  principle: number | null
  loanAmount: number | null
  interest: number | null
  termMonths: number | null
}

export type LongTermLoanMinAggregateOutputType = {
  id: string | null
  memberId: string | null
  penaltyId: string | null
  principle: number | null
  loanAmount: number | null
  interest: number | null
  termMonths: number | null
  dueDate: Date | null
  paymentDate: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type LongTermLoanMaxAggregateOutputType = {
  id: string | null
  memberId: string | null
  penaltyId: string | null
  principle: number | null
  loanAmount: number | null
  interest: number | null
  termMonths: number | null
  dueDate: Date | null
  paymentDate: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type LongTermLoanCountAggregateOutputType = {
  id: number
  memberId: number
  penaltyId: number
  principle: number
  loanAmount: number
  interest: number
  termMonths: number
  dueDate: number
  paymentDate: number
  status: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type LongTermLoanAvgAggregateInputType = {
  principle?: true
  loanAmount?: true
  interest?: true
  termMonths?: true
}

export type LongTermLoanSumAggregateInputType = {
  principle?: true
  loanAmount?: true
  interest?: true
  termMonths?: true
}

export type LongTermLoanMinAggregateInputType = {
  id?: true
  memberId?: true
  penaltyId?: true
  principle?: true
  loanAmount?: true
  interest?: true
  termMonths?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type LongTermLoanMaxAggregateInputType = {
  id?: true
  memberId?: true
  penaltyId?: true
  principle?: true
  loanAmount?: true
  interest?: true
  termMonths?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type LongTermLoanCountAggregateInputType = {
  id?: true
  memberId?: true
  penaltyId?: true
  principle?: true
  loanAmount?: true
  interest?: true
  termMonths?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type LongTermLoanAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which LongTermLoan to aggregate.
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoans to fetch.
   */
  orderBy?: Prisma.LongTermLoanOrderByWithRelationInput | Prisma.LongTermLoanOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.LongTermLoanWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoans from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoans.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned LongTermLoans
  **/
  _count?: true | LongTermLoanCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: LongTermLoanAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: LongTermLoanSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: LongTermLoanMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: LongTermLoanMaxAggregateInputType
}

export type GetLongTermLoanAggregateType<T extends LongTermLoanAggregateArgs> = {
      [P in keyof T & keyof AggregateLongTermLoan]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateLongTermLoan[P]>
    : Prisma.GetScalarType<T[P], AggregateLongTermLoan[P]>
}




export type LongTermLoanGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.LongTermLoanWhereInput
  orderBy?: Prisma.LongTermLoanOrderByWithAggregationInput | Prisma.LongTermLoanOrderByWithAggregationInput[]
  by: Prisma.LongTermLoanScalarFieldEnum[] | Prisma.LongTermLoanScalarFieldEnum
  having?: Prisma.LongTermLoanScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: LongTermLoanCountAggregateInputType | true
  _avg?: LongTermLoanAvgAggregateInputType
  _sum?: LongTermLoanSumAggregateInputType
  _min?: LongTermLoanMinAggregateInputType
  _max?: LongTermLoanMaxAggregateInputType
}

export type LongTermLoanGroupByOutputType = {
  id: string
  memberId: string
  penaltyId: string | null
  principle: number
  loanAmount: number
  interest: number
  termMonths: number
  dueDate: Date
  paymentDate: Date | null
  status: $Enums.TransactionStatus
  createdAt: Date
  updatedAt: Date
  _count: LongTermLoanCountAggregateOutputType | null
  _avg: LongTermLoanAvgAggregateOutputType | null
  _sum: LongTermLoanSumAggregateOutputType | null
  _min: LongTermLoanMinAggregateOutputType | null
  _max: LongTermLoanMaxAggregateOutputType | null
}

type GetLongTermLoanGroupByPayload<T extends LongTermLoanGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<LongTermLoanGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof LongTermLoanGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], LongTermLoanGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], LongTermLoanGroupByOutputType[P]>
      }
    >
  >



export type LongTermLoanWhereInput = {
  AND?: Prisma.LongTermLoanWhereInput | Prisma.LongTermLoanWhereInput[]
  OR?: Prisma.LongTermLoanWhereInput[]
  NOT?: Prisma.LongTermLoanWhereInput | Prisma.LongTermLoanWhereInput[]
  id?: Prisma.StringFilter<"LongTermLoan"> | string
  memberId?: Prisma.StringFilter<"LongTermLoan"> | string
  penaltyId?: Prisma.StringNullableFilter<"LongTermLoan"> | string | null
  principle?: Prisma.FloatFilter<"LongTermLoan"> | number
  loanAmount?: Prisma.FloatFilter<"LongTermLoan"> | number
  interest?: Prisma.FloatFilter<"LongTermLoan"> | number
  termMonths?: Prisma.IntFilter<"LongTermLoan"> | number
  dueDate?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoan"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoan"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
  penalty?: Prisma.XOR<Prisma.PenaltyNullableScalarRelationFilter, Prisma.PenaltyWhereInput> | null
  deductions?: Prisma.DeductionListRelationFilter
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentListRelationFilter
}

export type LongTermLoanOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  member?: Prisma.MemberOrderByWithRelationInput
  penalty?: Prisma.PenaltyOrderByWithRelationInput
  deductions?: Prisma.DeductionOrderByRelationAggregateInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentOrderByRelationAggregateInput
}

export type LongTermLoanWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.LongTermLoanWhereInput | Prisma.LongTermLoanWhereInput[]
  OR?: Prisma.LongTermLoanWhereInput[]
  NOT?: Prisma.LongTermLoanWhereInput | Prisma.LongTermLoanWhereInput[]
  memberId?: Prisma.StringFilter<"LongTermLoan"> | string
  penaltyId?: Prisma.StringNullableFilter<"LongTermLoan"> | string | null
  principle?: Prisma.FloatFilter<"LongTermLoan"> | number
  loanAmount?: Prisma.FloatFilter<"LongTermLoan"> | number
  interest?: Prisma.FloatFilter<"LongTermLoan"> | number
  termMonths?: Prisma.IntFilter<"LongTermLoan"> | number
  dueDate?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoan"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoan"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  member?: Prisma.XOR<Prisma.MemberScalarRelationFilter, Prisma.MemberWhereInput>
  penalty?: Prisma.XOR<Prisma.PenaltyNullableScalarRelationFilter, Prisma.PenaltyWhereInput> | null
  deductions?: Prisma.DeductionListRelationFilter
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentListRelationFilter
}, "id">

export type LongTermLoanOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.LongTermLoanCountOrderByAggregateInput
  _avg?: Prisma.LongTermLoanAvgOrderByAggregateInput
  _max?: Prisma.LongTermLoanMaxOrderByAggregateInput
  _min?: Prisma.LongTermLoanMinOrderByAggregateInput
  _sum?: Prisma.LongTermLoanSumOrderByAggregateInput
}

export type LongTermLoanScalarWhereWithAggregatesInput = {
  AND?: Prisma.LongTermLoanScalarWhereWithAggregatesInput | Prisma.LongTermLoanScalarWhereWithAggregatesInput[]
  OR?: Prisma.LongTermLoanScalarWhereWithAggregatesInput[]
  NOT?: Prisma.LongTermLoanScalarWhereWithAggregatesInput | Prisma.LongTermLoanScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"LongTermLoan"> | string
  memberId?: Prisma.StringWithAggregatesFilter<"LongTermLoan"> | string
  penaltyId?: Prisma.StringNullableWithAggregatesFilter<"LongTermLoan"> | string | null
  principle?: Prisma.FloatWithAggregatesFilter<"LongTermLoan"> | number
  loanAmount?: Prisma.FloatWithAggregatesFilter<"LongTermLoan"> | number
  interest?: Prisma.FloatWithAggregatesFilter<"LongTermLoan"> | number
  termMonths?: Prisma.IntWithAggregatesFilter<"LongTermLoan"> | number
  dueDate?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoan"> | Date | string
  paymentDate?: Prisma.DateTimeNullableWithAggregatesFilter<"LongTermLoan"> | Date | string | null
  status?: Prisma.EnumTransactionStatusWithAggregatesFilter<"LongTermLoan"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoan"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoan"> | Date | string
}

export type LongTermLoanCreateInput = {
  id?: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutLongTermLoansInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutLongTermLoansInput
  deductions?: Prisma.DeductionCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUncheckedCreateInput = {
  id?: string
  memberId: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  deductions?: Prisma.DeductionUncheckedCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUpdateInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutLongTermLoansNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutLongTermLoansNestedInput
  deductions?: Prisma.DeductionUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  deductions?: Prisma.DeductionUncheckedUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanCreateManyInput = {
  id?: string
  memberId: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanUpdateManyMutationInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanUncheckedUpdateManyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanListRelationFilter = {
  every?: Prisma.LongTermLoanWhereInput
  some?: Prisma.LongTermLoanWhereInput
  none?: Prisma.LongTermLoanWhereInput
}

export type LongTermLoanOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type LongTermLoanNullableScalarRelationFilter = {
  is?: Prisma.LongTermLoanWhereInput | null
  isNot?: Prisma.LongTermLoanWhereInput | null
}

export type LongTermLoanCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanAvgOrderByAggregateInput = {
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
}

export type LongTermLoanMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  memberId?: Prisma.SortOrder
  penaltyId?: Prisma.SortOrder
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanSumOrderByAggregateInput = {
  principle?: Prisma.SortOrder
  loanAmount?: Prisma.SortOrder
  interest?: Prisma.SortOrder
  termMonths?: Prisma.SortOrder
}

export type LongTermLoanScalarRelationFilter = {
  is?: Prisma.LongTermLoanWhereInput
  isNot?: Prisma.LongTermLoanWhereInput
}

export type LongTermLoanCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput> | Prisma.LongTermLoanCreateWithoutMemberInput[] | Prisma.LongTermLoanUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutMemberInput | Prisma.LongTermLoanCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.LongTermLoanCreateManyMemberInputEnvelope
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
}

export type LongTermLoanUncheckedCreateNestedManyWithoutMemberInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput> | Prisma.LongTermLoanCreateWithoutMemberInput[] | Prisma.LongTermLoanUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutMemberInput | Prisma.LongTermLoanCreateOrConnectWithoutMemberInput[]
  createMany?: Prisma.LongTermLoanCreateManyMemberInputEnvelope
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
}

export type LongTermLoanUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput> | Prisma.LongTermLoanCreateWithoutMemberInput[] | Prisma.LongTermLoanUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutMemberInput | Prisma.LongTermLoanCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.LongTermLoanUpsertWithWhereUniqueWithoutMemberInput | Prisma.LongTermLoanUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.LongTermLoanCreateManyMemberInputEnvelope
  set?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  delete?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  update?: Prisma.LongTermLoanUpdateWithWhereUniqueWithoutMemberInput | Prisma.LongTermLoanUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.LongTermLoanUpdateManyWithWhereWithoutMemberInput | Prisma.LongTermLoanUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
}

export type LongTermLoanUncheckedUpdateManyWithoutMemberNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput> | Prisma.LongTermLoanCreateWithoutMemberInput[] | Prisma.LongTermLoanUncheckedCreateWithoutMemberInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutMemberInput | Prisma.LongTermLoanCreateOrConnectWithoutMemberInput[]
  upsert?: Prisma.LongTermLoanUpsertWithWhereUniqueWithoutMemberInput | Prisma.LongTermLoanUpsertWithWhereUniqueWithoutMemberInput[]
  createMany?: Prisma.LongTermLoanCreateManyMemberInputEnvelope
  set?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  delete?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  update?: Prisma.LongTermLoanUpdateWithWhereUniqueWithoutMemberInput | Prisma.LongTermLoanUpdateWithWhereUniqueWithoutMemberInput[]
  updateMany?: Prisma.LongTermLoanUpdateManyWithWhereWithoutMemberInput | Prisma.LongTermLoanUpdateManyWithWhereWithoutMemberInput[]
  deleteMany?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
}

export type LongTermLoanCreateNestedOneWithoutDeductionsInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedCreateWithoutDeductionsInput>
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutDeductionsInput
  connect?: Prisma.LongTermLoanWhereUniqueInput
}

export type LongTermLoanUpdateOneWithoutDeductionsNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedCreateWithoutDeductionsInput>
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutDeductionsInput
  upsert?: Prisma.LongTermLoanUpsertWithoutDeductionsInput
  disconnect?: boolean
  delete?: Prisma.LongTermLoanWhereInput | boolean
  connect?: Prisma.LongTermLoanWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.LongTermLoanUpdateToOneWithWhereWithoutDeductionsInput, Prisma.LongTermLoanUpdateWithoutDeductionsInput>, Prisma.LongTermLoanUncheckedUpdateWithoutDeductionsInput>
}

export type LongTermLoanCreateNestedOneWithoutLongTermLoanRepaymentsInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedCreateWithoutLongTermLoanRepaymentsInput>
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutLongTermLoanRepaymentsInput
  connect?: Prisma.LongTermLoanWhereUniqueInput
}

export type LongTermLoanUpdateOneRequiredWithoutLongTermLoanRepaymentsNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedCreateWithoutLongTermLoanRepaymentsInput>
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutLongTermLoanRepaymentsInput
  upsert?: Prisma.LongTermLoanUpsertWithoutLongTermLoanRepaymentsInput
  connect?: Prisma.LongTermLoanWhereUniqueInput
  update?: Prisma.XOR<Prisma.XOR<Prisma.LongTermLoanUpdateToOneWithWhereWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUpdateWithoutLongTermLoanRepaymentsInput>, Prisma.LongTermLoanUncheckedUpdateWithoutLongTermLoanRepaymentsInput>
}

export type LongTermLoanCreateNestedManyWithoutPenaltyInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput> | Prisma.LongTermLoanCreateWithoutPenaltyInput[] | Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput | Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput[]
  createMany?: Prisma.LongTermLoanCreateManyPenaltyInputEnvelope
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
}

export type LongTermLoanUncheckedCreateNestedManyWithoutPenaltyInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput> | Prisma.LongTermLoanCreateWithoutPenaltyInput[] | Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput | Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput[]
  createMany?: Prisma.LongTermLoanCreateManyPenaltyInputEnvelope
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
}

export type LongTermLoanUpdateManyWithoutPenaltyNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput> | Prisma.LongTermLoanCreateWithoutPenaltyInput[] | Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput | Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput[]
  upsert?: Prisma.LongTermLoanUpsertWithWhereUniqueWithoutPenaltyInput | Prisma.LongTermLoanUpsertWithWhereUniqueWithoutPenaltyInput[]
  createMany?: Prisma.LongTermLoanCreateManyPenaltyInputEnvelope
  set?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  delete?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  update?: Prisma.LongTermLoanUpdateWithWhereUniqueWithoutPenaltyInput | Prisma.LongTermLoanUpdateWithWhereUniqueWithoutPenaltyInput[]
  updateMany?: Prisma.LongTermLoanUpdateManyWithWhereWithoutPenaltyInput | Prisma.LongTermLoanUpdateManyWithWhereWithoutPenaltyInput[]
  deleteMany?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
}

export type LongTermLoanUncheckedUpdateManyWithoutPenaltyNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput> | Prisma.LongTermLoanCreateWithoutPenaltyInput[] | Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput[]
  connectOrCreate?: Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput | Prisma.LongTermLoanCreateOrConnectWithoutPenaltyInput[]
  upsert?: Prisma.LongTermLoanUpsertWithWhereUniqueWithoutPenaltyInput | Prisma.LongTermLoanUpsertWithWhereUniqueWithoutPenaltyInput[]
  createMany?: Prisma.LongTermLoanCreateManyPenaltyInputEnvelope
  set?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  delete?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  connect?: Prisma.LongTermLoanWhereUniqueInput | Prisma.LongTermLoanWhereUniqueInput[]
  update?: Prisma.LongTermLoanUpdateWithWhereUniqueWithoutPenaltyInput | Prisma.LongTermLoanUpdateWithWhereUniqueWithoutPenaltyInput[]
  updateMany?: Prisma.LongTermLoanUpdateManyWithWhereWithoutPenaltyInput | Prisma.LongTermLoanUpdateManyWithWhereWithoutPenaltyInput[]
  deleteMany?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
}

export type LongTermLoanCreateWithoutMemberInput = {
  id?: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  penalty?: Prisma.PenaltyCreateNestedOneWithoutLongTermLoansInput
  deductions?: Prisma.DeductionCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUncheckedCreateWithoutMemberInput = {
  id?: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  deductions?: Prisma.DeductionUncheckedCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanCreateOrConnectWithoutMemberInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput>
}

export type LongTermLoanCreateManyMemberInputEnvelope = {
  data: Prisma.LongTermLoanCreateManyMemberInput | Prisma.LongTermLoanCreateManyMemberInput[]
}

export type LongTermLoanUpsertWithWhereUniqueWithoutMemberInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  update: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutMemberInput, Prisma.LongTermLoanUncheckedUpdateWithoutMemberInput>
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutMemberInput, Prisma.LongTermLoanUncheckedCreateWithoutMemberInput>
}

export type LongTermLoanUpdateWithWhereUniqueWithoutMemberInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutMemberInput, Prisma.LongTermLoanUncheckedUpdateWithoutMemberInput>
}

export type LongTermLoanUpdateManyWithWhereWithoutMemberInput = {
  where: Prisma.LongTermLoanScalarWhereInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateManyMutationInput, Prisma.LongTermLoanUncheckedUpdateManyWithoutMemberInput>
}

export type LongTermLoanScalarWhereInput = {
  AND?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
  OR?: Prisma.LongTermLoanScalarWhereInput[]
  NOT?: Prisma.LongTermLoanScalarWhereInput | Prisma.LongTermLoanScalarWhereInput[]
  id?: Prisma.StringFilter<"LongTermLoan"> | string
  memberId?: Prisma.StringFilter<"LongTermLoan"> | string
  penaltyId?: Prisma.StringNullableFilter<"LongTermLoan"> | string | null
  principle?: Prisma.FloatFilter<"LongTermLoan"> | number
  loanAmount?: Prisma.FloatFilter<"LongTermLoan"> | number
  interest?: Prisma.FloatFilter<"LongTermLoan"> | number
  termMonths?: Prisma.IntFilter<"LongTermLoan"> | number
  dueDate?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoan"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoan"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoan"> | Date | string
}

export type LongTermLoanCreateWithoutDeductionsInput = {
  id?: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutLongTermLoansInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutLongTermLoansInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUncheckedCreateWithoutDeductionsInput = {
  id?: string
  memberId: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanCreateOrConnectWithoutDeductionsInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedCreateWithoutDeductionsInput>
}

export type LongTermLoanUpsertWithoutDeductionsInput = {
  update: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedUpdateWithoutDeductionsInput>
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedCreateWithoutDeductionsInput>
  where?: Prisma.LongTermLoanWhereInput
}

export type LongTermLoanUpdateToOneWithWhereWithoutDeductionsInput = {
  where?: Prisma.LongTermLoanWhereInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutDeductionsInput, Prisma.LongTermLoanUncheckedUpdateWithoutDeductionsInput>
}

export type LongTermLoanUpdateWithoutDeductionsInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutLongTermLoansNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutLongTermLoansNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateWithoutDeductionsInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanCreateWithoutLongTermLoanRepaymentsInput = {
  id?: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutLongTermLoansInput
  penalty?: Prisma.PenaltyCreateNestedOneWithoutLongTermLoansInput
  deductions?: Prisma.DeductionCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUncheckedCreateWithoutLongTermLoanRepaymentsInput = {
  id?: string
  memberId: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  deductions?: Prisma.DeductionUncheckedCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanCreateOrConnectWithoutLongTermLoanRepaymentsInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedCreateWithoutLongTermLoanRepaymentsInput>
}

export type LongTermLoanUpsertWithoutLongTermLoanRepaymentsInput = {
  update: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedUpdateWithoutLongTermLoanRepaymentsInput>
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedCreateWithoutLongTermLoanRepaymentsInput>
  where?: Prisma.LongTermLoanWhereInput
}

export type LongTermLoanUpdateToOneWithWhereWithoutLongTermLoanRepaymentsInput = {
  where?: Prisma.LongTermLoanWhereInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutLongTermLoanRepaymentsInput, Prisma.LongTermLoanUncheckedUpdateWithoutLongTermLoanRepaymentsInput>
}

export type LongTermLoanUpdateWithoutLongTermLoanRepaymentsInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutLongTermLoansNestedInput
  penalty?: Prisma.PenaltyUpdateOneWithoutLongTermLoansNestedInput
  deductions?: Prisma.DeductionUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateWithoutLongTermLoanRepaymentsInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  deductions?: Prisma.DeductionUncheckedUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanCreateWithoutPenaltyInput = {
  id?: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  member: Prisma.MemberCreateNestedOneWithoutLongTermLoansInput
  deductions?: Prisma.DeductionCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanUncheckedCreateWithoutPenaltyInput = {
  id?: string
  memberId: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  deductions?: Prisma.DeductionUncheckedCreateNestedManyWithoutLongTermLoanInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedCreateNestedManyWithoutLongTermLoanInput
}

export type LongTermLoanCreateOrConnectWithoutPenaltyInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput>
}

export type LongTermLoanCreateManyPenaltyInputEnvelope = {
  data: Prisma.LongTermLoanCreateManyPenaltyInput | Prisma.LongTermLoanCreateManyPenaltyInput[]
}

export type LongTermLoanUpsertWithWhereUniqueWithoutPenaltyInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  update: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedUpdateWithoutPenaltyInput>
  create: Prisma.XOR<Prisma.LongTermLoanCreateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedCreateWithoutPenaltyInput>
}

export type LongTermLoanUpdateWithWhereUniqueWithoutPenaltyInput = {
  where: Prisma.LongTermLoanWhereUniqueInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateWithoutPenaltyInput, Prisma.LongTermLoanUncheckedUpdateWithoutPenaltyInput>
}

export type LongTermLoanUpdateManyWithWhereWithoutPenaltyInput = {
  where: Prisma.LongTermLoanScalarWhereInput
  data: Prisma.XOR<Prisma.LongTermLoanUpdateManyMutationInput, Prisma.LongTermLoanUncheckedUpdateManyWithoutPenaltyInput>
}

export type LongTermLoanCreateManyMemberInput = {
  id?: string
  penaltyId?: string | null
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanUpdateWithoutMemberInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  penalty?: Prisma.PenaltyUpdateOneWithoutLongTermLoansNestedInput
  deductions?: Prisma.DeductionUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateWithoutMemberInput = {
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  deductions?: Prisma.DeductionUncheckedUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateManyWithoutMemberInput = {
  penaltyId?: Prisma.NullableStringFieldUpdateOperationsInput | string | null
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanCreateManyPenaltyInput = {
  id?: string
  memberId: string
  principle?: number
  loanAmount?: number
  interest?: number
  termMonths?: number
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanUpdateWithoutPenaltyInput = {
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  member?: Prisma.MemberUpdateOneRequiredWithoutLongTermLoansNestedInput
  deductions?: Prisma.DeductionUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateWithoutPenaltyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  deductions?: Prisma.DeductionUncheckedUpdateManyWithoutLongTermLoanNestedInput
  longTermLoanRepayments?: Prisma.LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanNestedInput
}

export type LongTermLoanUncheckedUpdateManyWithoutPenaltyInput = {
  memberId?: Prisma.StringFieldUpdateOperationsInput | string
  principle?: Prisma.FloatFieldUpdateOperationsInput | number
  loanAmount?: Prisma.FloatFieldUpdateOperationsInput | number
  interest?: Prisma.FloatFieldUpdateOperationsInput | number
  termMonths?: Prisma.IntFieldUpdateOperationsInput | number
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}


/**
 * Count Type LongTermLoanCountOutputType
 */

export type LongTermLoanCountOutputType = {
  deductions: number
  longTermLoanRepayments: number
}

export type LongTermLoanCountOutputTypeSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  deductions?: boolean | LongTermLoanCountOutputTypeCountDeductionsArgs
  longTermLoanRepayments?: boolean | LongTermLoanCountOutputTypeCountLongTermLoanRepaymentsArgs
}

/**
 * LongTermLoanCountOutputType without action
 */
export type LongTermLoanCountOutputTypeDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanCountOutputType
   */
  select?: Prisma.LongTermLoanCountOutputTypeSelect<ExtArgs> | null
}

/**
 * LongTermLoanCountOutputType without action
 */
export type LongTermLoanCountOutputTypeCountDeductionsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.DeductionWhereInput
}

/**
 * LongTermLoanCountOutputType without action
 */
export type LongTermLoanCountOutputTypeCountLongTermLoanRepaymentsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.LongTermLoanRepaymentWhereInput
}


export type LongTermLoanSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  memberId?: boolean
  penaltyId?: boolean
  principle?: boolean
  loanAmount?: boolean
  interest?: boolean
  termMonths?: boolean
  dueDate?: boolean
  paymentDate?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
  penalty?: boolean | Prisma.LongTermLoan$penaltyArgs<ExtArgs>
  deductions?: boolean | Prisma.LongTermLoan$deductionsArgs<ExtArgs>
  longTermLoanRepayments?: boolean | Prisma.LongTermLoan$longTermLoanRepaymentsArgs<ExtArgs>
  _count?: boolean | Prisma.LongTermLoanCountOutputTypeDefaultArgs<ExtArgs>
}, ExtArgs["result"]["longTermLoan"]>



export type LongTermLoanSelectScalar = {
  id?: boolean
  memberId?: boolean
  penaltyId?: boolean
  principle?: boolean
  loanAmount?: boolean
  interest?: boolean
  termMonths?: boolean
  dueDate?: boolean
  paymentDate?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type LongTermLoanOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "memberId" | "penaltyId" | "principle" | "loanAmount" | "interest" | "termMonths" | "dueDate" | "paymentDate" | "status" | "createdAt" | "updatedAt", ExtArgs["result"]["longTermLoan"]>
export type LongTermLoanInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  member?: boolean | Prisma.MemberDefaultArgs<ExtArgs>
  penalty?: boolean | Prisma.LongTermLoan$penaltyArgs<ExtArgs>
  deductions?: boolean | Prisma.LongTermLoan$deductionsArgs<ExtArgs>
  longTermLoanRepayments?: boolean | Prisma.LongTermLoan$longTermLoanRepaymentsArgs<ExtArgs>
  _count?: boolean | Prisma.LongTermLoanCountOutputTypeDefaultArgs<ExtArgs>
}

export type $LongTermLoanPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "LongTermLoan"
  objects: {
    member: Prisma.$MemberPayload<ExtArgs>
    penalty: Prisma.$PenaltyPayload<ExtArgs> | null
    deductions: Prisma.$DeductionPayload<ExtArgs>[]
    longTermLoanRepayments: Prisma.$LongTermLoanRepaymentPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    memberId: string
    penaltyId: string | null
    principle: number
    loanAmount: number
    interest: number
    termMonths: number
    dueDate: Date
    paymentDate: Date | null
    status: $Enums.TransactionStatus
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["longTermLoan"]>
  composites: {}
}

export type LongTermLoanGetPayload<S extends boolean | null | undefined | LongTermLoanDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload, S>

export type LongTermLoanCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<LongTermLoanFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: LongTermLoanCountAggregateInputType | true
  }

export interface LongTermLoanDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['LongTermLoan'], meta: { name: 'LongTermLoan' } }
  /**
   * Find zero or one LongTermLoan that matches the filter.
   * @param {LongTermLoanFindUniqueArgs} args - Arguments to find a LongTermLoan
   * @example
   * // Get one LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends LongTermLoanFindUniqueArgs>(args: Prisma.SelectSubset<T, LongTermLoanFindUniqueArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one LongTermLoan that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {LongTermLoanFindUniqueOrThrowArgs} args - Arguments to find a LongTermLoan
   * @example
   * // Get one LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends LongTermLoanFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, LongTermLoanFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first LongTermLoan that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanFindFirstArgs} args - Arguments to find a LongTermLoan
   * @example
   * // Get one LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends LongTermLoanFindFirstArgs>(args?: Prisma.SelectSubset<T, LongTermLoanFindFirstArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first LongTermLoan that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanFindFirstOrThrowArgs} args - Arguments to find a LongTermLoan
   * @example
   * // Get one LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends LongTermLoanFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, LongTermLoanFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more LongTermLoans that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all LongTermLoans
   * const longTermLoans = await prisma.longTermLoan.findMany()
   * 
   * // Get first 10 LongTermLoans
   * const longTermLoans = await prisma.longTermLoan.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const longTermLoanWithIdOnly = await prisma.longTermLoan.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends LongTermLoanFindManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a LongTermLoan.
   * @param {LongTermLoanCreateArgs} args - Arguments to create a LongTermLoan.
   * @example
   * // Create one LongTermLoan
   * const LongTermLoan = await prisma.longTermLoan.create({
   *   data: {
   *     // ... data to create a LongTermLoan
   *   }
   * })
   * 
   */
  create<T extends LongTermLoanCreateArgs>(args: Prisma.SelectSubset<T, LongTermLoanCreateArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many LongTermLoans.
   * @param {LongTermLoanCreateManyArgs} args - Arguments to create many LongTermLoans.
   * @example
   * // Create many LongTermLoans
   * const longTermLoan = await prisma.longTermLoan.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends LongTermLoanCreateManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a LongTermLoan.
   * @param {LongTermLoanDeleteArgs} args - Arguments to delete one LongTermLoan.
   * @example
   * // Delete one LongTermLoan
   * const LongTermLoan = await prisma.longTermLoan.delete({
   *   where: {
   *     // ... filter to delete one LongTermLoan
   *   }
   * })
   * 
   */
  delete<T extends LongTermLoanDeleteArgs>(args: Prisma.SelectSubset<T, LongTermLoanDeleteArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one LongTermLoan.
   * @param {LongTermLoanUpdateArgs} args - Arguments to update one LongTermLoan.
   * @example
   * // Update one LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends LongTermLoanUpdateArgs>(args: Prisma.SelectSubset<T, LongTermLoanUpdateArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more LongTermLoans.
   * @param {LongTermLoanDeleteManyArgs} args - Arguments to filter LongTermLoans to delete.
   * @example
   * // Delete a few LongTermLoans
   * const { count } = await prisma.longTermLoan.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends LongTermLoanDeleteManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more LongTermLoans.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many LongTermLoans
   * const longTermLoan = await prisma.longTermLoan.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends LongTermLoanUpdateManyArgs>(args: Prisma.SelectSubset<T, LongTermLoanUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one LongTermLoan.
   * @param {LongTermLoanUpsertArgs} args - Arguments to update or create a LongTermLoan.
   * @example
   * // Update or create a LongTermLoan
   * const longTermLoan = await prisma.longTermLoan.upsert({
   *   create: {
   *     // ... data to create a LongTermLoan
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the LongTermLoan we want to update
   *   }
   * })
   */
  upsert<T extends LongTermLoanUpsertArgs>(args: Prisma.SelectSubset<T, LongTermLoanUpsertArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more LongTermLoans that matches the filter.
   * @param {LongTermLoanFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const longTermLoan = await prisma.longTermLoan.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.LongTermLoanFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a LongTermLoan.
   * @param {LongTermLoanAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const longTermLoan = await prisma.longTermLoan.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.LongTermLoanAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of LongTermLoans.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanCountArgs} args - Arguments to filter LongTermLoans to count.
   * @example
   * // Count the number of LongTermLoans
   * const count = await prisma.longTermLoan.count({
   *   where: {
   *     // ... the filter for the LongTermLoans we want to count
   *   }
   * })
  **/
  count<T extends LongTermLoanCountArgs>(
    args?: Prisma.Subset<T, LongTermLoanCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], LongTermLoanCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a LongTermLoan.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends LongTermLoanAggregateArgs>(args: Prisma.Subset<T, LongTermLoanAggregateArgs>): Prisma.PrismaPromise<GetLongTermLoanAggregateType<T>>

  /**
   * Group by LongTermLoan.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends LongTermLoanGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: LongTermLoanGroupByArgs['orderBy'] }
      : { orderBy?: LongTermLoanGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, LongTermLoanGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetLongTermLoanGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the LongTermLoan model
 */
readonly fields: LongTermLoanFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for LongTermLoan.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__LongTermLoanClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  member<T extends Prisma.MemberDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.MemberDefaultArgs<ExtArgs>>): Prisma.Prisma__MemberClient<runtime.Types.Result.GetResult<Prisma.$MemberPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  penalty<T extends Prisma.LongTermLoan$penaltyArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.LongTermLoan$penaltyArgs<ExtArgs>>): Prisma.Prisma__PenaltyClient<runtime.Types.Result.GetResult<Prisma.$PenaltyPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>
  deductions<T extends Prisma.LongTermLoan$deductionsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.LongTermLoan$deductionsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$DeductionPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  longTermLoanRepayments<T extends Prisma.LongTermLoan$longTermLoanRepaymentsArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.LongTermLoan$longTermLoanRepaymentsArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findMany", GlobalOmitOptions> | Null>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the LongTermLoan model
 */
export interface LongTermLoanFieldRefs {
  readonly id: Prisma.FieldRef<"LongTermLoan", 'String'>
  readonly memberId: Prisma.FieldRef<"LongTermLoan", 'String'>
  readonly penaltyId: Prisma.FieldRef<"LongTermLoan", 'String'>
  readonly principle: Prisma.FieldRef<"LongTermLoan", 'Float'>
  readonly loanAmount: Prisma.FieldRef<"LongTermLoan", 'Float'>
  readonly interest: Prisma.FieldRef<"LongTermLoan", 'Float'>
  readonly termMonths: Prisma.FieldRef<"LongTermLoan", 'Int'>
  readonly dueDate: Prisma.FieldRef<"LongTermLoan", 'DateTime'>
  readonly paymentDate: Prisma.FieldRef<"LongTermLoan", 'DateTime'>
  readonly status: Prisma.FieldRef<"LongTermLoan", 'TransactionStatus'>
  readonly createdAt: Prisma.FieldRef<"LongTermLoan", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"LongTermLoan", 'DateTime'>
}
    

// Custom InputTypes
/**
 * LongTermLoan findUnique
 */
export type LongTermLoanFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoan to fetch.
   */
  where: Prisma.LongTermLoanWhereUniqueInput
}

/**
 * LongTermLoan findUniqueOrThrow
 */
export type LongTermLoanFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoan to fetch.
   */
  where: Prisma.LongTermLoanWhereUniqueInput
}

/**
 * LongTermLoan findFirst
 */
export type LongTermLoanFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoan to fetch.
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoans to fetch.
   */
  orderBy?: Prisma.LongTermLoanOrderByWithRelationInput | Prisma.LongTermLoanOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for LongTermLoans.
   */
  cursor?: Prisma.LongTermLoanWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoans from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoans.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of LongTermLoans.
   */
  distinct?: Prisma.LongTermLoanScalarFieldEnum | Prisma.LongTermLoanScalarFieldEnum[]
}

/**
 * LongTermLoan findFirstOrThrow
 */
export type LongTermLoanFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoan to fetch.
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoans to fetch.
   */
  orderBy?: Prisma.LongTermLoanOrderByWithRelationInput | Prisma.LongTermLoanOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for LongTermLoans.
   */
  cursor?: Prisma.LongTermLoanWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoans from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoans.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of LongTermLoans.
   */
  distinct?: Prisma.LongTermLoanScalarFieldEnum | Prisma.LongTermLoanScalarFieldEnum[]
}

/**
 * LongTermLoan findMany
 */
export type LongTermLoanFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoans to fetch.
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoans to fetch.
   */
  orderBy?: Prisma.LongTermLoanOrderByWithRelationInput | Prisma.LongTermLoanOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing LongTermLoans.
   */
  cursor?: Prisma.LongTermLoanWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoans from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoans.
   */
  skip?: number
  distinct?: Prisma.LongTermLoanScalarFieldEnum | Prisma.LongTermLoanScalarFieldEnum[]
}

/**
 * LongTermLoan create
 */
export type LongTermLoanCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * The data needed to create a LongTermLoan.
   */
  data: Prisma.XOR<Prisma.LongTermLoanCreateInput, Prisma.LongTermLoanUncheckedCreateInput>
}

/**
 * LongTermLoan createMany
 */
export type LongTermLoanCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many LongTermLoans.
   */
  data: Prisma.LongTermLoanCreateManyInput | Prisma.LongTermLoanCreateManyInput[]
}

/**
 * LongTermLoan update
 */
export type LongTermLoanUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * The data needed to update a LongTermLoan.
   */
  data: Prisma.XOR<Prisma.LongTermLoanUpdateInput, Prisma.LongTermLoanUncheckedUpdateInput>
  /**
   * Choose, which LongTermLoan to update.
   */
  where: Prisma.LongTermLoanWhereUniqueInput
}

/**
 * LongTermLoan updateMany
 */
export type LongTermLoanUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update LongTermLoans.
   */
  data: Prisma.XOR<Prisma.LongTermLoanUpdateManyMutationInput, Prisma.LongTermLoanUncheckedUpdateManyInput>
  /**
   * Filter which LongTermLoans to update
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * Limit how many LongTermLoans to update.
   */
  limit?: number
}

/**
 * LongTermLoan upsert
 */
export type LongTermLoanUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * The filter to search for the LongTermLoan to update in case it exists.
   */
  where: Prisma.LongTermLoanWhereUniqueInput
  /**
   * In case the LongTermLoan found by the `where` argument doesn't exist, create a new LongTermLoan with this data.
   */
  create: Prisma.XOR<Prisma.LongTermLoanCreateInput, Prisma.LongTermLoanUncheckedCreateInput>
  /**
   * In case the LongTermLoan was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.LongTermLoanUpdateInput, Prisma.LongTermLoanUncheckedUpdateInput>
}

/**
 * LongTermLoan delete
 */
export type LongTermLoanDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
  /**
   * Filter which LongTermLoan to delete.
   */
  where: Prisma.LongTermLoanWhereUniqueInput
}

/**
 * LongTermLoan deleteMany
 */
export type LongTermLoanDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which LongTermLoans to delete
   */
  where?: Prisma.LongTermLoanWhereInput
  /**
   * Limit how many LongTermLoans to delete.
   */
  limit?: number
}

/**
 * LongTermLoan findRaw
 */
export type LongTermLoanFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * LongTermLoan aggregateRaw
 */
export type LongTermLoanAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * LongTermLoan.penalty
 */
export type LongTermLoan$penaltyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Penalty
   */
  select?: Prisma.PenaltySelect<ExtArgs> | null
  /**
   * Omit specific fields from the Penalty
   */
  omit?: Prisma.PenaltyOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.PenaltyInclude<ExtArgs> | null
  where?: Prisma.PenaltyWhereInput
}

/**
 * LongTermLoan.deductions
 */
export type LongTermLoan$deductionsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Deduction
   */
  select?: Prisma.DeductionSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Deduction
   */
  omit?: Prisma.DeductionOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.DeductionInclude<ExtArgs> | null
  where?: Prisma.DeductionWhereInput
  orderBy?: Prisma.DeductionOrderByWithRelationInput | Prisma.DeductionOrderByWithRelationInput[]
  cursor?: Prisma.DeductionWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.DeductionScalarFieldEnum | Prisma.DeductionScalarFieldEnum[]
}

/**
 * LongTermLoan.longTermLoanRepayments
 */
export type LongTermLoan$longTermLoanRepaymentsArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  where?: Prisma.LongTermLoanRepaymentWhereInput
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithRelationInput | Prisma.LongTermLoanRepaymentOrderByWithRelationInput[]
  cursor?: Prisma.LongTermLoanRepaymentWhereUniqueInput
  take?: number
  skip?: number
  distinct?: Prisma.LongTermLoanRepaymentScalarFieldEnum | Prisma.LongTermLoanRepaymentScalarFieldEnum[]
}

/**
 * LongTermLoan without action
 */
export type LongTermLoanDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoan
   */
  select?: Prisma.LongTermLoanSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoan
   */
  omit?: Prisma.LongTermLoanOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanInclude<ExtArgs> | null
}
