// @prisma/client/models/LongTermLoanRepayment.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `LongTermLoanRepayment` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model LongTermLoanRepayment
 * 
 */
export type LongTermLoanRepaymentModel = runtime.Types.Result.DefaultSelection<Prisma.$LongTermLoanRepaymentPayload>

export type AggregateLongTermLoanRepayment = {
  _count: LongTermLoanRepaymentCountAggregateOutputType | null
  _min: LongTermLoanRepaymentMinAggregateOutputType | null
  _max: LongTermLoanRepaymentMaxAggregateOutputType | null
}

export type LongTermLoanRepaymentMinAggregateOutputType = {
  id: string | null
  longTermLoanId: string | null
  dueDate: Date | null
  paymentDate: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type LongTermLoanRepaymentMaxAggregateOutputType = {
  id: string | null
  longTermLoanId: string | null
  dueDate: Date | null
  paymentDate: Date | null
  status: $Enums.TransactionStatus | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type LongTermLoanRepaymentCountAggregateOutputType = {
  id: number
  longTermLoanId: number
  dueDate: number
  paymentDate: number
  status: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type LongTermLoanRepaymentMinAggregateInputType = {
  id?: true
  longTermLoanId?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type LongTermLoanRepaymentMaxAggregateInputType = {
  id?: true
  longTermLoanId?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
}

export type LongTermLoanRepaymentCountAggregateInputType = {
  id?: true
  longTermLoanId?: true
  dueDate?: true
  paymentDate?: true
  status?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type LongTermLoanRepaymentAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which LongTermLoanRepayment to aggregate.
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoanRepayments to fetch.
   */
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithRelationInput | Prisma.LongTermLoanRepaymentOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.LongTermLoanRepaymentWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoanRepayments from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoanRepayments.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned LongTermLoanRepayments
  **/
  _count?: true | LongTermLoanRepaymentCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: LongTermLoanRepaymentMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: LongTermLoanRepaymentMaxAggregateInputType
}

export type GetLongTermLoanRepaymentAggregateType<T extends LongTermLoanRepaymentAggregateArgs> = {
      [P in keyof T & keyof AggregateLongTermLoanRepayment]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateLongTermLoanRepayment[P]>
    : Prisma.GetScalarType<T[P], AggregateLongTermLoanRepayment[P]>
}




export type LongTermLoanRepaymentGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.LongTermLoanRepaymentWhereInput
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithAggregationInput | Prisma.LongTermLoanRepaymentOrderByWithAggregationInput[]
  by: Prisma.LongTermLoanRepaymentScalarFieldEnum[] | Prisma.LongTermLoanRepaymentScalarFieldEnum
  having?: Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: LongTermLoanRepaymentCountAggregateInputType | true
  _min?: LongTermLoanRepaymentMinAggregateInputType
  _max?: LongTermLoanRepaymentMaxAggregateInputType
}

export type LongTermLoanRepaymentGroupByOutputType = {
  id: string
  longTermLoanId: string
  dueDate: Date
  paymentDate: Date | null
  status: $Enums.TransactionStatus
  createdAt: Date
  updatedAt: Date
  _count: LongTermLoanRepaymentCountAggregateOutputType | null
  _min: LongTermLoanRepaymentMinAggregateOutputType | null
  _max: LongTermLoanRepaymentMaxAggregateOutputType | null
}

type GetLongTermLoanRepaymentGroupByPayload<T extends LongTermLoanRepaymentGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<LongTermLoanRepaymentGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof LongTermLoanRepaymentGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], LongTermLoanRepaymentGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], LongTermLoanRepaymentGroupByOutputType[P]>
      }
    >
  >



export type LongTermLoanRepaymentWhereInput = {
  AND?: Prisma.LongTermLoanRepaymentWhereInput | Prisma.LongTermLoanRepaymentWhereInput[]
  OR?: Prisma.LongTermLoanRepaymentWhereInput[]
  NOT?: Prisma.LongTermLoanRepaymentWhereInput | Prisma.LongTermLoanRepaymentWhereInput[]
  id?: Prisma.StringFilter<"LongTermLoanRepayment"> | string
  longTermLoanId?: Prisma.StringFilter<"LongTermLoanRepayment"> | string
  dueDate?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoanRepayment"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoanRepayment"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  longTermLoan?: Prisma.XOR<Prisma.LongTermLoanScalarRelationFilter, Prisma.LongTermLoanWhereInput>
}

export type LongTermLoanRepaymentOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  longTermLoan?: Prisma.LongTermLoanOrderByWithRelationInput
}

export type LongTermLoanRepaymentWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.LongTermLoanRepaymentWhereInput | Prisma.LongTermLoanRepaymentWhereInput[]
  OR?: Prisma.LongTermLoanRepaymentWhereInput[]
  NOT?: Prisma.LongTermLoanRepaymentWhereInput | Prisma.LongTermLoanRepaymentWhereInput[]
  longTermLoanId?: Prisma.StringFilter<"LongTermLoanRepayment"> | string
  dueDate?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoanRepayment"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoanRepayment"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  longTermLoan?: Prisma.XOR<Prisma.LongTermLoanScalarRelationFilter, Prisma.LongTermLoanWhereInput>
}, "id">

export type LongTermLoanRepaymentOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.LongTermLoanRepaymentCountOrderByAggregateInput
  _max?: Prisma.LongTermLoanRepaymentMaxOrderByAggregateInput
  _min?: Prisma.LongTermLoanRepaymentMinOrderByAggregateInput
}

export type LongTermLoanRepaymentScalarWhereWithAggregatesInput = {
  AND?: Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput | Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput[]
  OR?: Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput[]
  NOT?: Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput | Prisma.LongTermLoanRepaymentScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"LongTermLoanRepayment"> | string
  longTermLoanId?: Prisma.StringWithAggregatesFilter<"LongTermLoanRepayment"> | string
  dueDate?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoanRepayment"> | Date | string
  paymentDate?: Prisma.DateTimeNullableWithAggregatesFilter<"LongTermLoanRepayment"> | Date | string | null
  status?: Prisma.EnumTransactionStatusWithAggregatesFilter<"LongTermLoanRepayment"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoanRepayment"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"LongTermLoanRepayment"> | Date | string
}

export type LongTermLoanRepaymentCreateInput = {
  id?: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
  longTermLoan: Prisma.LongTermLoanCreateNestedOneWithoutLongTermLoanRepaymentsInput
}

export type LongTermLoanRepaymentUncheckedCreateInput = {
  id?: string
  longTermLoanId: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanRepaymentUpdateInput = {
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  longTermLoan?: Prisma.LongTermLoanUpdateOneRequiredWithoutLongTermLoanRepaymentsNestedInput
}

export type LongTermLoanRepaymentUncheckedUpdateInput = {
  longTermLoanId?: Prisma.StringFieldUpdateOperationsInput | string
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanRepaymentCreateManyInput = {
  id?: string
  longTermLoanId: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanRepaymentUpdateManyMutationInput = {
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanRepaymentUncheckedUpdateManyInput = {
  longTermLoanId?: Prisma.StringFieldUpdateOperationsInput | string
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanRepaymentListRelationFilter = {
  every?: Prisma.LongTermLoanRepaymentWhereInput
  some?: Prisma.LongTermLoanRepaymentWhereInput
  none?: Prisma.LongTermLoanRepaymentWhereInput
}

export type LongTermLoanRepaymentOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type LongTermLoanRepaymentCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanRepaymentMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanRepaymentMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  longTermLoanId?: Prisma.SortOrder
  dueDate?: Prisma.SortOrder
  paymentDate?: Prisma.SortOrder
  status?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type LongTermLoanRepaymentCreateNestedManyWithoutLongTermLoanInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput> | Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput[] | Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput[]
  createMany?: Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInputEnvelope
  connect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
}

export type LongTermLoanRepaymentUncheckedCreateNestedManyWithoutLongTermLoanInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput> | Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput[] | Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput[]
  createMany?: Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInputEnvelope
  connect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
}

export type LongTermLoanRepaymentUpdateManyWithoutLongTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput> | Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput[] | Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput[]
  upsert?: Prisma.LongTermLoanRepaymentUpsertWithWhereUniqueWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpsertWithWhereUniqueWithoutLongTermLoanInput[]
  createMany?: Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInputEnvelope
  set?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  delete?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  connect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  update?: Prisma.LongTermLoanRepaymentUpdateWithWhereUniqueWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpdateWithWhereUniqueWithoutLongTermLoanInput[]
  updateMany?: Prisma.LongTermLoanRepaymentUpdateManyWithWhereWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpdateManyWithWhereWithoutLongTermLoanInput[]
  deleteMany?: Prisma.LongTermLoanRepaymentScalarWhereInput | Prisma.LongTermLoanRepaymentScalarWhereInput[]
}

export type LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanNestedInput = {
  create?: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput> | Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput[] | Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput[]
  connectOrCreate?: Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput[]
  upsert?: Prisma.LongTermLoanRepaymentUpsertWithWhereUniqueWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpsertWithWhereUniqueWithoutLongTermLoanInput[]
  createMany?: Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInputEnvelope
  set?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  disconnect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  delete?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  connect?: Prisma.LongTermLoanRepaymentWhereUniqueInput | Prisma.LongTermLoanRepaymentWhereUniqueInput[]
  update?: Prisma.LongTermLoanRepaymentUpdateWithWhereUniqueWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpdateWithWhereUniqueWithoutLongTermLoanInput[]
  updateMany?: Prisma.LongTermLoanRepaymentUpdateManyWithWhereWithoutLongTermLoanInput | Prisma.LongTermLoanRepaymentUpdateManyWithWhereWithoutLongTermLoanInput[]
  deleteMany?: Prisma.LongTermLoanRepaymentScalarWhereInput | Prisma.LongTermLoanRepaymentScalarWhereInput[]
}

export type LongTermLoanRepaymentCreateWithoutLongTermLoanInput = {
  id?: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput = {
  id?: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanRepaymentCreateOrConnectWithoutLongTermLoanInput = {
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
  create: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput>
}

export type LongTermLoanRepaymentCreateManyLongTermLoanInputEnvelope = {
  data: Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInput | Prisma.LongTermLoanRepaymentCreateManyLongTermLoanInput[]
}

export type LongTermLoanRepaymentUpsertWithWhereUniqueWithoutLongTermLoanInput = {
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
  update: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedUpdateWithoutLongTermLoanInput>
  create: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedCreateWithoutLongTermLoanInput>
}

export type LongTermLoanRepaymentUpdateWithWhereUniqueWithoutLongTermLoanInput = {
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
  data: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateWithoutLongTermLoanInput, Prisma.LongTermLoanRepaymentUncheckedUpdateWithoutLongTermLoanInput>
}

export type LongTermLoanRepaymentUpdateManyWithWhereWithoutLongTermLoanInput = {
  where: Prisma.LongTermLoanRepaymentScalarWhereInput
  data: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateManyMutationInput, Prisma.LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanInput>
}

export type LongTermLoanRepaymentScalarWhereInput = {
  AND?: Prisma.LongTermLoanRepaymentScalarWhereInput | Prisma.LongTermLoanRepaymentScalarWhereInput[]
  OR?: Prisma.LongTermLoanRepaymentScalarWhereInput[]
  NOT?: Prisma.LongTermLoanRepaymentScalarWhereInput | Prisma.LongTermLoanRepaymentScalarWhereInput[]
  id?: Prisma.StringFilter<"LongTermLoanRepayment"> | string
  longTermLoanId?: Prisma.StringFilter<"LongTermLoanRepayment"> | string
  dueDate?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  paymentDate?: Prisma.DateTimeNullableFilter<"LongTermLoanRepayment"> | Date | string | null
  status?: Prisma.EnumTransactionStatusFilter<"LongTermLoanRepayment"> | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"LongTermLoanRepayment"> | Date | string
}

export type LongTermLoanRepaymentCreateManyLongTermLoanInput = {
  id?: string
  dueDate?: Date | string
  paymentDate?: Date | string | null
  status?: $Enums.TransactionStatus
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type LongTermLoanRepaymentUpdateWithoutLongTermLoanInput = {
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanRepaymentUncheckedUpdateWithoutLongTermLoanInput = {
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type LongTermLoanRepaymentUncheckedUpdateManyWithoutLongTermLoanInput = {
  dueDate?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  paymentDate?: Prisma.NullableDateTimeFieldUpdateOperationsInput | Date | string | null
  status?: Prisma.EnumTransactionStatusFieldUpdateOperationsInput | $Enums.TransactionStatus
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}



export type LongTermLoanRepaymentSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  longTermLoanId?: boolean
  dueDate?: boolean
  paymentDate?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  longTermLoan?: boolean | Prisma.LongTermLoanDefaultArgs<ExtArgs>
}, ExtArgs["result"]["longTermLoanRepayment"]>



export type LongTermLoanRepaymentSelectScalar = {
  id?: boolean
  longTermLoanId?: boolean
  dueDate?: boolean
  paymentDate?: boolean
  status?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type LongTermLoanRepaymentOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "longTermLoanId" | "dueDate" | "paymentDate" | "status" | "createdAt" | "updatedAt", ExtArgs["result"]["longTermLoanRepayment"]>
export type LongTermLoanRepaymentInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  longTermLoan?: boolean | Prisma.LongTermLoanDefaultArgs<ExtArgs>
}

export type $LongTermLoanRepaymentPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "LongTermLoanRepayment"
  objects: {
    longTermLoan: Prisma.$LongTermLoanPayload<ExtArgs>
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    longTermLoanId: string
    dueDate: Date
    paymentDate: Date | null
    status: $Enums.TransactionStatus
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["longTermLoanRepayment"]>
  composites: {}
}

export type LongTermLoanRepaymentGetPayload<S extends boolean | null | undefined | LongTermLoanRepaymentDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload, S>

export type LongTermLoanRepaymentCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<LongTermLoanRepaymentFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: LongTermLoanRepaymentCountAggregateInputType | true
  }

export interface LongTermLoanRepaymentDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['LongTermLoanRepayment'], meta: { name: 'LongTermLoanRepayment' } }
  /**
   * Find zero or one LongTermLoanRepayment that matches the filter.
   * @param {LongTermLoanRepaymentFindUniqueArgs} args - Arguments to find a LongTermLoanRepayment
   * @example
   * // Get one LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends LongTermLoanRepaymentFindUniqueArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentFindUniqueArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one LongTermLoanRepayment that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {LongTermLoanRepaymentFindUniqueOrThrowArgs} args - Arguments to find a LongTermLoanRepayment
   * @example
   * // Get one LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends LongTermLoanRepaymentFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first LongTermLoanRepayment that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentFindFirstArgs} args - Arguments to find a LongTermLoanRepayment
   * @example
   * // Get one LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends LongTermLoanRepaymentFindFirstArgs>(args?: Prisma.SelectSubset<T, LongTermLoanRepaymentFindFirstArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first LongTermLoanRepayment that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentFindFirstOrThrowArgs} args - Arguments to find a LongTermLoanRepayment
   * @example
   * // Get one LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends LongTermLoanRepaymentFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, LongTermLoanRepaymentFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more LongTermLoanRepayments that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all LongTermLoanRepayments
   * const longTermLoanRepayments = await prisma.longTermLoanRepayment.findMany()
   * 
   * // Get first 10 LongTermLoanRepayments
   * const longTermLoanRepayments = await prisma.longTermLoanRepayment.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const longTermLoanRepaymentWithIdOnly = await prisma.longTermLoanRepayment.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends LongTermLoanRepaymentFindManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanRepaymentFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a LongTermLoanRepayment.
   * @param {LongTermLoanRepaymentCreateArgs} args - Arguments to create a LongTermLoanRepayment.
   * @example
   * // Create one LongTermLoanRepayment
   * const LongTermLoanRepayment = await prisma.longTermLoanRepayment.create({
   *   data: {
   *     // ... data to create a LongTermLoanRepayment
   *   }
   * })
   * 
   */
  create<T extends LongTermLoanRepaymentCreateArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentCreateArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many LongTermLoanRepayments.
   * @param {LongTermLoanRepaymentCreateManyArgs} args - Arguments to create many LongTermLoanRepayments.
   * @example
   * // Create many LongTermLoanRepayments
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends LongTermLoanRepaymentCreateManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanRepaymentCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a LongTermLoanRepayment.
   * @param {LongTermLoanRepaymentDeleteArgs} args - Arguments to delete one LongTermLoanRepayment.
   * @example
   * // Delete one LongTermLoanRepayment
   * const LongTermLoanRepayment = await prisma.longTermLoanRepayment.delete({
   *   where: {
   *     // ... filter to delete one LongTermLoanRepayment
   *   }
   * })
   * 
   */
  delete<T extends LongTermLoanRepaymentDeleteArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentDeleteArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one LongTermLoanRepayment.
   * @param {LongTermLoanRepaymentUpdateArgs} args - Arguments to update one LongTermLoanRepayment.
   * @example
   * // Update one LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends LongTermLoanRepaymentUpdateArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentUpdateArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more LongTermLoanRepayments.
   * @param {LongTermLoanRepaymentDeleteManyArgs} args - Arguments to filter LongTermLoanRepayments to delete.
   * @example
   * // Delete a few LongTermLoanRepayments
   * const { count } = await prisma.longTermLoanRepayment.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends LongTermLoanRepaymentDeleteManyArgs>(args?: Prisma.SelectSubset<T, LongTermLoanRepaymentDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more LongTermLoanRepayments.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many LongTermLoanRepayments
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends LongTermLoanRepaymentUpdateManyArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one LongTermLoanRepayment.
   * @param {LongTermLoanRepaymentUpsertArgs} args - Arguments to update or create a LongTermLoanRepayment.
   * @example
   * // Update or create a LongTermLoanRepayment
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.upsert({
   *   create: {
   *     // ... data to create a LongTermLoanRepayment
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the LongTermLoanRepayment we want to update
   *   }
   * })
   */
  upsert<T extends LongTermLoanRepaymentUpsertArgs>(args: Prisma.SelectSubset<T, LongTermLoanRepaymentUpsertArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanRepaymentClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanRepaymentPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more LongTermLoanRepayments that matches the filter.
   * @param {LongTermLoanRepaymentFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.LongTermLoanRepaymentFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a LongTermLoanRepayment.
   * @param {LongTermLoanRepaymentAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const longTermLoanRepayment = await prisma.longTermLoanRepayment.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.LongTermLoanRepaymentAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of LongTermLoanRepayments.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentCountArgs} args - Arguments to filter LongTermLoanRepayments to count.
   * @example
   * // Count the number of LongTermLoanRepayments
   * const count = await prisma.longTermLoanRepayment.count({
   *   where: {
   *     // ... the filter for the LongTermLoanRepayments we want to count
   *   }
   * })
  **/
  count<T extends LongTermLoanRepaymentCountArgs>(
    args?: Prisma.Subset<T, LongTermLoanRepaymentCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], LongTermLoanRepaymentCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a LongTermLoanRepayment.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends LongTermLoanRepaymentAggregateArgs>(args: Prisma.Subset<T, LongTermLoanRepaymentAggregateArgs>): Prisma.PrismaPromise<GetLongTermLoanRepaymentAggregateType<T>>

  /**
   * Group by LongTermLoanRepayment.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {LongTermLoanRepaymentGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends LongTermLoanRepaymentGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: LongTermLoanRepaymentGroupByArgs['orderBy'] }
      : { orderBy?: LongTermLoanRepaymentGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, LongTermLoanRepaymentGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetLongTermLoanRepaymentGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the LongTermLoanRepayment model
 */
readonly fields: LongTermLoanRepaymentFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for LongTermLoanRepayment.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__LongTermLoanRepaymentClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  longTermLoan<T extends Prisma.LongTermLoanDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.LongTermLoanDefaultArgs<ExtArgs>>): Prisma.Prisma__LongTermLoanClient<runtime.Types.Result.GetResult<Prisma.$LongTermLoanPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the LongTermLoanRepayment model
 */
export interface LongTermLoanRepaymentFieldRefs {
  readonly id: Prisma.FieldRef<"LongTermLoanRepayment", 'String'>
  readonly longTermLoanId: Prisma.FieldRef<"LongTermLoanRepayment", 'String'>
  readonly dueDate: Prisma.FieldRef<"LongTermLoanRepayment", 'DateTime'>
  readonly paymentDate: Prisma.FieldRef<"LongTermLoanRepayment", 'DateTime'>
  readonly status: Prisma.FieldRef<"LongTermLoanRepayment", 'TransactionStatus'>
  readonly createdAt: Prisma.FieldRef<"LongTermLoanRepayment", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"LongTermLoanRepayment", 'DateTime'>
}
    

// Custom InputTypes
/**
 * LongTermLoanRepayment findUnique
 */
export type LongTermLoanRepaymentFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoanRepayment to fetch.
   */
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
}

/**
 * LongTermLoanRepayment findUniqueOrThrow
 */
export type LongTermLoanRepaymentFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoanRepayment to fetch.
   */
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
}

/**
 * LongTermLoanRepayment findFirst
 */
export type LongTermLoanRepaymentFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoanRepayment to fetch.
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoanRepayments to fetch.
   */
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithRelationInput | Prisma.LongTermLoanRepaymentOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for LongTermLoanRepayments.
   */
  cursor?: Prisma.LongTermLoanRepaymentWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoanRepayments from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoanRepayments.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of LongTermLoanRepayments.
   */
  distinct?: Prisma.LongTermLoanRepaymentScalarFieldEnum | Prisma.LongTermLoanRepaymentScalarFieldEnum[]
}

/**
 * LongTermLoanRepayment findFirstOrThrow
 */
export type LongTermLoanRepaymentFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoanRepayment to fetch.
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoanRepayments to fetch.
   */
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithRelationInput | Prisma.LongTermLoanRepaymentOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for LongTermLoanRepayments.
   */
  cursor?: Prisma.LongTermLoanRepaymentWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoanRepayments from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoanRepayments.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of LongTermLoanRepayments.
   */
  distinct?: Prisma.LongTermLoanRepaymentScalarFieldEnum | Prisma.LongTermLoanRepaymentScalarFieldEnum[]
}

/**
 * LongTermLoanRepayment findMany
 */
export type LongTermLoanRepaymentFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter, which LongTermLoanRepayments to fetch.
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of LongTermLoanRepayments to fetch.
   */
  orderBy?: Prisma.LongTermLoanRepaymentOrderByWithRelationInput | Prisma.LongTermLoanRepaymentOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing LongTermLoanRepayments.
   */
  cursor?: Prisma.LongTermLoanRepaymentWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` LongTermLoanRepayments from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` LongTermLoanRepayments.
   */
  skip?: number
  distinct?: Prisma.LongTermLoanRepaymentScalarFieldEnum | Prisma.LongTermLoanRepaymentScalarFieldEnum[]
}

/**
 * LongTermLoanRepayment create
 */
export type LongTermLoanRepaymentCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * The data needed to create a LongTermLoanRepayment.
   */
  data: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateInput, Prisma.LongTermLoanRepaymentUncheckedCreateInput>
}

/**
 * LongTermLoanRepayment createMany
 */
export type LongTermLoanRepaymentCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many LongTermLoanRepayments.
   */
  data: Prisma.LongTermLoanRepaymentCreateManyInput | Prisma.LongTermLoanRepaymentCreateManyInput[]
}

/**
 * LongTermLoanRepayment update
 */
export type LongTermLoanRepaymentUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * The data needed to update a LongTermLoanRepayment.
   */
  data: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateInput, Prisma.LongTermLoanRepaymentUncheckedUpdateInput>
  /**
   * Choose, which LongTermLoanRepayment to update.
   */
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
}

/**
 * LongTermLoanRepayment updateMany
 */
export type LongTermLoanRepaymentUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update LongTermLoanRepayments.
   */
  data: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateManyMutationInput, Prisma.LongTermLoanRepaymentUncheckedUpdateManyInput>
  /**
   * Filter which LongTermLoanRepayments to update
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * Limit how many LongTermLoanRepayments to update.
   */
  limit?: number
}

/**
 * LongTermLoanRepayment upsert
 */
export type LongTermLoanRepaymentUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * The filter to search for the LongTermLoanRepayment to update in case it exists.
   */
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
  /**
   * In case the LongTermLoanRepayment found by the `where` argument doesn't exist, create a new LongTermLoanRepayment with this data.
   */
  create: Prisma.XOR<Prisma.LongTermLoanRepaymentCreateInput, Prisma.LongTermLoanRepaymentUncheckedCreateInput>
  /**
   * In case the LongTermLoanRepayment was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.LongTermLoanRepaymentUpdateInput, Prisma.LongTermLoanRepaymentUncheckedUpdateInput>
}

/**
 * LongTermLoanRepayment delete
 */
export type LongTermLoanRepaymentDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
  /**
   * Filter which LongTermLoanRepayment to delete.
   */
  where: Prisma.LongTermLoanRepaymentWhereUniqueInput
}

/**
 * LongTermLoanRepayment deleteMany
 */
export type LongTermLoanRepaymentDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which LongTermLoanRepayments to delete
   */
  where?: Prisma.LongTermLoanRepaymentWhereInput
  /**
   * Limit how many LongTermLoanRepayments to delete.
   */
  limit?: number
}

/**
 * LongTermLoanRepayment findRaw
 */
export type LongTermLoanRepaymentFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * LongTermLoanRepayment aggregateRaw
 */
export type LongTermLoanRepaymentAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * LongTermLoanRepayment without action
 */
export type LongTermLoanRepaymentDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the LongTermLoanRepayment
   */
  select?: Prisma.LongTermLoanRepaymentSelect<ExtArgs> | null
  /**
   * Omit specific fields from the LongTermLoanRepayment
   */
  omit?: Prisma.LongTermLoanRepaymentOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.LongTermLoanRepaymentInclude<ExtArgs> | null
}
