// @prisma/client/models/Projects.ts

/* !!! This is code generated by Prisma. Do not edit directly. !!! */
/* eslint-disable */
// biome-ignore-all lint: generated file
// @ts-nocheck 
/*
 * This file exports the `Projects` model and its related types.
 *
 * 🟢 You can import this file directly.
 */
import type * as runtime from "@prisma/client/runtime/client"
import type * as $Enums from "../enums"
import type * as Prisma from "../internal/prismaNamespace"

/**
 * Model Projects
 * 
 */
export type ProjectsModel = runtime.Types.Result.DefaultSelection<Prisma.$ProjectsPayload>

export type AggregateProjects = {
  _count: ProjectsCountAggregateOutputType | null
  _avg: ProjectsAvgAggregateOutputType | null
  _sum: ProjectsSumAggregateOutputType | null
  _min: ProjectsMinAggregateOutputType | null
  _max: ProjectsMaxAggregateOutputType | null
}

export type ProjectsAvgAggregateOutputType = {
  estimatedCost: number | null
  actualCost: number | null
  sharesUsed: number | null
}

export type ProjectsSumAggregateOutputType = {
  estimatedCost: number | null
  actualCost: number | null
  sharesUsed: number | null
}

export type ProjectsMinAggregateOutputType = {
  id: string | null
  chamaId: string | null
  projectName: string | null
  estimatedCost: number | null
  actualCost: number | null
  sharesUsed: number | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type ProjectsMaxAggregateOutputType = {
  id: string | null
  chamaId: string | null
  projectName: string | null
  estimatedCost: number | null
  actualCost: number | null
  sharesUsed: number | null
  createdAt: Date | null
  updatedAt: Date | null
}

export type ProjectsCountAggregateOutputType = {
  id: number
  chamaId: number
  projectName: number
  estimatedCost: number
  actualCost: number
  sharesUsed: number
  createdAt: number
  updatedAt: number
  _all: number
}


export type ProjectsAvgAggregateInputType = {
  estimatedCost?: true
  actualCost?: true
  sharesUsed?: true
}

export type ProjectsSumAggregateInputType = {
  estimatedCost?: true
  actualCost?: true
  sharesUsed?: true
}

export type ProjectsMinAggregateInputType = {
  id?: true
  chamaId?: true
  projectName?: true
  estimatedCost?: true
  actualCost?: true
  sharesUsed?: true
  createdAt?: true
  updatedAt?: true
}

export type ProjectsMaxAggregateInputType = {
  id?: true
  chamaId?: true
  projectName?: true
  estimatedCost?: true
  actualCost?: true
  sharesUsed?: true
  createdAt?: true
  updatedAt?: true
}

export type ProjectsCountAggregateInputType = {
  id?: true
  chamaId?: true
  projectName?: true
  estimatedCost?: true
  actualCost?: true
  sharesUsed?: true
  createdAt?: true
  updatedAt?: true
  _all?: true
}

export type ProjectsAggregateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Projects to aggregate.
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Projects to fetch.
   */
  orderBy?: Prisma.ProjectsOrderByWithRelationInput | Prisma.ProjectsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the start position
   */
  cursor?: Prisma.ProjectsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Projects from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Projects.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Count returned Projects
  **/
  _count?: true | ProjectsCountAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to average
  **/
  _avg?: ProjectsAvgAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to sum
  **/
  _sum?: ProjectsSumAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the minimum value
  **/
  _min?: ProjectsMinAggregateInputType
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/aggregations Aggregation Docs}
   * 
   * Select which fields to find the maximum value
  **/
  _max?: ProjectsMaxAggregateInputType
}

export type GetProjectsAggregateType<T extends ProjectsAggregateArgs> = {
      [P in keyof T & keyof AggregateProjects]: P extends '_count' | 'count'
    ? T[P] extends true
      ? number
      : Prisma.GetScalarType<T[P], AggregateProjects[P]>
    : Prisma.GetScalarType<T[P], AggregateProjects[P]>
}




export type ProjectsGroupByArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  where?: Prisma.ProjectsWhereInput
  orderBy?: Prisma.ProjectsOrderByWithAggregationInput | Prisma.ProjectsOrderByWithAggregationInput[]
  by: Prisma.ProjectsScalarFieldEnum[] | Prisma.ProjectsScalarFieldEnum
  having?: Prisma.ProjectsScalarWhereWithAggregatesInput
  take?: number
  skip?: number
  _count?: ProjectsCountAggregateInputType | true
  _avg?: ProjectsAvgAggregateInputType
  _sum?: ProjectsSumAggregateInputType
  _min?: ProjectsMinAggregateInputType
  _max?: ProjectsMaxAggregateInputType
}

export type ProjectsGroupByOutputType = {
  id: string
  chamaId: string
  projectName: string
  estimatedCost: number
  actualCost: number
  sharesUsed: number
  createdAt: Date
  updatedAt: Date
  _count: ProjectsCountAggregateOutputType | null
  _avg: ProjectsAvgAggregateOutputType | null
  _sum: ProjectsSumAggregateOutputType | null
  _min: ProjectsMinAggregateOutputType | null
  _max: ProjectsMaxAggregateOutputType | null
}

type GetProjectsGroupByPayload<T extends ProjectsGroupByArgs> = Prisma.PrismaPromise<
  Array<
    Prisma.PickEnumerable<ProjectsGroupByOutputType, T['by']> &
      {
        [P in ((keyof T) & (keyof ProjectsGroupByOutputType))]: P extends '_count'
          ? T[P] extends boolean
            ? number
            : Prisma.GetScalarType<T[P], ProjectsGroupByOutputType[P]>
          : Prisma.GetScalarType<T[P], ProjectsGroupByOutputType[P]>
      }
    >
  >



export type ProjectsWhereInput = {
  AND?: Prisma.ProjectsWhereInput | Prisma.ProjectsWhereInput[]
  OR?: Prisma.ProjectsWhereInput[]
  NOT?: Prisma.ProjectsWhereInput | Prisma.ProjectsWhereInput[]
  id?: Prisma.StringFilter<"Projects"> | string
  chamaId?: Prisma.StringFilter<"Projects"> | string
  projectName?: Prisma.StringFilter<"Projects"> | string
  estimatedCost?: Prisma.FloatFilter<"Projects"> | number
  actualCost?: Prisma.FloatFilter<"Projects"> | number
  sharesUsed?: Prisma.FloatFilter<"Projects"> | number
  createdAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
  chama?: Prisma.XOR<Prisma.ChamaScalarRelationFilter, Prisma.ChamaWhereInput>
}

export type ProjectsOrderByWithRelationInput = {
  id?: Prisma.SortOrder
  chamaId?: Prisma.SortOrder
  projectName?: Prisma.SortOrder
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  chama?: Prisma.ChamaOrderByWithRelationInput
}

export type ProjectsWhereUniqueInput = Prisma.AtLeast<{
  id?: string
  AND?: Prisma.ProjectsWhereInput | Prisma.ProjectsWhereInput[]
  OR?: Prisma.ProjectsWhereInput[]
  NOT?: Prisma.ProjectsWhereInput | Prisma.ProjectsWhereInput[]
  chamaId?: Prisma.StringFilter<"Projects"> | string
  projectName?: Prisma.StringFilter<"Projects"> | string
  estimatedCost?: Prisma.FloatFilter<"Projects"> | number
  actualCost?: Prisma.FloatFilter<"Projects"> | number
  sharesUsed?: Prisma.FloatFilter<"Projects"> | number
  createdAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
  chama?: Prisma.XOR<Prisma.ChamaScalarRelationFilter, Prisma.ChamaWhereInput>
}, "id">

export type ProjectsOrderByWithAggregationInput = {
  id?: Prisma.SortOrder
  chamaId?: Prisma.SortOrder
  projectName?: Prisma.SortOrder
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
  _count?: Prisma.ProjectsCountOrderByAggregateInput
  _avg?: Prisma.ProjectsAvgOrderByAggregateInput
  _max?: Prisma.ProjectsMaxOrderByAggregateInput
  _min?: Prisma.ProjectsMinOrderByAggregateInput
  _sum?: Prisma.ProjectsSumOrderByAggregateInput
}

export type ProjectsScalarWhereWithAggregatesInput = {
  AND?: Prisma.ProjectsScalarWhereWithAggregatesInput | Prisma.ProjectsScalarWhereWithAggregatesInput[]
  OR?: Prisma.ProjectsScalarWhereWithAggregatesInput[]
  NOT?: Prisma.ProjectsScalarWhereWithAggregatesInput | Prisma.ProjectsScalarWhereWithAggregatesInput[]
  id?: Prisma.StringWithAggregatesFilter<"Projects"> | string
  chamaId?: Prisma.StringWithAggregatesFilter<"Projects"> | string
  projectName?: Prisma.StringWithAggregatesFilter<"Projects"> | string
  estimatedCost?: Prisma.FloatWithAggregatesFilter<"Projects"> | number
  actualCost?: Prisma.FloatWithAggregatesFilter<"Projects"> | number
  sharesUsed?: Prisma.FloatWithAggregatesFilter<"Projects"> | number
  createdAt?: Prisma.DateTimeWithAggregatesFilter<"Projects"> | Date | string
  updatedAt?: Prisma.DateTimeWithAggregatesFilter<"Projects"> | Date | string
}

export type ProjectsCreateInput = {
  id?: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
  chama: Prisma.ChamaCreateNestedOneWithoutProjectsInput
}

export type ProjectsUncheckedCreateInput = {
  id?: string
  chamaId: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ProjectsUpdateInput = {
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  chama?: Prisma.ChamaUpdateOneRequiredWithoutProjectsNestedInput
}

export type ProjectsUncheckedUpdateInput = {
  chamaId?: Prisma.StringFieldUpdateOperationsInput | string
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ProjectsCreateManyInput = {
  id?: string
  chamaId: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ProjectsUpdateManyMutationInput = {
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ProjectsUncheckedUpdateManyInput = {
  chamaId?: Prisma.StringFieldUpdateOperationsInput | string
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ProjectsListRelationFilter = {
  every?: Prisma.ProjectsWhereInput
  some?: Prisma.ProjectsWhereInput
  none?: Prisma.ProjectsWhereInput
}

export type ProjectsOrderByRelationAggregateInput = {
  _count?: Prisma.SortOrder
}

export type ProjectsCountOrderByAggregateInput = {
  id?: Prisma.SortOrder
  chamaId?: Prisma.SortOrder
  projectName?: Prisma.SortOrder
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ProjectsAvgOrderByAggregateInput = {
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
}

export type ProjectsMaxOrderByAggregateInput = {
  id?: Prisma.SortOrder
  chamaId?: Prisma.SortOrder
  projectName?: Prisma.SortOrder
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ProjectsMinOrderByAggregateInput = {
  id?: Prisma.SortOrder
  chamaId?: Prisma.SortOrder
  projectName?: Prisma.SortOrder
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
  createdAt?: Prisma.SortOrder
  updatedAt?: Prisma.SortOrder
}

export type ProjectsSumOrderByAggregateInput = {
  estimatedCost?: Prisma.SortOrder
  actualCost?: Prisma.SortOrder
  sharesUsed?: Prisma.SortOrder
}

export type ProjectsCreateNestedManyWithoutChamaInput = {
  create?: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput> | Prisma.ProjectsCreateWithoutChamaInput[] | Prisma.ProjectsUncheckedCreateWithoutChamaInput[]
  connectOrCreate?: Prisma.ProjectsCreateOrConnectWithoutChamaInput | Prisma.ProjectsCreateOrConnectWithoutChamaInput[]
  createMany?: Prisma.ProjectsCreateManyChamaInputEnvelope
  connect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
}

export type ProjectsUncheckedCreateNestedManyWithoutChamaInput = {
  create?: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput> | Prisma.ProjectsCreateWithoutChamaInput[] | Prisma.ProjectsUncheckedCreateWithoutChamaInput[]
  connectOrCreate?: Prisma.ProjectsCreateOrConnectWithoutChamaInput | Prisma.ProjectsCreateOrConnectWithoutChamaInput[]
  createMany?: Prisma.ProjectsCreateManyChamaInputEnvelope
  connect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
}

export type ProjectsUpdateManyWithoutChamaNestedInput = {
  create?: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput> | Prisma.ProjectsCreateWithoutChamaInput[] | Prisma.ProjectsUncheckedCreateWithoutChamaInput[]
  connectOrCreate?: Prisma.ProjectsCreateOrConnectWithoutChamaInput | Prisma.ProjectsCreateOrConnectWithoutChamaInput[]
  upsert?: Prisma.ProjectsUpsertWithWhereUniqueWithoutChamaInput | Prisma.ProjectsUpsertWithWhereUniqueWithoutChamaInput[]
  createMany?: Prisma.ProjectsCreateManyChamaInputEnvelope
  set?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  disconnect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  delete?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  connect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  update?: Prisma.ProjectsUpdateWithWhereUniqueWithoutChamaInput | Prisma.ProjectsUpdateWithWhereUniqueWithoutChamaInput[]
  updateMany?: Prisma.ProjectsUpdateManyWithWhereWithoutChamaInput | Prisma.ProjectsUpdateManyWithWhereWithoutChamaInput[]
  deleteMany?: Prisma.ProjectsScalarWhereInput | Prisma.ProjectsScalarWhereInput[]
}

export type ProjectsUncheckedUpdateManyWithoutChamaNestedInput = {
  create?: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput> | Prisma.ProjectsCreateWithoutChamaInput[] | Prisma.ProjectsUncheckedCreateWithoutChamaInput[]
  connectOrCreate?: Prisma.ProjectsCreateOrConnectWithoutChamaInput | Prisma.ProjectsCreateOrConnectWithoutChamaInput[]
  upsert?: Prisma.ProjectsUpsertWithWhereUniqueWithoutChamaInput | Prisma.ProjectsUpsertWithWhereUniqueWithoutChamaInput[]
  createMany?: Prisma.ProjectsCreateManyChamaInputEnvelope
  set?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  disconnect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  delete?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  connect?: Prisma.ProjectsWhereUniqueInput | Prisma.ProjectsWhereUniqueInput[]
  update?: Prisma.ProjectsUpdateWithWhereUniqueWithoutChamaInput | Prisma.ProjectsUpdateWithWhereUniqueWithoutChamaInput[]
  updateMany?: Prisma.ProjectsUpdateManyWithWhereWithoutChamaInput | Prisma.ProjectsUpdateManyWithWhereWithoutChamaInput[]
  deleteMany?: Prisma.ProjectsScalarWhereInput | Prisma.ProjectsScalarWhereInput[]
}

export type ProjectsCreateWithoutChamaInput = {
  id?: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ProjectsUncheckedCreateWithoutChamaInput = {
  id?: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ProjectsCreateOrConnectWithoutChamaInput = {
  where: Prisma.ProjectsWhereUniqueInput
  create: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput>
}

export type ProjectsCreateManyChamaInputEnvelope = {
  data: Prisma.ProjectsCreateManyChamaInput | Prisma.ProjectsCreateManyChamaInput[]
}

export type ProjectsUpsertWithWhereUniqueWithoutChamaInput = {
  where: Prisma.ProjectsWhereUniqueInput
  update: Prisma.XOR<Prisma.ProjectsUpdateWithoutChamaInput, Prisma.ProjectsUncheckedUpdateWithoutChamaInput>
  create: Prisma.XOR<Prisma.ProjectsCreateWithoutChamaInput, Prisma.ProjectsUncheckedCreateWithoutChamaInput>
}

export type ProjectsUpdateWithWhereUniqueWithoutChamaInput = {
  where: Prisma.ProjectsWhereUniqueInput
  data: Prisma.XOR<Prisma.ProjectsUpdateWithoutChamaInput, Prisma.ProjectsUncheckedUpdateWithoutChamaInput>
}

export type ProjectsUpdateManyWithWhereWithoutChamaInput = {
  where: Prisma.ProjectsScalarWhereInput
  data: Prisma.XOR<Prisma.ProjectsUpdateManyMutationInput, Prisma.ProjectsUncheckedUpdateManyWithoutChamaInput>
}

export type ProjectsScalarWhereInput = {
  AND?: Prisma.ProjectsScalarWhereInput | Prisma.ProjectsScalarWhereInput[]
  OR?: Prisma.ProjectsScalarWhereInput[]
  NOT?: Prisma.ProjectsScalarWhereInput | Prisma.ProjectsScalarWhereInput[]
  id?: Prisma.StringFilter<"Projects"> | string
  chamaId?: Prisma.StringFilter<"Projects"> | string
  projectName?: Prisma.StringFilter<"Projects"> | string
  estimatedCost?: Prisma.FloatFilter<"Projects"> | number
  actualCost?: Prisma.FloatFilter<"Projects"> | number
  sharesUsed?: Prisma.FloatFilter<"Projects"> | number
  createdAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
  updatedAt?: Prisma.DateTimeFilter<"Projects"> | Date | string
}

export type ProjectsCreateManyChamaInput = {
  id?: string
  projectName: string
  estimatedCost?: number
  actualCost?: number
  sharesUsed?: number
  createdAt?: Date | string
  updatedAt?: Date | string
}

export type ProjectsUpdateWithoutChamaInput = {
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ProjectsUncheckedUpdateWithoutChamaInput = {
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}

export type ProjectsUncheckedUpdateManyWithoutChamaInput = {
  projectName?: Prisma.StringFieldUpdateOperationsInput | string
  estimatedCost?: Prisma.FloatFieldUpdateOperationsInput | number
  actualCost?: Prisma.FloatFieldUpdateOperationsInput | number
  sharesUsed?: Prisma.FloatFieldUpdateOperationsInput | number
  createdAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
  updatedAt?: Prisma.DateTimeFieldUpdateOperationsInput | Date | string
}



export type ProjectsSelect<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetSelect<{
  id?: boolean
  chamaId?: boolean
  projectName?: boolean
  estimatedCost?: boolean
  actualCost?: boolean
  sharesUsed?: boolean
  createdAt?: boolean
  updatedAt?: boolean
  chama?: boolean | Prisma.ChamaDefaultArgs<ExtArgs>
}, ExtArgs["result"]["projects"]>



export type ProjectsSelectScalar = {
  id?: boolean
  chamaId?: boolean
  projectName?: boolean
  estimatedCost?: boolean
  actualCost?: boolean
  sharesUsed?: boolean
  createdAt?: boolean
  updatedAt?: boolean
}

export type ProjectsOmit<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = runtime.Types.Extensions.GetOmit<"id" | "chamaId" | "projectName" | "estimatedCost" | "actualCost" | "sharesUsed" | "createdAt" | "updatedAt", ExtArgs["result"]["projects"]>
export type ProjectsInclude<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  chama?: boolean | Prisma.ChamaDefaultArgs<ExtArgs>
}

export type $ProjectsPayload<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  name: "Projects"
  objects: {
    chama: Prisma.$ChamaPayload<ExtArgs>
  }
  scalars: runtime.Types.Extensions.GetPayloadResult<{
    id: string
    chamaId: string
    projectName: string
    estimatedCost: number
    actualCost: number
    sharesUsed: number
    createdAt: Date
    updatedAt: Date
  }, ExtArgs["result"]["projects"]>
  composites: {}
}

export type ProjectsGetPayload<S extends boolean | null | undefined | ProjectsDefaultArgs> = runtime.Types.Result.GetResult<Prisma.$ProjectsPayload, S>

export type ProjectsCountArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> =
  Omit<ProjectsFindManyArgs, 'select' | 'include' | 'distinct' | 'omit'> & {
    select?: ProjectsCountAggregateInputType | true
  }

export interface ProjectsDelegate<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> {
  [K: symbol]: { types: Prisma.TypeMap<ExtArgs>['model']['Projects'], meta: { name: 'Projects' } }
  /**
   * Find zero or one Projects that matches the filter.
   * @param {ProjectsFindUniqueArgs} args - Arguments to find a Projects
   * @example
   * // Get one Projects
   * const projects = await prisma.projects.findUnique({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUnique<T extends ProjectsFindUniqueArgs>(args: Prisma.SelectSubset<T, ProjectsFindUniqueArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findUnique", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find one Projects that matches the filter or throw an error with `error.code='P2025'`
   * if no matches were found.
   * @param {ProjectsFindUniqueOrThrowArgs} args - Arguments to find a Projects
   * @example
   * // Get one Projects
   * const projects = await prisma.projects.findUniqueOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findUniqueOrThrow<T extends ProjectsFindUniqueOrThrowArgs>(args: Prisma.SelectSubset<T, ProjectsFindUniqueOrThrowArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Projects that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsFindFirstArgs} args - Arguments to find a Projects
   * @example
   * // Get one Projects
   * const projects = await prisma.projects.findFirst({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirst<T extends ProjectsFindFirstArgs>(args?: Prisma.SelectSubset<T, ProjectsFindFirstArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findFirst", GlobalOmitOptions> | null, null, ExtArgs, GlobalOmitOptions>

  /**
   * Find the first Projects that matches the filter or
   * throw `PrismaKnownClientError` with `P2025` code if no matches were found.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsFindFirstOrThrowArgs} args - Arguments to find a Projects
   * @example
   * // Get one Projects
   * const projects = await prisma.projects.findFirstOrThrow({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   */
  findFirstOrThrow<T extends ProjectsFindFirstOrThrowArgs>(args?: Prisma.SelectSubset<T, ProjectsFindFirstOrThrowArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findFirstOrThrow", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Projects that matches the filter.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsFindManyArgs} args - Arguments to filter and select certain fields only.
   * @example
   * // Get all Projects
   * const projects = await prisma.projects.findMany()
   * 
   * // Get first 10 Projects
   * const projects = await prisma.projects.findMany({ take: 10 })
   * 
   * // Only select the `id`
   * const projectsWithIdOnly = await prisma.projects.findMany({ select: { id: true } })
   * 
   */
  findMany<T extends ProjectsFindManyArgs>(args?: Prisma.SelectSubset<T, ProjectsFindManyArgs<ExtArgs>>): Prisma.PrismaPromise<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "findMany", GlobalOmitOptions>>

  /**
   * Create a Projects.
   * @param {ProjectsCreateArgs} args - Arguments to create a Projects.
   * @example
   * // Create one Projects
   * const Projects = await prisma.projects.create({
   *   data: {
   *     // ... data to create a Projects
   *   }
   * })
   * 
   */
  create<T extends ProjectsCreateArgs>(args: Prisma.SelectSubset<T, ProjectsCreateArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "create", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Create many Projects.
   * @param {ProjectsCreateManyArgs} args - Arguments to create many Projects.
   * @example
   * // Create many Projects
   * const projects = await prisma.projects.createMany({
   *   data: [
   *     // ... provide data here
   *   ]
   * })
   *     
   */
  createMany<T extends ProjectsCreateManyArgs>(args?: Prisma.SelectSubset<T, ProjectsCreateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Delete a Projects.
   * @param {ProjectsDeleteArgs} args - Arguments to delete one Projects.
   * @example
   * // Delete one Projects
   * const Projects = await prisma.projects.delete({
   *   where: {
   *     // ... filter to delete one Projects
   *   }
   * })
   * 
   */
  delete<T extends ProjectsDeleteArgs>(args: Prisma.SelectSubset<T, ProjectsDeleteArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "delete", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Update one Projects.
   * @param {ProjectsUpdateArgs} args - Arguments to update one Projects.
   * @example
   * // Update one Projects
   * const projects = await prisma.projects.update({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  update<T extends ProjectsUpdateArgs>(args: Prisma.SelectSubset<T, ProjectsUpdateArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "update", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Delete zero or more Projects.
   * @param {ProjectsDeleteManyArgs} args - Arguments to filter Projects to delete.
   * @example
   * // Delete a few Projects
   * const { count } = await prisma.projects.deleteMany({
   *   where: {
   *     // ... provide filter here
   *   }
   * })
   * 
   */
  deleteMany<T extends ProjectsDeleteManyArgs>(args?: Prisma.SelectSubset<T, ProjectsDeleteManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Update zero or more Projects.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsUpdateManyArgs} args - Arguments to update one or more rows.
   * @example
   * // Update many Projects
   * const projects = await prisma.projects.updateMany({
   *   where: {
   *     // ... provide filter here
   *   },
   *   data: {
   *     // ... provide data here
   *   }
   * })
   * 
   */
  updateMany<T extends ProjectsUpdateManyArgs>(args: Prisma.SelectSubset<T, ProjectsUpdateManyArgs<ExtArgs>>): Prisma.PrismaPromise<Prisma.BatchPayload>

  /**
   * Create or update one Projects.
   * @param {ProjectsUpsertArgs} args - Arguments to update or create a Projects.
   * @example
   * // Update or create a Projects
   * const projects = await prisma.projects.upsert({
   *   create: {
   *     // ... data to create a Projects
   *   },
   *   update: {
   *     // ... in case it already exists, update
   *   },
   *   where: {
   *     // ... the filter for the Projects we want to update
   *   }
   * })
   */
  upsert<T extends ProjectsUpsertArgs>(args: Prisma.SelectSubset<T, ProjectsUpsertArgs<ExtArgs>>): Prisma.Prisma__ProjectsClient<runtime.Types.Result.GetResult<Prisma.$ProjectsPayload<ExtArgs>, T, "upsert", GlobalOmitOptions>, never, ExtArgs, GlobalOmitOptions>

  /**
   * Find zero or more Projects that matches the filter.
   * @param {ProjectsFindRawArgs} args - Select which filters you would like to apply.
   * @example
   * const projects = await prisma.projects.findRaw({
   *   filter: { age: { $gt: 25 } }
   * })
   */
  findRaw(args?: Prisma.ProjectsFindRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>

  /**
   * Perform aggregation operations on a Projects.
   * @param {ProjectsAggregateRawArgs} args - Select which aggregations you would like to apply.
   * @example
   * const projects = await prisma.projects.aggregateRaw({
   *   pipeline: [
   *     { $match: { status: "registered" } },
   *     { $group: { _id: "$country", total: { $sum: 1 } } }
   *   ]
   * })
   */
  aggregateRaw(args?: Prisma.ProjectsAggregateRawArgs): Prisma.PrismaPromise<Prisma.JsonObject>


  /**
   * Count the number of Projects.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsCountArgs} args - Arguments to filter Projects to count.
   * @example
   * // Count the number of Projects
   * const count = await prisma.projects.count({
   *   where: {
   *     // ... the filter for the Projects we want to count
   *   }
   * })
  **/
  count<T extends ProjectsCountArgs>(
    args?: Prisma.Subset<T, ProjectsCountArgs>,
  ): Prisma.PrismaPromise<
    T extends runtime.Types.Utils.Record<'select', any>
      ? T['select'] extends true
        ? number
        : Prisma.GetScalarType<T['select'], ProjectsCountAggregateOutputType>
      : number
  >

  /**
   * Allows you to perform aggregations operations on a Projects.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsAggregateArgs} args - Select which aggregations you would like to apply and on what fields.
   * @example
   * // Ordered by age ascending
   * // Where email contains prisma.io
   * // Limited to the 10 users
   * const aggregations = await prisma.user.aggregate({
   *   _avg: {
   *     age: true,
   *   },
   *   where: {
   *     email: {
   *       contains: "prisma.io",
   *     },
   *   },
   *   orderBy: {
   *     age: "asc",
   *   },
   *   take: 10,
   * })
  **/
  aggregate<T extends ProjectsAggregateArgs>(args: Prisma.Subset<T, ProjectsAggregateArgs>): Prisma.PrismaPromise<GetProjectsAggregateType<T>>

  /**
   * Group by Projects.
   * Note, that providing `undefined` is treated as the value not being there.
   * Read more here: https://pris.ly/d/null-undefined
   * @param {ProjectsGroupByArgs} args - Group by arguments.
   * @example
   * // Group by city, order by createdAt, get count
   * const result = await prisma.user.groupBy({
   *   by: ['city', 'createdAt'],
   *   orderBy: {
   *     createdAt: true
   *   },
   *   _count: {
   *     _all: true
   *   },
   * })
   * 
  **/
  groupBy<
    T extends ProjectsGroupByArgs,
    HasSelectOrTake extends Prisma.Or<
      Prisma.Extends<'skip', Prisma.Keys<T>>,
      Prisma.Extends<'take', Prisma.Keys<T>>
    >,
    OrderByArg extends Prisma.True extends HasSelectOrTake
      ? { orderBy: ProjectsGroupByArgs['orderBy'] }
      : { orderBy?: ProjectsGroupByArgs['orderBy'] },
    OrderFields extends Prisma.ExcludeUnderscoreKeys<Prisma.Keys<Prisma.MaybeTupleToUnion<T['orderBy']>>>,
    ByFields extends Prisma.MaybeTupleToUnion<T['by']>,
    ByValid extends Prisma.Has<ByFields, OrderFields>,
    HavingFields extends Prisma.GetHavingFields<T['having']>,
    HavingValid extends Prisma.Has<ByFields, HavingFields>,
    ByEmpty extends T['by'] extends never[] ? Prisma.True : Prisma.False,
    InputErrors extends ByEmpty extends Prisma.True
    ? `Error: "by" must not be empty.`
    : HavingValid extends Prisma.False
    ? {
        [P in HavingFields]: P extends ByFields
          ? never
          : P extends string
          ? `Error: Field "${P}" used in "having" needs to be provided in "by".`
          : [
              Error,
              'Field ',
              P,
              ` in "having" needs to be provided in "by"`,
            ]
      }[HavingFields]
    : 'take' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "take", you also need to provide "orderBy"'
    : 'skip' extends Prisma.Keys<T>
    ? 'orderBy' extends Prisma.Keys<T>
      ? ByValid extends Prisma.True
        ? {}
        : {
            [P in OrderFields]: P extends ByFields
              ? never
              : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
          }[OrderFields]
      : 'Error: If you provide "skip", you also need to provide "orderBy"'
    : ByValid extends Prisma.True
    ? {}
    : {
        [P in OrderFields]: P extends ByFields
          ? never
          : `Error: Field "${P}" in "orderBy" needs to be provided in "by"`
      }[OrderFields]
  >(args: Prisma.SubsetIntersection<T, ProjectsGroupByArgs, OrderByArg> & InputErrors): {} extends InputErrors ? GetProjectsGroupByPayload<T> : Prisma.PrismaPromise<InputErrors>
/**
 * Fields of the Projects model
 */
readonly fields: ProjectsFieldRefs;
}

/**
 * The delegate class that acts as a "Promise-like" for Projects.
 * Why is this prefixed with `Prisma__`?
 * Because we want to prevent naming conflicts as mentioned in
 * https://github.com/prisma/prisma-client-js/issues/707
 */
export interface Prisma__ProjectsClient<T, Null = never, ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs, GlobalOmitOptions = {}> extends Prisma.PrismaPromise<T> {
  readonly [Symbol.toStringTag]: "PrismaPromise"
  chama<T extends Prisma.ChamaDefaultArgs<ExtArgs> = {}>(args?: Prisma.Subset<T, Prisma.ChamaDefaultArgs<ExtArgs>>): Prisma.Prisma__ChamaClient<runtime.Types.Result.GetResult<Prisma.$ChamaPayload<ExtArgs>, T, "findUniqueOrThrow", GlobalOmitOptions> | Null, Null, ExtArgs, GlobalOmitOptions>
  /**
   * Attaches callbacks for the resolution and/or rejection of the Promise.
   * @param onfulfilled The callback to execute when the Promise is resolved.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of which ever callback is executed.
   */
  then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): runtime.Types.Utils.JsPromise<TResult1 | TResult2>
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): runtime.Types.Utils.JsPromise<T | TResult>
  /**
   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
   * resolved value cannot be modified from the callback.
   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
   * @returns A Promise for the completion of the callback.
   */
  finally(onfinally?: (() => void) | undefined | null): runtime.Types.Utils.JsPromise<T>
}




/**
 * Fields of the Projects model
 */
export interface ProjectsFieldRefs {
  readonly id: Prisma.FieldRef<"Projects", 'String'>
  readonly chamaId: Prisma.FieldRef<"Projects", 'String'>
  readonly projectName: Prisma.FieldRef<"Projects", 'String'>
  readonly estimatedCost: Prisma.FieldRef<"Projects", 'Float'>
  readonly actualCost: Prisma.FieldRef<"Projects", 'Float'>
  readonly sharesUsed: Prisma.FieldRef<"Projects", 'Float'>
  readonly createdAt: Prisma.FieldRef<"Projects", 'DateTime'>
  readonly updatedAt: Prisma.FieldRef<"Projects", 'DateTime'>
}
    

// Custom InputTypes
/**
 * Projects findUnique
 */
export type ProjectsFindUniqueArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter, which Projects to fetch.
   */
  where: Prisma.ProjectsWhereUniqueInput
}

/**
 * Projects findUniqueOrThrow
 */
export type ProjectsFindUniqueOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter, which Projects to fetch.
   */
  where: Prisma.ProjectsWhereUniqueInput
}

/**
 * Projects findFirst
 */
export type ProjectsFindFirstArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter, which Projects to fetch.
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Projects to fetch.
   */
  orderBy?: Prisma.ProjectsOrderByWithRelationInput | Prisma.ProjectsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Projects.
   */
  cursor?: Prisma.ProjectsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Projects from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Projects.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Projects.
   */
  distinct?: Prisma.ProjectsScalarFieldEnum | Prisma.ProjectsScalarFieldEnum[]
}

/**
 * Projects findFirstOrThrow
 */
export type ProjectsFindFirstOrThrowArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter, which Projects to fetch.
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Projects to fetch.
   */
  orderBy?: Prisma.ProjectsOrderByWithRelationInput | Prisma.ProjectsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for searching for Projects.
   */
  cursor?: Prisma.ProjectsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Projects from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Projects.
   */
  skip?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/distinct Distinct Docs}
   * 
   * Filter by unique combinations of Projects.
   */
  distinct?: Prisma.ProjectsScalarFieldEnum | Prisma.ProjectsScalarFieldEnum[]
}

/**
 * Projects findMany
 */
export type ProjectsFindManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter, which Projects to fetch.
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/sorting Sorting Docs}
   * 
   * Determine the order of Projects to fetch.
   */
  orderBy?: Prisma.ProjectsOrderByWithRelationInput | Prisma.ProjectsOrderByWithRelationInput[]
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination Cursor Docs}
   * 
   * Sets the position for listing Projects.
   */
  cursor?: Prisma.ProjectsWhereUniqueInput
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Take `±n` Projects from the position of the cursor.
   */
  take?: number
  /**
   * {@link https://www.prisma.io/docs/concepts/components/prisma-client/pagination Pagination Docs}
   * 
   * Skip the first `n` Projects.
   */
  skip?: number
  distinct?: Prisma.ProjectsScalarFieldEnum | Prisma.ProjectsScalarFieldEnum[]
}

/**
 * Projects create
 */
export type ProjectsCreateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * The data needed to create a Projects.
   */
  data: Prisma.XOR<Prisma.ProjectsCreateInput, Prisma.ProjectsUncheckedCreateInput>
}

/**
 * Projects createMany
 */
export type ProjectsCreateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to create many Projects.
   */
  data: Prisma.ProjectsCreateManyInput | Prisma.ProjectsCreateManyInput[]
}

/**
 * Projects update
 */
export type ProjectsUpdateArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * The data needed to update a Projects.
   */
  data: Prisma.XOR<Prisma.ProjectsUpdateInput, Prisma.ProjectsUncheckedUpdateInput>
  /**
   * Choose, which Projects to update.
   */
  where: Prisma.ProjectsWhereUniqueInput
}

/**
 * Projects updateMany
 */
export type ProjectsUpdateManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The data used to update Projects.
   */
  data: Prisma.XOR<Prisma.ProjectsUpdateManyMutationInput, Prisma.ProjectsUncheckedUpdateManyInput>
  /**
   * Filter which Projects to update
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * Limit how many Projects to update.
   */
  limit?: number
}

/**
 * Projects upsert
 */
export type ProjectsUpsertArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * The filter to search for the Projects to update in case it exists.
   */
  where: Prisma.ProjectsWhereUniqueInput
  /**
   * In case the Projects found by the `where` argument doesn't exist, create a new Projects with this data.
   */
  create: Prisma.XOR<Prisma.ProjectsCreateInput, Prisma.ProjectsUncheckedCreateInput>
  /**
   * In case the Projects was found with the provided `where` argument, update it with this data.
   */
  update: Prisma.XOR<Prisma.ProjectsUpdateInput, Prisma.ProjectsUncheckedUpdateInput>
}

/**
 * Projects delete
 */
export type ProjectsDeleteArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
  /**
   * Filter which Projects to delete.
   */
  where: Prisma.ProjectsWhereUniqueInput
}

/**
 * Projects deleteMany
 */
export type ProjectsDeleteManyArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Filter which Projects to delete
   */
  where?: Prisma.ProjectsWhereInput
  /**
   * Limit how many Projects to delete.
   */
  limit?: number
}

/**
 * Projects findRaw
 */
export type ProjectsFindRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * The query predicate filter. If unspecified, then all documents in the collection will match the predicate. ${@link https://docs.mongodb.com/manual/reference/operator/query MongoDB Docs}.
   */
  filter?: runtime.InputJsonValue
  /**
   * Additional options to pass to the `find` command ${@link https://docs.mongodb.com/manual/reference/command/find/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Projects aggregateRaw
 */
export type ProjectsAggregateRawArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * An array of aggregation stages to process and transform the document stream via the aggregation pipeline. ${@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline MongoDB Docs}.
   */
  pipeline?: runtime.InputJsonValue[]
  /**
   * Additional options to pass to the `aggregate` command ${@link https://docs.mongodb.com/manual/reference/command/aggregate/#command-fields MongoDB Docs}.
   */
  options?: runtime.InputJsonValue
}

/**
 * Projects without action
 */
export type ProjectsDefaultArgs<ExtArgs extends runtime.Types.Extensions.InternalArgs = runtime.Types.Extensions.DefaultArgs> = {
  /**
   * Select specific fields to fetch from the Projects
   */
  select?: Prisma.ProjectsSelect<ExtArgs> | null
  /**
   * Omit specific fields from the Projects
   */
  omit?: Prisma.ProjectsOmit<ExtArgs> | null
  /**
   * Choose, which related nodes to fetch as well
   */
  include?: Prisma.ProjectsInclude<ExtArgs> | null
}
